Generating plots of gene expression by time course and treatment
================================================================

Review paths to confirm they are still correct:
```{r}
sharedPath <- "/isilon/biodiversity/users/shared/Pythium_ultimum_RNAseq/"
```

Over here Emily to read the big file
```{r}
rpkmPlotsFullImport <- read.table(paste(sharedPath, 
                                        "rpkmForBarPlots_full.csv", 
                                        sep = ""), 
                                  sep = ",", 
                                  header = TRUE, 
                                  as.is = FALSE, 
                                  stringsAsFactors = FALSE)

rpkmPlotsFullImport$xLab <- paste(rpkmPlotsFullImport$Condition,
                                  rpkmPlotsFullImport$TimePoint, 
                                  sep = "_")
unique(rpkmPlotsFullImport$xLab)

rpkmPlotsFullImport$xLab <- sub("OoCTimeCourse_", "", rpkmPlotsFullImport$xLab)
rpkmPlotsFullImport$xLab <- sub("Various_0", "Various", rpkmPlotsFullImport$xLab)
unique(rpkmPlotsFullImport$xLab)
unique(rpkmPlotsFullImport$Experiment)

labOrdered <- c("Various", "Control_0", "Control_3", "Cholesterol_3", 
                "Control_7", "Cholesterol_7", "Control_24", "Cholesterol_24", 
                "0", "12", "24", "48", "72", "120", "240") 
rpkmPlotsFullImport$xLab <- factor(rpkmPlotsFullImport$xLab, levels = c(labOrdered))   
levels(rpkmPlotsFullImport$xLab)

expOrdered <- c("Mycelium", "Oosporogenesis", "OosporeConversion") 
rpkmPlotsFullImport$Experiment <- factor(rpkmPlotsFullImport$Experiment, 
                                         levels = c(expOrdered))   
levels(rpkmPlotsFullImport$Experiment)
```


Add a column in the large data file about being CAZy (secreted or not) on non CAZy
```{r}

mainFolder <- "GGPLOTS_FROM_FULL_DATA_continuous_X/from_Emily_24_March_2017"

library(data.table)
listLoci <- fread(paste(sharedPath, mainFolder, "/CAZy_of_interest_list.csv", sep = ""),
                  sep = "auto", header = TRUE)

listLoci_duplicated <- listLoci[which(duplicated(listLoci[,c(2,4)])),]
listLoci_notduplicated <- listLoci[which(!duplicated(listLoci[,c(2,4)])),]

which(duplicated(listLoci[,c(2,4)]))

rpkmPlotsFullImport <- merge(listLoci_notduplicated[,2:4], rpkmPlotsFullImport[,2:9], by = "SequenceID", all.y = TRUE)

rpkmPlotsFullImport$Secreted[is.na(rpkmPlotsFullImport$Secreted)] <- "non-CAZy"


```


Create Main plot folder
```{r}
mainFolder <- "GGPLOTS_STATS_FROM_FULL_DATA_model_fitting"
dir.create(paste(sharedPath, mainFolder , sep = ""), 
           showWarnings = TRUE, recursive = FALSE)
```

Create a folder for the plot outputs:
```{r}
analysisDir <- "24_March_2017"
dir.create(paste(sharedPath, mainFolder , "/", analysisDir, sep = ""), 
           showWarnings = TRUE, recursive = FALSE)
analysisPath <- paste(sharedPath, mainFolder , "/", analysisDir, sep = "")
```



EdgeR for Oosporogenesis
```{r}

library(edgeR)

# select which experiment to analyze
Exp <- levels(rpkmPlotsFullImport$Experiment)[2]
Exp
CAZy <- levels(as.factor(rpkmPlotsFullImport$Secreted))[2]

#Get only certain columns and oosporogenesis experiment
subset_count <-  rpkmPlotsFullImport[which(rpkmPlotsFullImport$Experiment==Exp & rpkmPlotsFullImport$Secreted!="non-CAZy"),]

subset_count$TRT <- paste(subset_count$Condition,subset_count$TimePoint,subset_count$RNA_Replicate,sep="_")

# make table in proper format
my_counts <- dcast(data = subset_count, SequenceID ~ Condition + TimePoint + RNA_Replicate, value.var = "Read_Counts")
rownames(my_counts) <- my_counts$SequenceID
my_counts$SequenceID <- NULL

is.data.frame(my_counts)


colSums( my_counts ) # Library Sizes
colSums( my_counts ) / 1e06 # Library Sizes in millions of reads
table( rowSums( my_counts ) )[ 1:30 ]

Condition <- sub("_.*","",colnames(my_counts))
TimePoint <- sub("Control_|Cholesterol_","",colnames(my_counts))
TimePoint <- sub("_.*","",TimePoint)
CondTime  <- sub("_[0-9]$","",colnames(my_counts))


cds2 <- DGEList( my_counts ,  group = Condition )
cds2 <- DGEList( my_counts ,  group = CondTime )
# Somehow rownames did not follow
dimnames(cds2)[[1]] <- rownames(my_counts)

names( cds2 )
sum( cds2$all.zeros )

cds2$samples$lib.size

cds2 <- estimateCommonDisp( cds2 )
names( cds2 )
# The estimate
cds2$common.dispersion
# 
# # Default Setting
# cds2 <- estimateTagwiseDisp( cds2  )
# names( cds2 )
# summary( cds$tagwise.dispersion )
# # More shrinkage/sqeezing toward the common
# cds2 <- estimateTagwiseDisp( cds2 , prior.n = 25 )
# summary( cds2$tagwise.dispersion ) # not much changed, but the ends got squeezed in quite a bit.
# # # The recommended setting for this data set is the default of 10. Let's stick with that.
# cds2 <- estimateTagwiseDisp( cds2 , prior.n = 10 )
# 
# de.cmn2 <- exactTest( cds2 , common.disp = TRUE )
# de.tgw2 <- exactTest( cds2 , common.disp = FALSE , pair = c( "Control" , "Cholesterol" ) )
# de.poi2 <- exactTest( cds2 , dispersion = 1e-06  , pair = c( "Control" , "Cholesterol" ) )
# et <- exactTest(cds2)
# topTags(et)

# from edgeR User Guide

design <- model.matrix(~CondTime -1)
#design <- model.matrix(~Condition)

cds2 <- estimateDisp(cds2,design)

fit <- glmQLFit(cds2, design )
qlf <- glmQLFTest(fit)
topTags(qlf)

Fit_results2 <- data.frame(qlf$coefficients, qlf$table)

write.table(Fit_results2, paste(sharedPath,mainFolder,"/",analysisDir, "/", Exp,"EdgeR_glmQLFit_CAZy_Anyone_diff_0.csv", sep=""), append = FALSE, quote = TRUE, sep = ",", col.names = NA, row.names = TRUE) 


fit2 <- glmFit(cds2,design) 
lrt <- glmLRT(fit2)
topTags(lrt)

Fit_results <- data.frame(lrt$coefficients, lrt$table)
                          
write.table(Fit_results, paste(sharedPath,mainFolder,"/",analysisDir, "/", Exp, "EdgeR_glmFit_CAZy_Anyone_diff_0.csv", sep=""), append = FALSE, quote = TRUE, sep = ",", col.names = NA, row.names = TRUE)

cds2 <- estimateGLMCommonDisp(cds2, design)
#To estimate trended dispersions:
cds2 <- estimateGLMTrendedDisp(cds2, design)
#To estimate tagwise dispersions:
cds2 <- estimateGLMTagwiseDisp(cds2, design)


```



Get a subset for my own GLM analysis
```{r}
library(data.table)
# #Get only certain columns and oosporogenesis experiment
# subset_count <-  rpkmPlotsFullImport[which(rpkmPlotsFullImport$Experiment=="Oosporogenesis"),c("Experiment", "SequenceID", "Condition", "TimePoint", "RNA_Replicate","Read_Counts")]
# 
# # get the secreted subset
# subset_count <- subset_count[subset_count$SequenceID %in% unique(Secreted_CAZy$sequence),]

my_counts_GLM <- dcast(data = subset_count, Condition + TimePoint + RNA_Replicate ~ SequenceID, value.var = "Read_Counts")

sub_my_counts_GLM <-  my_counts_GLM[ , c("Condition", "TimePoint", "RNA_Replicate", "PYU1_T005443", "PYU1_T007490")]

sub_my_counts_GLM$Condition <- as.factor(sub_my_counts_GLM$Condition)

fit2 <- glm(PYU1_T005443 ~ Condition, family = poisson(link="log"), sub_my_counts_GLM)

print(anova(fit2, test = "Chisq")) 
# Defines the fitted model
print(summary(fit2))


fit3 <- glm(PYU1_T005443 ~ Condition, family = quasipoisson, sub_my_counts_GLM)

print(anova(fit3, test = "Chisq")) 

anova3 <- anova(fit3, test = "Chisq")
# Defines the fitted model
coeff3 <- coefficients(summary(fit3))

cbind(anova3,rownames(coeff3),coeff3)


```






```{r}

All_CAZy <- read.table("data/dbCAN_Subset_AllRefs_columnSubset.csv", 
                                  sep = ",", 
                                  header = TRUE, 
                                  as.is = FALSE, 
                                  stringsAsFactors = FALSE)

All_CAZy_names <- unique(All_CAZy[which(All_CAZy$species=="Pythium_ultimum_var_ultimum"),c("sequence")])

```

** Non linear modelling

Get a subset for analysis
```{r}
library(data.table)
#Get only certain columns and oosporogenesis experiment
subset_count2 <-  rpkmPlotsFullImport[which(rpkmPlotsFullImport$Experiment=="OosporeConversion"),c("Experiment", "SequenceID", "Condition", "TimePoint", "RNA_Replicate","Read_Counts")]

# get the secreted subset
subset_count2 <- subset_count2[subset_count2$SequenceID %in% All_CAZy_names,]

my_counts_NL <- dcast(data = subset_count2, Condition + TimePoint + RNA_Replicate ~ SequenceID, value.var = "Read_Counts")

my_counts_NL$sqrt_time <- sqrt(my_counts_NL$TimePoint)


```


Gaussian fit (see https://en.wikipedia.org/wiki/Gaussian_function)

```{r}
library("nlsMicrobio")
library(nlstools)
sub_my_counts_NL$log_PYU1_T005252 <- log(sub_my_counts_NL$PYU1_T005252 +1)

   NL_fit <- nls(PYU1_T005252 ~   exp(max*exp(-((sqrt_time - cent)^2) / (2 * StDev^2))), 
              start=list(max = 1, cent = 2, StDev =1), data=sub_my_counts_NL, trace=TRUE)

      summary(NL_fit)
   plotfit(NL_fit, smooth = TRUE)

   overview(NL_fit)
   
   coefficients(summary(NL_fit))["max",1]


t <- seq(from = 1, to = 16, by = 0.2)
fitted <- exp(coefficients(summary(NL_fit))["max",1]*exp(-((t - coefficients(summary(NL_fit))["cent",1])^2) / 
        (2 * coefficients(summary(NL_fit))["StDev",1]^2)))
plot(t,fitted)
   

```



```{r}

#dat_coeff3 <- data.frame()
remove(NL_fit)
remove(Temporary3)
ii <- 92

  for(ii in 192:(ncol(my_counts_NL)-1)) { 
    print("######################################################################################################")   
    print(c(ii, colnames(my_counts_NL)[ii]))
    
# upper_max <- 5 * log(max(my_counts_NL[,ii]))
# start_max <- log(mean(my_counts_NL[,ii]))
 start_StDev <- sd(my_counts_NL[,ii])
# start_cent <- log(my_counts_NL[which.max(my_counts_NL[,ii]), 2] +1)
 
  whichmax <- my_counts_NL[,c(2,ii)]
  whichmax <- whichmax[order(-whichmax[,2]),] 
  whichmax$log_count  <- log(whichmax[,2] + 1)
  whichmax_sub <- dcast(whichmax[1:6, c(1,3),], TimePoint ~ . , value.var = "log_count", mean)

 start_cent  <- mean(whichmax_sub[,2])
 start_max  <- mean(whichmax$log_count[1:8])
 upper_max  <- 5 * start_max

 try(
   NL_fit <- nls(my_counts_NL[,ii] ~   exp(max*exp(-((sqrt_time - cent)^2) / (2 * StDev^2))),
              algorithm = "port",
              start=list(max = start_max, cent = start_cent, StDev =start_StDev), 
              lower= list(max = 0, cent = 0, StDev =0.001* start_StDev),
              upper= list(max = upper_max, cent = 16, StDev =100 *start_StDev),
#              weights = 1/my_counts_NL[,ii],
                data=my_counts_NL, trace=TRUE)
, silent=FALSE)

  if (class(NL_fit) != "try-error")  {

    #    summary(NL_fit)
    # plotfit(NL_fit, smooth = TRUE)
    # overview(NL_fit)
    # 
    # coefficients(summary(NL_fit))

    Temporary3 <- data.frame(colnames(my_counts_NL)[ii],row.names(coefficients(summary(NL_fit))),coefficients(summary(NL_fit)))
    dat_coeff3 <- rbind(dat_coeff3,Temporary3) 
    
  } else {
   Temporary3 <- data.frame(colnames(my_counts_NL)[ii],  "NA", "NA",  "NA", "NA", "NA")
   colnames(Temporary3) <- colnames(dat_coeff3)
    dat_coeff3 <- rbind(dat_coeff3,Temporary3) 
 }
  }

```


```{r}
write.table(dat_coeff3, paste(sharedPath,mainFolder,"/coefficients_conversion_CAZy_output.csv", sep=""), append = FALSE, quote = FALSE, sep = ",",col.names = TRUE)
```


EdgeR tutorial at https://cgrlucb.wikispaces.com/file/view/edgeR_Tutorial.pdf
```{r}
library(edgeR)
pnas_file <- "https://sites.google.com/site/davismcc/useful-documents/pnas_expression.txt?attredirects=0&d=1"

raw.data <- read.table( file = pnas_file , header = TRUE )

counts <- raw.data[ , -c(1,ncol(raw.data)) ]

is.data.frame(counts)

rownames( counts ) <- raw.data[ , 1 ] # gene names

colnames( counts ) <- paste(c(rep("C_R",4),rep("T_R",3)),c(1:4,1:3),sep="") # sample names

dim( counts )
colSums( counts ) # Library Sizes
colSums( counts ) / 1e06 # Library Sizes in millions of reads
table( rowSums( counts ) )[ 1:30 ] # Number of genes with low counts

group <- c(rep("C", 4) , rep("T", 3))
cds <- DGEList( counts , group = group )
names( cds )
dimnames(cds)
head(cds$samples)
head(cds$counts) # original count matrix
cds$samples # contains a summary of your samples
sum( cds$all.zeros ) # How many genes have 0 counts across all samples
cds # or type the name of the object

cds <- cds[rowSums(1e+06 * cds$counts/expandAsMatrix(cds$samples$lib.size, dim(cds)) > 1) >= 3, ]
dim( cds )
cds <- calcNormFactors( cds )
cds$samples
# effective library sizes
cds$samples$lib.size * cds$samples$norm.factors

cds <- estimateCommonDisp( cds )
names( cds )
# The estimate
cds$common.dispersion


```






Dump list file as single column and read it.  
The list could have 2 column, one with better labels.
List was sent by e-mail we think
Use next chunk instead
```{r}
# #listName <- "Loci_list.txt"
# #listName <- "listRefCandidates.txt"
# listName <- "CAZy_of_interest_list"
# 
# library(data.table)
# listLoci <- fread(paste(analysisPath, "/", listName, "", sep = ""), 
#                   sep = "\t", header = TRUE,  
#                   stringsAsFactors = FALSE)
# colnames(listLoci)[1] <- "SequenceID"
# colnames(listLoci)[2] <- "CAZyFamily"
# 
# #listLoci$geneInfo <- paste(listLoci$SequenceID, listLoci$CAZyFamily, sep = " -\n ")
```



Read CAZy of entire recent run of dbcan pipeline (from March 2016 version)
```{r}

library(data.table)

cazySecreted <- fread(file = "data/dbCAN_SecretedPyuu_Final_subsetColumns.csv", sep = "auto", header = TRUE, fill=TRUE)

```



Single file, multiple Bar plots for gene in listLoci
```{r}
library("ggplot2")

svg(file   = paste(analysisPath, "/", listName,  ".svg", sep = ""),
    width  = 8, 
    height = 7*ceiling(nrow(temp)/200))
    # 200 in seems to be the maximum length for a pdf file, so I used sgv
    #height = 200)

bp <- ggplot(aes(y = Read_Counts, x = xLab), data  = temp) +
  theme_bw() +
  theme(axis.text.y = element_text(size  = 6, 
                                   hjust = 1, 
                                   vjust = 0.4)) +
  theme(axis.text.x = element_text(colour = 'black', 
                                   size   = 6, 
                                   angle  = 45, 
                                   hjust  = 1, 
                                   vjust  = 1)) +
  theme(axis.ticks = element_line(colour = 'black', 
                                  size   = 0.5)) +
  labs(y = "Total Number of Reads\n (rpkm-normalized)") + 
  labs(x = "Treatment") +
  theme(axis.title.y = element_text(colour = 'black', 
                                    size   = 10, 
                                    angle  = 90, 
                                    hjust  = 0.5, 
                                    vjust  = 0.2, 
                                    face   = 'bold')) +
  theme(axis.title.x = element_text(colour = 'black', 
                                    size   = 10, 
                                    angle  = 0, 
                                    hjust  = 0.5, 
                                    vjust  =  -0.2, 
                                    face   = 'bold')) +
  geom_boxplot(position = position_dodge(0.8), 
               width    = 0.8, 
               outlier.size   = 1, 
               outlier.colour = "black", 
               outlier.shape  = 20) +
  stat_summary(fun.y = mean, 
               geom  = "point", 
               shape = 5, 
               size  = 1, 
               position = position_dodge(0.8))  +
#   facet_wrap(~SequenceID, 
#   facet_grid(SequenceID~Experiment,
  facet_grid(geneInfoTab~Experiment,
             scales = "free", space = "free_x") +
  theme(strip.text.x = element_text(size   = 6, 
                                    face   = 'bold.italic', 
                                    colour = "black", 
                                    angle  = 0))
print(bp)
dev.off()

```

Chunk to generate bar plots for individual genes and output them
to the specified folder for "analysisPath"
```{r}
library("ggplot2")

dir.create(paste(analysisPath, "/Individual_plots", sep = ""), 
           showWarnings = TRUE, recursive = FALSE)
i<- 1
for(i in 1:length(listLoci$geneInfoTab)){
  temp2 <- subset(temp, temp$SequenceID == listLoci$SequenceID[i])
  temp2$Experiment <- sub("Mycelium", "Mycel.", temp2$Experiment )
  png(file   = paste(analysisPath, "/Individual_plots/", 
                     listLoci$SequenceID[i], ".png", sep = ""), 
      width  = 7, 
      height = 3, 
      units  = "in", 
      res    = 300, 
      bg     = "white")
  bp <- ggplot(aes(y = Read_Counts, x = xLab), data  = temp2) +
  theme_bw() +
  theme(axis.text.y = element_text(size  = 8, 
                                   hjust = 1, 
                                   vjust = 0.4)) +
  theme(axis.text.x = element_text(colour = 'black', 
                                   size   = 8, 
                                   angle  = 45, 
                                   hjust  = 1, 
                                   vjust  = 1)) +
  theme(axis.ticks = element_line(colour = 'black', 
                                  size   = 0.5)) +
  labs(y = "Total Number of Reads\n(rpkm-normalized)") +
  labs(x = "Treatment") +
  theme(axis.title.y = element_text(colour = 'black', 
                                    size   = 8, 
                                    angle  = 90, 
                                    hjust  = 0.5, 
                                    vjust  = 1.5, 
                                    face   = 'bold')) +
  theme(axis.title.x = element_text(colour = 'black', 
                                    size   = 10, 
                                    angle  = 0, 
                                    hjust  = 0.5, 
                                    vjust  =  0, 
                                    face   = 'bold')) +
  geom_boxplot(position = position_dodge(0.8), 
               width    = 0.8, 
               outlier.size   = 1, 
               outlier.colour = "black", 
               outlier.shape  = 20) +
  stat_summary(fun.y = mean, 
               geom  = "point", 
               shape = 5, 
               size  = 0.5, 
               position = position_dodge(0.8))  +
#   facet_wrap(~SequenceID, 
#   facet_grid(SequenceID~Experiment,
  facet_grid(geneInfoTab~Experiment,
             scales = "free", space= "free_x") +
  theme(strip.text.x = element_text(size   = 9, 
                                    face   = 'bold', 
                                    colour = "black", 
                                    angle  = 0)) +
  theme(strip.text.y = element_text(size   = 7, 
                                    face   = 'bold', 
                                    colour = "black", 
                                    angle  = -90))
print(bp)
dev.off()
}
```


*** Over here Andre - just ran this after you left, 19March2017:
Bar plots per CAZy family group:
```{r}
library("ggplot2")
cazyFam <- unique(listLoci$CAZyFamily)
dirPlotName <- "ggplotsByCAZyFam"

dir.create(paste(analysisPath, "/", dirPlotName, sep = ""), 
           showWarnings = TRUE, recursive = FALSE)

ggplotsByCazyFamPath <- paste(analysisPath, "/", dirPlotName, "/", sep = "")

i <- 1
for(i in 1:length(cazyFam)){
  temp3 <- subset(temp, temp$CAZyFamily == cazyFam[i])
  temp3$Experiment <- sub("Mycelium", "Mycel.", temp3$Experiment )

pdf(file   = paste(ggplotsByCazyFamPath, cazyFam[i], ".pdf", sep = ""),
    width  = 8, 
    height = 7*ceiling(nrow(temp3)/90))

bp <- ggplot(aes(y = Read_Counts, x = xLab), data  = temp3) +
theme_bw() +
  theme(axis.text.y = element_text(size  = 8, 
                                   hjust = 1, 
                                   vjust = 0.4)) +
  theme(axis.text.x = element_text(colour = 'black', 
                                   size   = 8, 
                                   angle  = 45, 
                                   hjust  = 1, 
                                   vjust  = 1)) +
  theme(axis.ticks = element_line(colour = 'black', 
                                  size   = 0.5)) +
  labs(y = "Total Number of Reads\n(rpkm-normalized)") +
  labs(x = "Treatment") +
  theme(axis.title.y = element_text(colour = 'black', 
                                    size   = 8, 
                                    angle  = 90, 
                                    hjust  = 0.5, 
                                    vjust  = 1.5, 
                                    face   = 'bold')) +
  theme(axis.title.x = element_text(colour = 'black', 
                                    size   = 10, 
                                    angle  = 0, 
                                    hjust  = 0.5, 
                                    vjust  =  0, 
                                    face   = 'bold')) +
  geom_boxplot(position = position_dodge(0.8), 
               width    = 0.8, 
               outlier.size   = 1, 
               outlier.colour = "black", 
               outlier.shape  = 20) +
  stat_summary(fun.y = mean, 
               geom  = "point", 
               shape = 5, 
               size  = 0.5, 
               position = position_dodge(0.8))  +
#   facet_wrap(~SequenceID, 
#   facet_grid(SequenceID~Experiment,
  facet_grid(geneInfoTab~Experiment,
             scales = "free", space= "free_x") +
  theme(strip.text.x = element_text(size   = 9, 
                                    face   = 'bold', 
                                    colour = "black", 
                                    angle  = 0)) +
  theme(strip.text.y = element_text(size   = 7, 
                                    face   = 'bold', 
                                    colour = "black", 
                                    angle  = -90))
print(bp)
dev.off()
}

```