Generating plots of gene expression by time course and treatment
================================================================

Review paths to confirm they are still correct:
```{r}
sharedPath <- "/isilon/biodiversity/users/shared/Pythium_ultimum_RNAseq/"
```

Over here Emily to read the big file
```{r}
rpkmPlotsFullImport <- read.table(paste(sharedPath, 
                                        "rpkmForBarPlots_full.csv", 
                                        sep = ""), 
                                  sep = ",", 
                                  header = TRUE, 
                                  as.is = FALSE, 
                                  stringsAsFactors = FALSE)

rpkmPlotsFullImport$xLab <- paste(rpkmPlotsFullImport$Condition,
                                  rpkmPlotsFullImport$TimePoint, 
                                  sep = "_")
unique(rpkmPlotsFullImport$xLab)

rpkmPlotsFullImport$xLab <- sub("OoCTimeCourse_", "", rpkmPlotsFullImport$xLab)
rpkmPlotsFullImport$xLab <- sub("Various_0", "Various", rpkmPlotsFullImport$xLab)
unique(rpkmPlotsFullImport$xLab)
unique(rpkmPlotsFullImport$Experiment)

labOrdered <- c("Various", "Control_0", "Control_3", "Cholesterol_3", 
                "Control_7", "Cholesterol_7", "Control_24", "Cholesterol_24", 
                "0", "12", "24", "48", "72", "120", "240") 
rpkmPlotsFullImport$xLab <- factor(rpkmPlotsFullImport$xLab, levels = c(labOrdered))   
levels(rpkmPlotsFullImport$xLab)

expOrdered <- c("Mycelium", "Oosporogenesis", "OosporeConversion") 
rpkmPlotsFullImport$Experiment <- factor(rpkmPlotsFullImport$Experiment, 
                                         levels = c(expOrdered))   
levels(rpkmPlotsFullImport$Experiment)
```


Create Main plot folder
```{r}
mainFolder <- "GGPLOTS_FROM_FULL_DATA"
dir.create(paste(sharedPath, mainFolder , sep = ""), 
           showWarnings = TRUE, recursive = FALSE)
```

Create a folder for the plot outputs:
```{r}
analysisDir <- "from_Emily_24_March_2017"
dir.create(paste(sharedPath, mainFolder , "/", analysisDir, sep = ""), 
           showWarnings = TRUE, recursive = FALSE)
analysisPath <- paste(sharedPath, mainFolder , "/", analysisDir, sep = "")
```

Dump list file as single column and read it.  
The list could have 2 column, one with better labels.
List was sent by e-mail we think
Use next chunk instead
```{r}
# #listName <- "Loci_list.txt"
# #listName <- "listRefCandidates.txt"
# listName <- "CAZy_of_interest_list"
# 
# library(data.table)
# listLoci <- fread(paste(analysisPath, "/", listName, "", sep = ""), 
#                   sep = "\t", header = TRUE,  
#                   stringsAsFactors = FALSE)
# colnames(listLoci)[1] <- "SequenceID"
# colnames(listLoci)[2] <- "CAZyFamily"
# 
# #listLoci$geneInfo <- paste(listLoci$SequenceID, listLoci$CAZyFamily, sep = " -\n ")
```

Read CAZy of entire recent run of dbcan pipeline (from Emily's lists)
```{r}

listName <- "CAZy_of_interest_list"

library(data.table)

# Read full dataset
cazyPyuuAll <- fread(file = "data/dbCAN_Subset_AllRefs_columnSubset.csv",
#                        "CAZy_prediction_test/output/Pythium_ultimum_var_ultimum/Pythium_ultimum_var_ultimum.CAZy.out.dm", 
#                       sep = ""),
                  sep = "auto", header = TRUE, fill=TRUE)

cazyPyuuAll <- cazyPyuuAll[which(cazyPyuuAll$species=="Pythium_ultimum_var_ultimum"),c("cazyFam", "sequence")]

# Remove one strange code for GT2 (duplicated gene)
cazyPyuuAll <-cazyPyuuAll[!cazyPyuuAll$cazyFam == "GT2_Cellulose_synt",]


# Read Secreted dataset
cazyPyuuSecreted <- fread(file = "data/dbCAN_SecretedPyuu_Final_subsetColumns.csv",
                  sep = "auto", header = TRUE, fill=TRUE)

cazyPyuuSecreted <- cazyPyuuSecreted[ ,c("cazyFam", "sequence")]
cazyPyuuSecreted$Secreted <- "secreted"




# Now make a new table that has only the cazyFam and sequence columns plus info about Secreted
length(unique(paste(cazyPyuuSecreted$cazyFam,cazyPyuuSecreted$sequence, sep="_")))
length(unique(paste(cazyPyuuAll$cazyFam,cazyPyuuAll$sequence, sep="_")))


cazyFamGene <- merge(cazyPyuuAll, cazyPyuuSecreted, by = c("cazyFam", "sequence"), all.x = TRUE)

cazyFamGene$Secreted[is.na(cazyFamGene$Secreted)] <- "not secreted"

cazyFamGene <- cazyFamGene[!duplicated(cazyFamGene), ]

colnames(cazyFamGene)[2] <- "T_code"

# For some reason I can't get openxlsx package:

library(openxlsx)
LEV_2010_annotations <- read.xlsx(paste(sharedPath, 
                                 "References/Genome_Biology_paper_AdditionalDataFile3.xlsx", 
                                 sep = ""), 
                           sheet = "S9", 
                           startRow = 2)

T_G_conversion_table <- read.csv(paste(sharedPath, 
                                       "References/T_G_name_conversion_from_gff_file.csv", 
                                       sep = ""), 
                                 stringsAsFactors = FALSE)

length(unique(T_G_conversion_table$T_code))
length(unique(T_G_conversion_table$G_code))

T_G_conversion_table$G_code[duplicated(T_G_conversion_table$G_code)]
T_G_conversion_table$T_codes[duplicated(T_G_conversion_table$T_code)]

colnames(T_G_conversion_table)[colnames(T_G_conversion_table)=="G_code"] <- "Locus"

LEV_2010_df <- merge(T_G_conversion_table, LEV_2010_annotations, by = "Locus", all.x = TRUE)


CAZy_of_interest <- merge(cazyFamGene, LEV_2010_df[,2:3], by = "T_code", all.x = TRUE)

# one protein was duopklicated because of different annotations, remove it
CAZy_of_interest <- CAZy_of_interest[!duplicated(paste(CAZy_of_interest$cazyFam,CAZy_of_interest$T_code)),]


CAZy_of_interest$geneInfo <-    paste(CAZy_of_interest$T_code, " - ", CAZy_of_interest$Secreted, " - ", CAZy_of_interest$Functional.Annotation, sep = "")
CAZy_of_interest$geneInfoTab <- paste(CAZy_of_interest$T_code, " - ", CAZy_of_interest$Secreted, " \n ", CAZy_of_interest$Functional.Annotation, sep = "") 

# To add another line break after several spaces
for(i in 1:8){
  CAZy_of_interest$geneInfoTab <- sub(" ", "$", CAZy_of_interest$geneInfoTab, ignore.case = FALSE)
}
# add a space after last character denoted by $
CAZy_of_interest$geneInfoTab <- sub("$", " ", CAZy_of_interest$geneInfoTab, ignore.case = FALSE)
CAZy_of_interest$geneInfoTab <- sub(" ", "\n", CAZy_of_interest$geneInfoTab, ignore.case = FALSE)
CAZy_of_interest$geneInfoTab <- gsub("\\$", " ", CAZy_of_interest$geneInfoTab, ignore.case = FALSE)

print(CAZy_of_interest$geneInfoTab[1])

listLoci <- data.frame(CAZy_of_interest$T_code,CAZy_of_interest$cazyFam,CAZy_of_interest$Secreted,CAZy_of_interest$geneInfo,
                       CAZy_of_interest$geneInfoTab) 

colnames(listLoci)
colnames(listLoci) <- c("SequenceID", "CAZyFamily",  "Secreted", "geneInfo", "geneInfoTab")
print(listLoci$geneInfoTab[1])


write.table(listLoci, paste(analysisPath, "/", listName, ".csv", sep = ""),
            append = FALSE, sep = ",", col.names = NA)

# I'll need to read in the table - since I can't get openxlsx to work:
listLoci <- fread(paste(analysisPath, "/CAZy_of_interest_list.csv", sep = ""),
                  sep = "auto", header = TRUE)

# to chack if \n is there
print(listLoci$geneInfoTab[1])

```

Creates a data subset for plotting and exports this subset.  
This can help us when we want to diagnosticate outliers.
```{r}
temp <- merge(listLoci3, rpkmPlotsFullImport, by = "SequenceID", all.x = TRUE)
temp$Experiment <- sub("OosporeConversion", "Oospore Conversion", temp$Experiment)

# To order rows
temp <- temp[order(temp$geneInfo, temp$xLab),] 

# To select and order columns
selectCol <- c("SequenceID", "CAZyFamily", "geneInfo", "geneInfoTab", "ExpUnit", "Condition", 
               "TimePoint", "RNA_Replicate", "Experiment", "xLab", "Read_Counts") 

temp <- subset(temp, select = selectCol)

write.table(temp, paste(analysisPath, "/", listName,"_all.csv", sep = ""),
            append = FALSE, sep = ",", col.names = NA)

```


Single file, multiple Bar plots for gene in listLoci
```{r}
library("ggplot2")

svg(file   = paste(analysisPath, "/", listName,  ".svg", sep = ""),
    width  = 8, 
    height = 7*ceiling(nrow(temp)/200))
    # 200 in seems to be the maximum length for a pdf file, so I used sgv
    #height = 200)

bp <- ggplot(aes(y = Read_Counts, x = xLab), data  = temp) +
  theme_bw() +
  theme(axis.text.y = element_text(size  = 6, 
                                   hjust = 1, 
                                   vjust = 0.4)) +
  theme(axis.text.x = element_text(colour = 'black', 
                                   size   = 6, 
                                   angle  = 45, 
                                   hjust  = 1, 
                                   vjust  = 1)) +
  theme(axis.ticks = element_line(colour = 'black', 
                                  size   = 0.5)) +
  labs(y = "Total Number of Reads\n (rpkm-normalized)") + 
  labs(x = "Treatment") +
  theme(axis.title.y = element_text(colour = 'black', 
                                    size   = 10, 
                                    angle  = 90, 
                                    hjust  = 0.5, 
                                    vjust  = 0.2, 
                                    face   = 'bold')) +
  theme(axis.title.x = element_text(colour = 'black', 
                                    size   = 10, 
                                    angle  = 0, 
                                    hjust  = 0.5, 
                                    vjust  =  -0.2, 
                                    face   = 'bold')) +
  geom_boxplot(position = position_dodge(0.8), 
               width    = 0.8, 
               outlier.size   = 1, 
               outlier.colour = "black", 
               outlier.shape  = 20) +
  stat_summary(fun.y = mean, 
               geom  = "point", 
               shape = 5, 
               size  = 1, 
               position = position_dodge(0.8))  +
#   facet_wrap(~SequenceID, 
#   facet_grid(SequenceID~Experiment,
  facet_grid(geneInfoTab~Experiment,
             scales = "free", space = "free_x") +
  theme(strip.text.x = element_text(size   = 6, 
                                    face   = 'bold', 
                                    colour = "black", 
                                    angle  = 0))  +
  theme(strip.text.y = element_text(size   = 5, 
                                    colour = "black",
                                    hjust = 0,
                                    angle  =-90))

print(bp)
dev.off()

```

Chunk to generate bar plots for individual genes and output them
to the specified folder for "analysisPath"
```{r}
library("ggplot2")

create_dir <- paste(analysisPath, "/Individual_plots", sep = "")
if (dir.exists(create_dir)) {} else {
  dir.create(create_dir, showWarnings = TRUE, recursive = FALSE)  }

i <- 3
for(i in 1:length(listLoci$geneInfoTab)){
  temp2 <- subset(temp, temp$SequenceID == listLoci$SequenceID[i])
  temp2$Experiment <- sub("Mycelium", "Mycel.", temp2$Experiment )
  png(file   = paste(create_dir, "/", listLoci$SequenceID[i], ".png", sep = ""), 
      width  = 7, 
      height = 3, 
      units  = "in", 
      res    = 300, 
      bg     = "white")
  bp <- ggplot(aes(y = Read_Counts, x = xLab), data  = temp2) +
  theme_bw() +
  theme(axis.text.y = element_text(size  = 8, 
                                   hjust = 1, 
                                   vjust = 0.4)) +
  theme(axis.text.x = element_text(colour = 'black', 
                                   size   = 8, 
                                   angle  = 45, 
                                   hjust  = 1, 
                                   vjust  = 1)) +
  theme(axis.ticks = element_line(colour = 'black', 
                                  size   = 0.5)) +
  labs(y = "Total Number of Reads\n(rpkm-normalized)") +
  labs(x = "Treatment") +
  theme(axis.title.y = element_text(colour = 'black', 
                                    size   = 8, 
                                    angle  = 90, 
                                    hjust  = 0.5, 
                                    vjust  = 1.5, 
                                    face   = 'bold')) +
  theme(axis.title.x = element_text(colour = 'black', 
                                    size   = 10, 
                                    angle  = 0, 
                                    hjust  = 0.5, 
                                    vjust  =  0, 
                                    face   = 'bold')) +
  geom_boxplot(position = position_dodge(0.8), 
               width    = 0.8, 
               outlier.size   = 1, 
               outlier.colour = "black", 
               outlier.shape  = 20) +
  stat_summary(fun.y = mean, 
               geom  = "point", 
               shape = 5, 
               size  = 0.5, 
               position = position_dodge(0.8))  +
#   facet_wrap(~SequenceID, 
#   facet_grid(SequenceID~Experiment,
  facet_grid(geneInfoTab~Experiment,
             scales = "free", space= "free_x") +
  theme(strip.text.x = element_text(size   = 8, 
                                    face   = 'bold', 
                                    colour = "black", 
                                    angle  = 0)) +
  theme(strip.text.y = element_text(size   = 6, 
                                #    face   = 'bold', 
                                    colour = "black", 
                                    angle  = -90))
print(bp)
dev.off()
}
```


*** Over here Andre - just ran this after you left, 19March2017:
Bar plots per CAZy family group:
```{r}
library("ggplot2")
cazyFam <- unique(listLoci$CAZyFamily)

ggplotsByCazyFamPath <- paste(analysisPath, "/ggplotsByCAZyFam", sep = "")
if (dir.exists(ggplotsByCazyFamPath)) {} else {
  dir.create(ggplotsByCazyFamPath, showWarnings = TRUE, recursive = FALSE)  }

i <- 1
for(i in 1:length(cazyFam)){
  temp3 <- subset(temp, temp$CAZyFamily == cazyFam[i])
  temp3$Experiment <- sub("Mycelium", "Mycel.", temp3$Experiment )

pdf(file   = paste(ggplotsByCazyFamPath, "/", cazyFam[i], ".pdf", sep = ""),
    width  = 8, 
    height = 7*ceiling(nrow(temp3)/90))

bp <- ggplot(aes(y = Read_Counts, x = xLab), data  = temp3) +
theme_bw() +
  theme(axis.text.y = element_text(size  = 8, 
                                   hjust = 1, 
                                   vjust = 0.4)) +
  theme(axis.text.x = element_text(colour = 'black', 
                                   size   = 8, 
                                   angle  = 45, 
                                   hjust  = 1, 
                                   vjust  = 1)) +
  theme(axis.ticks = element_line(colour = 'black', 
                                  size   = 0.5)) +
  labs(y = "Total Number of Reads\n(rpkm-normalized)") +
  labs(x = "Treatment") +
  theme(axis.title.y = element_text(colour = 'black', 
                                    size   = 8, 
                                    angle  = 90, 
                                    hjust  = 0.5, 
                                    vjust  = 1.5, 
                                    face   = 'bold')) +
  theme(axis.title.x = element_text(colour = 'black', 
                                    size   = 10, 
                                    angle  = 0, 
                                    hjust  = 0.5, 
                                    vjust  =  0, 
                                    face   = 'bold')) +
  geom_boxplot(position = position_dodge(0.8), 
               width    = 0.8, 
               outlier.size   = 1, 
               outlier.colour = "black", 
               outlier.shape  = 20) +
  stat_summary(fun.y = mean, 
               geom  = "point", 
               shape = 5, 
               size  = 0.5, 
               position = position_dodge(0.8))  +
#   facet_wrap(~SequenceID, 
#   facet_grid(SequenceID~Experiment,
  facet_grid(geneInfoTab~Experiment,
             scales = "free", space= "free_x") +
  theme(strip.text.x = element_text(size   = 8, 
                                    face   = 'bold', 
                                    colour = "black", 
                                    angle  = 0)) +
  theme(strip.text.y = element_text(size   = 6, 
                          #          face   = 'bold', 
                                    colour = "black", 
                                    angle  = -90))
print(bp)
dev.off()
}

```





Emily's quick looks into patterns:
Looking at the CAZy total, versus secreted and not secreted, and by CAZy family
```{r}
######
listLoci3 <- fread(paste(dataPath, "/secretome/pyuu_listLoci_notSecreted_CAZy.csv", sep = ""),
                  sep = "auto", header = TRUE)
listLoci4 <- fread(paste(dataPath, "/secretome/pyuu_listLoci_Secreted_CAZy.csv", sep = ""),
                  sep = "auto", header = TRUE)



setkey(listLoci, "SequenceID")
secSubList <- list(listLoci4$SequenceID)
secSubtbl <- listLoci[c(secSubList)]
listLoci4 <- secSubtbl

setkey(listLoci, "SequenceID")
notSecList <- list(listLoci3$SequenceID)
notSecSubtbl <- listLoci[c(notSecList)]
listLoci3 <- secSubtbl

secCAZyList <- unique(listLoci4$CAZyFamily)
length(secCAZyList) # 41
notSecCAZyList <- unique(listLoci3$CAZyFamily)
length(notSecCAZyList) # 80

allCAZy <- unique(union(secCAZyList, notSecCAZyList))  
length(allCAZy) # 98


commonCAZy <- intersect(secCAZyList, notSecCAZyList)
length(commonCAZy) #23

secOnly <- setdiff(secCAZyList, notSecCAZyList)
length(secOnly) # 18

notSecOnly <- setdiff(notSecCAZyList, secCAZyList)
length(notSecOnly) # 57


dir.create(paste(dataPath, "/cazyFams_png", sep = ""), 
           showWarnings = TRUE, recursive = FALSE)
install.packages("VennDiagram")
library(VennDiagram)

# All CAZy families
png(file = paste(dataPath, "/cazyFams_png/allCAZyFamsPyuu.png", sep = ""), 
      width  = 7, 
      height = 3, 
      units  = "in", 
      res    = 300, 
      bg     = "white")
grid.newpage()
bp <- draw.pairwise.venn(41, 80, 23, category = c("Secreted CAZy families", "Not secreted CAZy families"),
                   lty = rep("blank", 2), fill = c("light blue", "pink"), alpha = rep(0.5, 2), cat.pos = c(0, 0), 
                   cat.dist = rep(0.025, 2))
print(bp)
dev.off()


# GH CAZy families
secGHfams <- c("GH28", "GH31", "GH3", "GH17", "GH5", "GH38", "GH43", "GH89", "GH13", "GH53", "GH7", "GH16", "GH131", 
               "GH6", "GH81", "GH2", "GH15", "GH72", "GH30", "GH18")
length(unique(secGHfams)) # 20
notSecGHfams <- c("GH109", "GH3", "GH30", "GH123", "GH17", "GH63", "GH5", "GH6", "GH72", "GH35", "GH47", "GH15", "GH18", 
                  "GH1", "GH37", "GH16", "GH114", "GH81", "GH31", "GH85")
length(unique(notSecGHfams)) # 20
intersect(secGHfams, notSecGHfams)
# "GH31" "GH3"  "GH17" "GH5"  "GH16" "GH6"  "GH81" "GH15" "GH72" "GH30" "GH18"
length(intersect(secGHfams, notSecGHfams)) # 11
length(setdiff(secGHfams, notSecGHfams)) # 9
# "GH28"  "GH38"  "GH43"  "GH89"  "GH13"  "GH53"  "GH7"   "GH131" "GH2"
length(setdiff(notSecGHfams, secGHfams)) # 9
# "GH109" "GH123" "GH63"  "GH35"  "GH47"  "GH1"   "GH37"  "GH114" "GH85"

png(file = paste(dataPath, "/cazyFams_png/GH_CAZyFamsPyuu.png", sep = ""), 
      width  = 7, 
      height = 3, 
      units  = "in", 
      res    = 300, 
      bg     = "white")
grid.newpage()
bp <- draw.pairwise.venn(20, 20, 11, category = c("Secreted GH CAZy", "Not secreted GH CAZy"),
                   lty = rep("blank", 2), fill = c("light blue", "pink"), alpha = rep(0.5, 2), cat.pos = c(0, 0), 
                   cat.dist = rep(0.025, 2))
print(bp)
dev.off()


# GT CAZy families
secGTfams <- c("GT4", "GT31", "GT2", "GT96", "GT41", "GT10")
length(secGTfams) # 6
notSecGTfams <- c("GT20", "GT71", "GT76", "GT48", "GT83", "GT59", "GT22", "GT4", "GT1", "GT41", "GT31", "GT8", "GT2", 
                  "GT33", "GT60", "GT66", "GT58", "GT96", "GT32", "GT57", "GT19", "GT90", "GT45", "GT39", "GT95", "GT24",
                  "GT28", "GT74", "GT30", "GT50", "GT7")
length(unique(notSecGTfams)) # 31
length(intersect(secGTfams, notSecGTfams)) # 5
# "GT4"  "GT31" "GT2"  "GT96" "GT41"
setdiff(secGTfams, notSecGTfams) # "GT10" 1
setdiff(notSecGTfams, secGTfams) # 26
# [1] "GT20" "GT71" "GT76" "GT48" "GT83" "GT59" "GT22" "GT1"  "GT8"  "GT33" "GT60" "GT66" "GT58" "GT32" "GT57" "GT19" "GT90"
# [18] "GT45" "GT39" "GT95" "GT24" "GT28" "GT74" "GT30" "GT50" "GT7" 

png(file = paste(dataPath, "/cazyFams_png/GT_CAZyFamsPyuu.png", sep = ""), 
      width  = 7, 
      height = 3, 
      units  = "in", 
      res    = 300, 
      bg     = "white")
grid.newpage()
bp <- draw.pairwise.venn(6, 31, 5, category = c("Secreted GT CAZy", "Not secreted GT CAZy"),
                   lty = rep("blank", 2), fill = c("light blue", "pink"), alpha = rep(0.5, 2), cat.pos = c(0, 0), 
                   cat.dist = rep(0.025, 2))
print(bp)
dev.off()
# CBM CAZy families
secCBMfams <- c("CBM63", "CBM25", "CBM1", "CBM21")
length(secCBMfams) # 4
notSecCBMfams <- c("CBM63", "CBM38", "CBM32", "CBM47", "CBM13", "CBM50", "CBM1", "CBM9", "CBM20", "CBM43", "CBM51")
length(notSecCBMfams) # ll
intersect(secCBMfams, notSecCBMfams) # "CBM63" "CBM1" 2
setdiff(secCBMfams, notSecCBMfams) # "CBM25" "CBM21" 2
setdiff(notSecCBMfams, secCBMfams) # "CBM38" "CBM32" "CBM47" "CBM13" "CBM50" "CBM9"  "CBM20" "CBM43" "CBM51" 9
png(file = paste(dataPath, "/cazyFams_png/CBM_CAZyFamsPyuu.png", sep = ""), 
      width  = 7, 
      height = 3, 
      units  = "in", 
      res    = 300, 
      bg     = "white")
grid.newpage()
bp <- draw.pairwise.venn(4, 11, 2, category = c("Secreted CBM CAZy", "Not secreted CBM CAZy"),
                   lty = rep("blank", 2), fill = c("light blue", "pink"), alpha = rep(0.5, 2), cat.pos = c(0, 0), 
                   cat.dist = rep(0.025, 2))
print(bp)
dev.off()
# CE CAZy families
secCEfams <- c("CE5", "CE10", "CE1", "CE13") # 4
notSecCEfams <- c("CE3", "CE1", "CE10", "CE11", "CE14", "CE4", "CE12", "CE9", "CE2", "CE7") # 10
intersect(secCEfams, notSecCEfams) # "CE10" "CE1" 2
setdiff(secCEfams, notSecCEfams) # "CE5"  "CE13" 2
setdiff(notSecCEfams, secCEfams) # "CE3"  "CE11" "CE14" "CE4"  "CE12" "CE9"  "CE2"  "CE7" 8
png(file = paste(dataPath, "/cazyFams_png/CE_CAZyFamsPyuu.png", sep = ""), 
      width  = 7, 
      height = 3, 
      units  = "in", 
      res    = 300, 
      bg     = "white")
grid.newpage()
bp <- draw.pairwise.venn(4, 10, 2, category = c("Secreted CE CAZy", "Not secreted CE CAZy"),
                   lty = rep("blank", 2), fill = c("light blue", "pink"), alpha = rep(0.5, 2), cat.pos = c(0, 0), 
                   cat.dist = rep(0.025, 2))
print(bp)
dev.off()
# PL CAZy families
secPLfams <- c("PL3", "PL1","PL4") # 3
notSecPLfams <- c("PL3") # 1
png(file = paste(dataPath, "/cazyFams_png/PL_CAZyFamsPyuu.png", sep = ""), 
      width  = 7, 
      height = 3, 
      units  = "in", 
      res    = 300, 
      bg     = "white")
grid.newpage()
bp <- draw.pairwise.venn(3, 1, 1, category = c("Secreted PL CAZy", "Not secreted PL CAZy"),
                   lty = rep("blank", 2), fill = c("light blue", "pink"), alpha = rep(0.5, 2), cat.pos = c(0, 0), 
                   cat.dist = rep(0.025, 2))
print(bp)
dev.off()
# AA CAZy families
secAAfams <- c("AA9", "AA8", "AA2") # 3
notSecAAfams <- c("AA6", "AA4", "AA8", "AA2", "AA7", "AA3") # 6
intersect(secAAfams, notSecAAfams) # "AA8" "AA2" 2
setdiff(secAAfams, notSecAAfams) # "AA9" 1
setdiff(notSecAAfams, secAAfams) # "AA6" "AA4" "AA7" "AA3" 4
png(file = paste(dataPath, "/cazyFams_png/AA_CAZyFamsPyuu.png", sep = ""), 
      width  = 7, 
      height = 3, 
      units  = "in", 
      res    = 300, 
      bg     = "white")
grid.newpage()
bp <- draw.pairwise.venn(3, 6, 2, category = c("Secreted AA CAZy", "Not secreted AA CAZy"),
                   lty = rep("blank", 2), fill = c("light blue", "pink"), alpha = rep(0.5, 2), cat.pos = c(0, 0), 
                   cat.dist = rep(0.025, 2))
print(bp)
dev.off()

library(png)
AApng <- readPNG(paste(dataPath, "/cazyFams_png/AA_CAZyFamsPyuu.png", sep = ""))
g1 <- rasterGrob(AApng, interpolate = T)

aLLpng <- readPNG(paste(dataPath, "/cazyFams_png/allCAZyFamsPyuu.png", sep = ""))
g2 <- rasterGrob(aLLpng, interpolate = T)

cbmPNG <- readPNG(paste(dataPath, "/cazyFams_png/CBM_CAZyFamsPyuu.png", sep = ""))
g3 <- rasterGrob(cbmPNG, interpolate = T)

cePNG <- readPNG(paste(dataPath, "/cazyFams_png/CE_CAZyFamsPyuu.png", sep = ""))
g4 <- rasterGrob(cePNG, interpolate = T)

ghPNG <- readPNG(paste(dataPath, "/cazyFams_png/GH_CAZyFamsPyuu.png", sep = ""))
g5 <- rasterGrob(ghPNG, interpolate = T)

gtPNG <- readPNG(paste(dataPath, "/cazyFams_png/GT_CAZyFamsPyuu.png", sep = ""))
g6 <- rasterGrob(gtPNG, interpolate = T)

plPNG <- readPNG(paste(dataPath, "/cazyFams_png/PL_CAZyFamsPyuu.png", sep = ""))
g7 <- rasterGrob(plPNG, interpolate = T)

library(gridExtra)
png(file = paste(dataPath, "/cazyFams_png/sevenCAZySetPlots.png", sep = ""), 
      width  = 7, 
      height = 3, 
      units  = "in", 
      res    = 300, 
      bg     = "white")
bp <- grid.arrange(g7, g1, g3, g4, g5, g6, g2, nrow = 3)
# dev.copy(bp, "sevenCAZySetPlots.png")
print(bp)
dev.off()

#########
```



CE13- T009144
CE13 - T006952
GH13/CBM25 - T001346
CBM63 - T010810
GH3 - T008966
GH3 - T012530
GH7 - T011896
GH15/CBM21 - T005445
GH16 - T001777
GH17 - T014660
GH30 - T006030
GH72 - T009223
GH72 - T009228
GH131 - T001945
GH131 - T001949
SLH - T003553

Making a larger image with multiple PNG plots for results:
```{r}
library(data.table)
pngList <- fread(file = paste(dataPath, "/cazyFams_png/cazyPNGlist.csv", sep = ""),
                 sep = "auto", header = TRUE, fill=TRUE, stringsAsFactors = F)

install.packages("imager")
library(imager)
install.packages("TeachingDemos")
library(TeachingDemos)
# I want images in sets of 4. There are 16 total in the list. 
# So for cropping:
# images 1, 5, 9, 13 are to be bottom-cropped
# images 2, 3, 6, 7, 10, 11, 14, 15 are top- and bottom-cropped
# images 4, 8, 12, 16 are to be top-cropped
# I'm doing them in sets of 4, and combining them
# There is an off-set between the oospore conversion and oosporogenesis dividing lines in image 3
# or those where the y-axis label is in the 100s, and not 10s because it requires more axis-label space

# pngList1 <- pngList[1:4]
# pngList1 <- pngList[5:8]
# pngList1 <- pngList[9:12]
pngList1 <- pngList[13:16]

# png1 <- load.image(pngList1$PathPNG[1])
# plot(png1)
# width(png1) # 2100
# height(png1) # 900
# depth(png1) # 1

png1 <- load.image(pngList1$PathPNG[1])
png1b <- imsub(png1, y <770)
png1c <- implot(png1b, TeachingDemos::shadowtext(1, 130, paste(pngList1$imgNum[1], ": ", pngList1$CazyFamily[1], sep = ""),
                                                 cex = 3.5, font = 4.5, col="darkblue", bg="blue", pos=4))
#plot(png1c)

png2 <- load.image(pngList1$PathPNG[2])
png2b <- imsub(png2, y < 770, y > 89)
png2c <- implot(png2b, TeachingDemos::shadowtext(1, 50, paste(pngList1$imgNum[2], ": ", pngList1$CazyFamily[2], sep = ""),
                                                 cex = 3.5, font = 4.5, col="darkblue", bg="blue", pos=4))
# plot(png2c)

png3 <- load.image(pngList1$PathPNG[3])
png3b <- imsub(png3, y < 770, y > 89)
png3c <- implot(png3b, TeachingDemos::shadowtext(1, 50, paste(pngList1$imgNum[3], ": ", pngList1$CazyFamily[3], sep = ""),
                                                 cex = 3.5, font = 4.5, col="darkblue", bg="blue", pos=4))
#plot(png3c)

png4 <- load.image(pngList1$PathPNG[4])
png4b <- imsub(png4, y > 89)
png4c <- implot(png4b, TeachingDemos::shadowtext(1, 50, paste(pngList1$imgNum[4], ": ", pngList1$CazyFamily[4], sep = ""),
                                                 cex = 3.5, font = 4.5, col="darkblue", bg="blue", pos=4))
#plot(png4c)

png5 <- list(png1c, png2c, png3c, png4c) %>% imappend("y")
plot(png5)
imager::save.image(png5, paste(figPath, "/plots_continuousX/combinedPNG-4b.png", sep = ""))

# Below are ways to get the info of the png, including the height and width, and 
# what different sections look like to decide on cropping
imsplit(png1, "c")
imsplit(png1,"c") %>% laply(mean) # [1] 0.9436821 0.9444731 0.9458000
imsplit(png1,"x") %>% laply(mean) %>% head # [1] 1 1 1 1 1 1
imsplit(png1,"y",11) %>% plot
```


```{r}

library(grid)
library(png)

i <- 1 

plots <- lapply(ll <- pngList$PathPNG,function(x){
  img <- as.raster(readPNG(x))
  rasterGrob(img, interpolate = TRUE)
})

library(ggplot2)
library(gridExtra)

# ggsave(paste(dataPath, "/cazyFams_png/Plots_Combined.png", sep = ""),
#        width=8.5, height=11, 
#        do.call(marrangeGrob, c(plots, list(nrow=2, ncol=1,top=NULL))))


library(gridExtra)
png(file = paste(dataPath, "/cazyFams_png/Plots_Combined.png", sep = ""), 
      width  = 7, 
      height = 3, 
      units  = "in", 
      res    = 300, 
      bg     = "white")
bp <- grid.arrange(g7, g1, g3, g4, g5, g6, g2, nrow = 3)

print(bp)
dev.off()




```