Generating plots of gene expression by time course and treatment
================================================================

Review paths to confirm they are still correct:
```{r}
sharedPath <- "/isilon/biodiversity/users/shared/Pythium_ultimum_RNAseq/"
```

Over here Emily to read the big file
```{r}
rpkmPlotsFullImport <- read.table(paste(sharedPath, 
                                        "rpkmForBarPlots_full.csv", 
                                        sep = ""), 
                                  sep = ",", 
                                  header = TRUE, 
                                  as.is = FALSE, 
                                  stringsAsFactors = FALSE)

rpkmPlotsFullImport$xLab <- paste(rpkmPlotsFullImport$Condition,
                                  rpkmPlotsFullImport$TimePoint, 
                                  sep = "_")
unique(rpkmPlotsFullImport$xLab)

rpkmPlotsFullImport$xLab <- sub("OoCTimeCourse_", "", rpkmPlotsFullImport$xLab)
rpkmPlotsFullImport$xLab <- sub("Various_0", "Various", rpkmPlotsFullImport$xLab)
unique(rpkmPlotsFullImport$xLab)
unique(rpkmPlotsFullImport$Experiment)

labOrdered <- c("Various", "Control_0", "Control_3", "Cholesterol_3", 
                "Control_7", "Cholesterol_7", "Control_24", "Cholesterol_24", 
                "0", "12", "24", "48", "72", "120", "240") 
rpkmPlotsFullImport$xLab <- factor(rpkmPlotsFullImport$xLab, levels = c(labOrdered))   
levels(rpkmPlotsFullImport$xLab)

expOrdered <- c("Mycelium", "Oosporogenesis", "OosporeConversion") 
rpkmPlotsFullImport$Experiment <- factor(rpkmPlotsFullImport$Experiment, 
                                         levels = c(expOrdered))   
levels(rpkmPlotsFullImport$Experiment)
```


Create Main plot folder
```{r}
mainFolder <- "GGPLOTS_FROM_FULL_DATA"
dir.create(paste(sharedPath, mainFolder , sep = ""), 
           showWarnings = TRUE, recursive = FALSE)
```

Create a folder for the plot outputs:
```{r}
analysisDir <- "from_Emily_19_March_2017"
dir.create(paste(sharedPath, mainFolder , "/", analysisDir, sep = ""), 
           showWarnings = TRUE, recursive = FALSE)
analysisPath <- paste(sharedPath, mainFolder , "/", analysisDir, sep = "")
```

Dump list file as single column and read it.  
The list could have 2 column, one with better labels.
List was sent by e-mail we think
Use next chunk instead
```{r}
# #listName <- "Loci_list.txt"
# #listName <- "listRefCandidates.txt"
# listName <- "CAZy_of_interest_list"
# 
# library(data.table)
# listLoci <- fread(paste(analysisPath, "/", listName, "", sep = ""), 
#                   sep = "\t", header = TRUE,  
#                   stringsAsFactors = FALSE)
# colnames(listLoci)[1] <- "SequenceID"
# colnames(listLoci)[2] <- "CAZyFamily"
# 
# #listLoci$geneInfo <- paste(listLoci$SequenceID, listLoci$CAZyFamily, sep = " -\n ")
```

Read CAZy of entire recent run of dbcan pipeline (from March 2016 version)
```{r}

library(data.table)
# test[[i]] <- fread(dbCANList[i], sep = "auto", header = FALSE)
cazyPyuuAll <- fread(paste(sharedPath, 
                        "CAZy_prediction_test/output/Pythium_ultimum_var_ultimum/Pythium_ultimum_var_ultimum.CAZy.out.dm", 
                        sep = ""),
                  sep = "auto", header = TRUE, fill=TRUE)
cazyPyuuAll <- cazyPyuuAll[-c(1, 2), ]

# rename some columns:
colnames(cazyPyuuAll)[1] <- "cazyFam"

# Now make a new table that has only the cazyFam and sequence columns
cazyFamGene <- subset(cazyPyuuAll, select=c("cazyFam", "sequence"))
cazyFamGene$cazyFam <- sub(".hmm", "", cazyFamGene$cazyFam)

# For some reason I can't get openxlsx package:

library(openxlsx)
LEV_2010_annotations <- read.xlsx(paste(sharedPath, 
                                 "References/Genome_Biology_paper_AdditionalDataFile3.xlsx", 
                                 sep = ""), 
                           sheet = "S9", 
                           startRow = 2)

T_G_conversion_table <- read.csv(paste(sharedPath, 
                                       "References/T_G_name_conversion_from_gff_file.csv", 
                                       sep = ""), 
                                 stringsAsFactors = FALSE)

length(unique(T_G_conversion_table$T_code))
length(unique(T_G_conversion_table$G_code))

T_G_conversion_table$G_code[duplicated(T_G_conversion_table$G_code)]
T_G_conversion_table$T_codes[duplicated(T_G_conversion_table$T_code)]

colnames(T_G_conversion_table)[colnames(T_G_conversion_table)=="G_code"] <- "Locus"

LEV_2010_df <- merge(T_G_conversion_table, LEV_2010_annotations, by = "Locus", all.x = TRUE)

colnames(cazyFamGene)[2] <- "T_code"

CAZy_of_interest <- merge(cazyFamGene, LEV_2010_df[,2:3], by = "T_code", all.x = TRUE)


CAZy_of_interest$geneInfo <- paste(CAZy_of_interest$T_code, CAZy_of_interest$Functional.Annotation, sep = " - ")
CAZy_of_interest$geneInfoTab <- paste(CAZy_of_interest$T_code, CAZy_of_interest$Functional.Annotation, sep = " -\n ")

colnames(listLoci)
listLoci <- data.frame(CAZy_of_interest$T_code,CAZy_of_interest$cazyFam,CAZy_of_interest$geneInfo,
                       CAZy_of_interest$geneInfoTab) 
colnames(listLoci) <- c("SequenceID", "CAZyFamily", "geneInfo", "geneInfoTab")

write.table(listLoci, paste(analysisPath, "/", listName, ".csv", sep = ""),
            append = FALSE, sep = ",", col.names = NA)

# I'll need to read in the table - since I can't get openxlsx to work:
listLoci <- fread(paste(analysisPath, "/CAZy_of_interest_list.csv", sep = ""),
                  sep = "auto", header = TRUE)

```


Creates a data subset for plotting and exports this subset.  
This can help us when we want to diagnosticate outliers.
```{r}
temp <- merge(listLoci, rpkmPlotsFullImport, by = "SequenceID", all.x = TRUE)
temp$Experiment <- sub("OosporeConversion", "Oospore Conversion", temp$Experiment)

# To order rows
temp <- temp[order(temp$geneInfo, temp$xLab),] 

# To select and order columns
selectCol <- c("SequenceID", "CAZyFamily", "geneInfo", "geneInfoTab", "ExpUnit", "Condition", 
               "TimePoint", "RNA_Replicate", "Experiment", "xLab", "Read_Counts") 

temp <- subset(temp, select = selectCol)

write.table(temp, paste(analysisPath, "/", listName,"_all.csv", sep = ""),
            append = FALSE, sep = ",", col.names = NA)

```


Single file, multiple Bar plots for gene in listLoci
```{r}
library("ggplot2")

pdf(file   = paste(analysisPath, "/", listName,  ".pdf", sep = ""),
    width  = 8, 
    height = 7*ceiling(nrow(temp)/200))

bp <- ggplot(aes(y = Read_Counts, x = xLab), data  = temp) +
  theme_bw() +
  theme(axis.text.y = element_text(size  = 6, 
                                   hjust = 1, 
                                   vjust = 0.4)) +
  theme(axis.text.x = element_text(colour = 'black', 
                                   size   = 6, 
                                   angle  = 45, 
                                   hjust  = 1, 
                                   vjust  = 1)) +
  theme(axis.ticks = element_line(colour = 'black', 
                                  size   = 0.5)) +
  labs(y = "Total Number of Reads\n (rpkm-normalized)") + 
  labs(x = "Treatment") +
  theme(axis.title.y = element_text(colour = 'black', 
                                    size   = 10, 
                                    angle  = 90, 
                                    hjust  = 0.5, 
                                    vjust  = 0.2, 
                                    face   = 'bold')) +
  theme(axis.title.x = element_text(colour = 'black', 
                                    size   = 10, 
                                    angle  = 0, 
                                    hjust  = 0.5, 
                                    vjust  =  -0.2, 
                                    face   = 'bold')) +
  geom_boxplot(position = position_dodge(0.8), 
               width    = 0.8, 
               outlier.size   = 1, 
               outlier.colour = "black", 
               outlier.shape  = 20) +
  stat_summary(fun.y = mean, 
               geom  = "point", 
               shape = 5, 
               size  = 1, 
               position = position_dodge(0.8))  +
#   facet_wrap(~SequenceID, 
#   facet_grid(SequenceID~Experiment,
  facet_grid(geneInfoTab~Experiment,
             scales = "free", space = "free_x") +
  theme(strip.text.x = element_text(size   = 6, 
                                    face   = 'bold.italic', 
                                    colour = "black", 
                                    angle  = 0))
print(bp)
dev.off()
```

Chunk to generate bar plots for individual genes and output them
to the specified folder for "analysisPath"
```{r}
library("ggplot2")

dir.create(paste(analysisPath, "/Individual_plots", sep = ""), 
           showWarnings = TRUE, recursive = FALSE)
i<- 1
for(i in 1:length(listLoci$geneInfoTab)){
  temp2 <- subset(temp, temp$SequenceID == listLoci$SequenceID[i])
  temp2$Experiment <- sub("Mycelium", "Mycel.", temp2$Experiment )
  png(file   = paste(analysisPath, "/Individual_plots/", 
                     listLoci$SequenceID[i], ".png", sep = ""), 
      width  = 7, 
      height = 3, 
      units  = "in", 
      res    = 300, 
      bg     = "white")
  bp <- ggplot(aes(y = Read_Counts, x = xLab), data  = temp2) +
  theme_bw() +
  theme(axis.text.y = element_text(size  = 8, 
                                   hjust = 1, 
                                   vjust = 0.4)) +
  theme(axis.text.x = element_text(colour = 'black', 
                                   size   = 8, 
                                   angle  = 45, 
                                   hjust  = 1, 
                                   vjust  = 1)) +
  theme(axis.ticks = element_line(colour = 'black', 
                                  size   = 0.5)) +
  labs(y = "Total Number of Reads\n(rpkm-normalized)") +
  labs(x = "Treatment") +
  theme(axis.title.y = element_text(colour = 'black', 
                                    size   = 8, 
                                    angle  = 90, 
                                    hjust  = 0.5, 
                                    vjust  = 1.5, 
                                    face   = 'bold')) +
  theme(axis.title.x = element_text(colour = 'black', 
                                    size   = 10, 
                                    angle  = 0, 
                                    hjust  = 0.5, 
                                    vjust  =  0, 
                                    face   = 'bold')) +
  geom_boxplot(position = position_dodge(0.8), 
               width    = 0.8, 
               outlier.size   = 1, 
               outlier.colour = "black", 
               outlier.shape  = 20) +
  stat_summary(fun.y = mean, 
               geom  = "point", 
               shape = 5, 
               size  = 0.5, 
               position = position_dodge(0.8))  +
#   facet_wrap(~SequenceID, 
#   facet_grid(SequenceID~Experiment,
  facet_grid(geneInfoTab~Experiment,
             scales = "free", space= "free_x") +
  theme(strip.text.x = element_text(size   = 9, 
                                    face   = 'bold', 
                                    colour = "black", 
                                    angle  = 0)) +
  theme(strip.text.y = element_text(size   = 7, 
                                    face   = 'bold', 
                                    colour = "black", 
                                    angle  = -90))
print(bp)
dev.off()
}
```


*** Over here Andre - just ran this after you left, 19March2017:
Bar plots per CAZy family group:
```{r}
library("ggplot2")
cazyFam <- unique(listLoci$CAZyFamily)
dirPlotName <- "ggplotsByCAZyFam"

dir.create(paste(analysisPath, "/", dirPlotName, sep = ""), 
           showWarnings = TRUE, recursive = FALSE)

ggplotsByCazyFamPath <- paste(analysisPath, "/", dirPlotName, "/", sep = "")

i <- 1
for(i in 1:length(cazyFam)){
  temp3 <- subset(temp, temp$CAZyFamily == cazyFam[i])
  temp3$Experiment <- sub("Mycelium", "Mycel.", temp3$Experiment )

pdf(file   = paste(ggplotsByCazyFamPath, cazyFam[i], ".pdf", sep = ""),
    width  = 8, 
    height = 7*ceiling(nrow(temp3)/90))

bp <- ggplot(aes(y = Read_Counts, x = xLab), data  = temp3) +
theme_bw() +
  theme(axis.text.y = element_text(size  = 8, 
                                   hjust = 1, 
                                   vjust = 0.4)) +
  theme(axis.text.x = element_text(colour = 'black', 
                                   size   = 8, 
                                   angle  = 45, 
                                   hjust  = 1, 
                                   vjust  = 1)) +
  theme(axis.ticks = element_line(colour = 'black', 
                                  size   = 0.5)) +
  labs(y = "Total Number of Reads\n(rpkm-normalized)") +
  labs(x = "Treatment") +
  theme(axis.title.y = element_text(colour = 'black', 
                                    size   = 8, 
                                    angle  = 90, 
                                    hjust  = 0.5, 
                                    vjust  = 1.5, 
                                    face   = 'bold')) +
  theme(axis.title.x = element_text(colour = 'black', 
                                    size   = 10, 
                                    angle  = 0, 
                                    hjust  = 0.5, 
                                    vjust  =  0, 
                                    face   = 'bold')) +
  geom_boxplot(position = position_dodge(0.8), 
               width    = 0.8, 
               outlier.size   = 1, 
               outlier.colour = "black", 
               outlier.shape  = 20) +
  stat_summary(fun.y = mean, 
               geom  = "point", 
               shape = 5, 
               size  = 0.5, 
               position = position_dodge(0.8))  +
#   facet_wrap(~SequenceID, 
#   facet_grid(SequenceID~Experiment,
  facet_grid(geneInfoTab~Experiment,
             scales = "free", space= "free_x") +
  theme(strip.text.x = element_text(size   = 9, 
                                    face   = 'bold', 
                                    colour = "black", 
                                    angle  = 0)) +
  theme(strip.text.y = element_text(size   = 7, 
                                    face   = 'bold', 
                                    colour = "black", 
                                    angle  = -90))
print(bp)
dev.off()
}

```