 RNA-Seq Data Analysis After Read Processing and TopHat2 alignment
=================================================================
This will help us when finding our files to source functions:
```{r}
install.packages("rprojroot")
library(rprojroot)
# We specify ours is an RStudio project
# The root object contains a function that will help us locate our package r files
# regarless of our current working directory
root <- rprojroot::is_rstudio_project
scriptsPath <- root$make_fix_file(".")("R")
scripts <- dir(root$find_file("R", path = root$find_file()))
scriptsl <- paste(scriptsPath, scripts, sep = "//")
lapply(scriptsl, source)
```

Chunk: Path Setting
The following paths are to directories where the references, tools and general requirements are 
located, this depends on the directories actually having been put there:
```{r}
sharedPath <- "/isilon/biodiversity/users/shared/Pythium_ultimum_RNAseq/"
sharedPathAn <- "/isilon/biodiversity/users/shared/Pythium_ultimum_RNAseq/"
referencesPath   <- paste(sharedPath, "References/", sep="")
cazyPath         <- paste(sharedPath, "CAZy/",       sep="")
htseqCountPath   <- "/opt/bio/HTSeq/bin/htseq-count"
pyuuRefPath      <- paste(referencesPath, "Pyuu_ref_1ribo_mito_no_repeats.fa", sep = "")
pyuugff3Path     <- paste(referencesPath, "Pyuu_ref_1ribo_mito_no_repeats.gff3", sep = "")
pyuuTranscripts  <- paste(referencesPath, "pythium_ultimum_transcripts.fasta", sep = "")
pyuuProteins     <- paste(referencesPath, "pythium_ultimum_proteins.fasta", sep = "")
```

Read in the metadata tables that contain all the information on the processing
outputs from time-course sets.
```{r}

ooGenMetaPath <- paste(sharedPath, 
                       "Oosporogenesis_Final_metadataTable_afterProcessing.csv",
                       sep = "")
Conversion_path   <- "Oospore_Conversion_TimeCourse/HiSeq_Analyses2/"
ooConvMetaPath <- paste(sharedPath,Conversion_path,
                        "OosporeConversion_Final_metadataTable_afterProcessing.csv",
                        sep = "")

```
**** Significant contamination observed, majority of reads unmapped - see which 
top species are the majority of our contamination from. 
```{r}

# ncbiDbPath     <- "/isilon/biodiversity/reference/ncbi/blastdb/reference/nt/nt"
# ncbiBlastnPath <- "/opt/bio/ncbi-blast+/bin/blastn"
# testContamPath <- paste(sharedPathAn, "T24-2_BC20/T24-2_BC20.TopHat.2016-05-20/", 
#                         "testContaminants/", sep = "")
# blastnOutFrmt  <- paste(" '", 
#                         "6", " qseqid", " sallacc", " pident", " length", 
#                         " mismatch", " gapopen", " qstart", " qend", " sstart", 
#                         " send", " evalue", " bitscore", 
#                         "'", sep = "") 
# maxTargetsSeqs <- 1
#
# prefix <- "testContamT24-2_BC20-2"
# cmd <- paste(ncbiBlastnPath,
#              " -db ",     ncbiDbPath,
#              " -query ",  paste(testContamPath, "T24-2_BC20.unmapped.DeRep.fasta", sep = ""),
#              " -max_target_seqs ", maxTargetsSeqs,
#              " -outfmt ", blastnOutFrmt,
#              " -out ",    paste(testContamPath, "T24-2_BC20.unmapped.DeRep.fasta2.bls", sep = ""),
#              sep = "")
# suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix, node)
```


HTSeq-count for TopHat2 hits:  
Prepare the metadata table for recording TopHat counts:
```{r}
# For paired-end not merged
# metadataAdapRem$countfTopHat = paste(metadataAdapRem$LibraryName, "TopHat2Count", sep = ".")

# # For merged or single-end reads:
# metadataAdapRM$countfTopHat = paste(metadataAdapRM$LibraryName, "TopHat2Count", sep = ".")
```


Read count with HT-Seq:
```{r}
# Set HT-Seq options:
# stranded <- "no"
# MINAQUAL <- 10
# prefix   <- "L_HTSeq_Qsub"
# node     <- 1
# 
# # cmd <- with(metadataAdapRM, # For merged or single-end reads
# cmd <- with(metadataAdapRem,  # For paired-end not merged
#             paste(htseqCountPath, 
#                   " -s ", stranded,
#                   " -a ", MINAQUAL,
#                   " --idattr=Parent ", 
#                   paste(sharedPathAn, LibraryName, "/", LibraryName,".TopHat", 
#                         topHatDate, "/", LibraryName, "_sn.sam ", sep = ""),
#                   pyuugff3Path, " > ",
#                   paste(sharedPathAn, LibraryName, "/", LibraryName,".TopHat", 
#                         topHatDate, "/", metadataAdapRem$countfTopHat, sep = ""),  # PE not merged
#                         # topHatDate, "/", metadataAdapRM$countfTopHat, sep = ""), # Merged/SE reads
#                   sep = ""))
# 
# suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix)
```

To remove the output files after you are done:
```{r}
system("/opt/gridengine/bin/linux-x64/qstat") # Remove qsub temp when qstat returns nothing.
RemoveQsubTempFiles(sharedPathAn, prefix)
```

Fix permissions for files:
```{r}
cmd <- paste( "find ", sharedPathAn, " -type d -exec chmod 770 {} + ", sep = ""); system(cmd)
cmd <- paste( "find ", sharedPathAn, " -type f -exec chmod 770 {} + ", sep = ""); system(cmd)
```

Get all data together for oosporogenesis and oospore conversion and add the counts for 
libraries representing the same treatments:
```{r}

metadataAdapRem1 <- read.csv(ooGenMetaPath, stringsAsFactors=FALSE)
metadataAdapRem1$countfTopHat <- paste(metadataAdapRem1$LibraryName, ".TopHat2_count", sep="")
metadataAdapRem1$Experiment <- "Oosporogenesis"

metadataAdapRem2 <- read.csv(ooConvMetaPath, stringsAsFactors=FALSE)
topHatDate <- ".2016-07-19"
metadataAdapRem2$TopHatOutPath <- paste(sharedPath,Conversion_path,metadataAdapRem2$LibraryName,"/",
            metadataAdapRem2$LibraryName, ".TopHat", topHatDate,"/", sep = "")
metadataAdapRem2$countfTopHat <- paste(metadataAdapRem2$LibraryName, ".TopHat2Count", sep="")
colnames(metadataAdapRem2)[colnames(metadataAdapRem2)=="RNASeq_Replicate"] <- "RNA_Replicate"
     metadataAdapRem2$Experiment <- "OosporeConversion"

common_cols <- intersect(colnames(metadataAdapRem1),colnames(metadataAdapRem2))

Metadata_2exp <- rbind(
  subset(metadataAdapRem1, select = common_cols), 
  subset(metadataAdapRem2, select = common_cols)
)

write.table(Metadata_2exp, file = file.path(sharedPath, "Metadata.Oospore_genesis_conversion.csv"),
            sep  = ",", 
            # sep = "\t", 
            row.names = TRUE, 
            col.names = NA, 
            quote = FALSE)
```


```{r}
rm(list=ls(pattern="metadataAdapRem"))
```


EdgeR on combines datasets
```{r}
library("edgeR")
# Identify the count files and read them into R using readDGE:

# For paired-end not merged:
countsListTopHat <- paste(Metadata_2exp$TopHatOutPath, Metadata_2exp$countfTopHat, sep = "")

# # For merged or single-end reads:
# countsListTopHat <- paste(sharedPathAn, metadataAdapRM$LibraryName, "/", metadataAdapRM$LibraryName,
#                           ".TopHat", topHatDate, "/", metadataAdapRM$countfTopHat, sep = "")

# Turn the count of the list into a data.frame
countsTopHat <- readDGE(countsListTopHat)$counts

colnames(countsTopHat) <- basename(colnames(countsTopHat))

colnames(countsTopHat) <- sub("\\.TopHat.*","",colnames(countsTopHat))

write.table(data.frame(countsTopHat), 
            file = file.path(sharedPath, 
                             "counts_readDGE_TopHat.Oospore_genesis_conversion.annotated2.csv"),
            sep  = ",", 
            # sep = "\t", 
            row.names = TRUE, 
            col.names = NA, 
            quote = FALSE)
```

* First option to branch for plotting since we have the same csv table for both 
oosporogenesis and oospore conversion countsTopHat[c(3166:3170),]

I thought there were a problem with the read readDGE.  The following would 
replace it but it did not help as the problem was before readDGE step
Kind of useless now but nice to see we can replace readDGE kind of manually.
```{r}
countsListTopHat <- paste(sharedPathAn, metadataAdapRem$LibraryName, "/", metadataAdapRem$LibraryName,
                          ".TopHat", topHatDate, "/", metadataAdapRem$countfTopHat, sep = "")

i <- 1
test <- list()
for(i in 1:length(countsListTopHat)){
test[[i]] <- read.table(file = countsListTopHat[i], sep = "\t", stringsAsFactors = FALSE)
#rownames(test[[i]]) <- unlist(test[[i]][1])
colnames(test[[i]])[2]   <- basename(countsListTopHat[i])
#test[[i]][1] <- NULL
print(nrow(test[[i]]))
}

DF_test2 <- Reduce(function(x, y) merge(x, y, by="V1", all=TRUE), test)

grep("PYU1_T003129",DF_test2)
```

Extract control or unwanted columns based on their names:
```{r}
# List the controls that you want to drop from count analyses, if there are none do: c("","") 
controlToDrop <- c("T7-c3_BC30_a1", "T7-c3_BC30_a2", "T7-c3_BC31_a1", "T7-c3_BC31_a2", 
                   "Tneg-H2O_BC32_a1", "Tneg-H2O_BC32_a2", "T48-2_BC22_a1", "T48-2_BC22_a2",
                   "T7-c3_BC29_a1", "T7-c3_BC29_a2") 
tempCountsTopHat <- data.frame(countsTopHat)
colnames(tempCountsTopHat) <- colnames(countsTopHat)
countsTopHatDf   <- tempCountsTopHat[, !(colnames(tempCountsTopHat) %in% controlToDrop)]

# Drop it in metadata as well:
Metadata_2exp_reduced   <- Metadata_2exp[!Metadata_2exp$LibraryName %in% controlToDrop, ]

write.table(Metadata_2exp_reduced, 
            file = file.path(sharedPath, "Metadata.Oospore_genesis_conversion_reduced.csv"),
            sep  = ",", 
            # sep = "\t", 
            row.names = TRUE, 
            col.names = NA, 
            quote = FALSE)
#<<<<<<< HEAD

remove(tempCountsTopHat)

#=======
#>>>>>>> b272e7c0e0c1cf0ccd5efa8628919339a1d1fff8
```

The trick was from here
http://stackoverflow.com/questions/19321053/sum-together-columns-of-data-frame-based-on-name-type
Make the names of the columns with different libraries of the same experimental unit identical - some 
columns have the same name. Then a row sum for each unique name is done. When a library is sequenced 
twice, the numbers will be added up.
```{r}
library("reshape2")
# to make sure that the T0 oosporogenesis does not get lumped with T0 Oospore conversion, adds a "c"
for (j in 1:nrow(Metadata_2exp_reduced)) {
if (Metadata_2exp_reduced$Experiment[j] == "Oosporogenesis") {
  Metadata_2exp_reduced$LibraryName[j] <- sub("^T0-","T0-c", Metadata_2exp_reduced$LibraryName[j])
} else { 
}}

# Must do the same on columns of data frame of counts, does 
# the sub command on those selected only to add a c
colnames(countsTopHatDf)[grep(
  "T0-1_BC01|T0-1_BC02|T0-2_Lib1|T0-4_BC02",
  colnames(countsTopHatDf))] <- sub("^T0-","T0-c", 
                                    colnames(countsTopHatDf)
                                    [grep("T0-1_BC01|T0-1_BC02|T0-2_Lib1|T0-4_BC02", 
                                          colnames(countsTopHatDf))])

# For paired-end not merged:
Metadata_2exp_reduced$ExpUnit <- sub("_.*", "", Metadata_2exp_reduced$LibraryName)
metadataName <- "metadata_2exp_reduced.csv"
write.table(data.frame(Metadata_2exp_reduced), file = file.path(sharedPath, metadataName),
            sep  = ",", 
            # sep = "\t", 
            row.names = TRUE, 
            col.names = NA, 
            quote = FALSE)

# For merged or single-end reads:
# metadataAdapRM$ExpUnit <- sub("_.*", "", metadataAdapRM$LibraryName)

countsTopHatDf$gene <- rownames(countsTopHatDf)
countsTopHatMelt    <- melt(countsTopHatDf)
colnames(countsTopHatMelt)[2] <- "LibraryName"

# For paired-end not merged:
df <- merge(countsTopHatMelt, Metadata_2exp_reduced[,c("LibraryName", "ExpUnit")], 
            by = "LibraryName", all.x = TRUE)

# # For merged or single-end reads:
# df <- merge(countsTopHatMelt, metadataAdapRM[,c("LibraryName", "ExpUnit")], 
#             by = "LibraryName", all.x = TRUE)

countsTopHatSum <- dcast(data = df[,c("gene", "value", "ExpUnit")], 
                         gene ~ ExpUnit, sum, value.var = "value")
rownames(countsTopHatSum) <- countsTopHatSum$gene
countsTopHatSum$gene      <- NULL

remove(countsTopHatMelt)
```

Normalization after removing rDNA data.  
The best approach to compare with Levesque et al 2010.
```{r}
library(Biostrings)
library("edgeR")
library("reshape2")
rowsRemove1 <- row.names(countsTopHatSum)[1:37] # These are unwanted ribosomal or mitochondrial genes 
rowsRemove2 <- c("__no_feature", 
                 "__ambiguous", 
                 "__too_low_aQual", 
                 "__not_aligned", 
                 "__alignment_not_unique")

countsTopHatGenes <- countsTopHatSum[-which(rownames(countsTopHatSum) 
                                            %in% c(rowsRemove1, rowsRemove2)), ]

# To see the proportion of reads not ribosomal compared to ribosomal and all else (for fun):
countsTopHatDf$gene <- NULL
mean(colSums(countsTopHatGenes)/colSums(countsTopHatDf))

# cpms <- cpm(countsTopHatGenes)  # counts per million

pyuuTranscripts  <- paste(sharedPath, "References/pythium_ultimum_transcripts.fasta", sep = "")

tempPyuuTrans <- readDNAStringSet(pyuuTranscripts)
namesTranscripts <- sub(" .*$", "", names(tempPyuuTrans))

# Create a vector with names:
geneLength <- setNames(width(tempPyuuTrans), namesTranscripts)
temp2      <- merge(countsTopHatGenes, geneLength, by.x = 0, by.y = 0)
rownames(temp2) <- temp2$Row.names
temp2$Row.names <- NULL

write.table(temp2, file = file.path(paste(sharedPath, sep = ""), "readDGEcountsWithGeneLength_2exp.csv"),
            append = FALSE, sep = ",", col.names = NA)
```

* Second option to Branch for plotting but we do not have that csv for oosporogenesis *

RPKM normalization
```{r}
# *** User, look at temp2 and count columns to select all up to y.
colSums(temp2)
min(colSums(temp2))
max(colSums(temp2))
mean(colSums(temp2[,1:37]))
mean(temp2$y)

rpkmCount <- rpkm(temp2[,1:37], # All columns except y, in HiSeq it's col 19
                  gene.length = temp2$y, normalized.lib.sizes = TRUE) 
colSums(rpkmCount)
min(colSums(rpkmCount))
max(colSums(rpkmCount))
mean_rpkm_2exp <- mean(colSums(rpkmCount))

remove(temp2)

```

Get data from Levesque et al.:
```{r}
library(openxlsx)

LEV_2010_rpkm <- read.xlsx(paste(sharedPath, 
                                 "References/Genome_Biology_paper_AdditionalDataFile3.xlsx", 
                                 sep = ""), 
                           sheet = "S9", 
                           startRow = 2)

#<<<<<<< HEAD
LEV_2010_rpkm[LEV_2010_rpkm=="NA"]  <- NA
colnames(LEV_2010_rpkm)[5]  <- "YEB_Pilch"

sapply(LEV_2010_rpkm,mode)
sapply(LEV_2010_rpkm,class)

LEV_2010_rpkm$Hypoxia = as.numeric(as.character(LEV_2010_rpkm$Hypoxia))
LEV_2010_rpkm$At.Infection = as.numeric(as.character(LEV_2010_rpkm$At.Infection))
LEV_2010_rpkm$YEB_Pilch = as.numeric(as.character(LEV_2010_rpkm$YEB_Pilch))
LEV_2010_rpkm$Mefenoxam.T = as.numeric(as.character(LEV_2010_rpkm$Mefenoxam.T))
LEV_2010_rpkm$Mefenoxam.C = as.numeric(as.character(LEV_2010_rpkm$Mefenoxam.C))
LEV_2010_rpkm$Temp.0C = as.numeric(as.character(LEV_2010_rpkm$Temp.0C))
LEV_2010_rpkm$Temp.35C = as.numeric(as.character(LEV_2010_rpkm$Temp.35C))
LEV_2010_rpkm$Temp.Room = as.numeric(as.character(LEV_2010_rpkm$Temp.Room))

sapply(LEV_2010_rpkm,mode)
sapply(LEV_2010_rpkm,class)

LEV_2010_rpkm$Locus[duplicated(LEV_2010_rpkm$Locus)]

LEV_2010_rpkm_unique <- LEV_2010_rpkm[!duplicated(LEV_2010_rpkm[,"Locus"]),]
                                                                
T_G_conversion_table <- read.csv(paste(sharedPath,"References/T_G_name_conversion_from_gff_file.csv", sep=""), stringsAsFactors=FALSE)

#T_G_conversion_table$TG <- paste(T_G_conversion_table$T_code, T_G_conversion_table$G_code,sep="_")

length(unique(T_G_conversion_table$T_code))
length(unique(T_G_conversion_table$G_code))

T_G_conversion_table$G_code[duplicated(T_G_conversion_table$G_code)]
T_G_conversion_table$T_codes[duplicated(T_G_conversion_table$T_code)]

colnames(T_G_conversion_table)[colnames(T_G_conversion_table)=="G_code"] <- "Locus"
#=======
T_G_conversion_table <- read.csv(paste(sharedPath, 
                                       "References/T_G_name_conversion_from_gff_file.csv", 
                                       sep = ""), 
                                 stringsAsFactors = FALSE)

colnames(T_G_conversion_table)[colnames(T_G_conversion_table) == "G_code"] <- "Locus"
#>>>>>>> b272e7c0e0c1cf0ccd5efa8628919339a1d1fff8

LEV_2010_rpkm_df <- merge(T_G_conversion_table, LEV_2010_rpkm_unique, by = "Locus", all.x = TRUE)

LEV_2010_rpkm_df$T_code[duplicated(LEV_2010_rpkm_df$T_code)]

rownames(LEV_2010_rpkm_df) <- LEV_2010_rpkm_df$T_code

LEV_2010_rpkm_df <- LEV_2010_rpkm_df[, -3:-1]


```



```{r}

Lev_2010_metadata <- data.frame(colnames(LEV_2010_rpkm_df), "Mycelium")

colnames(Lev_2010_metadata)  <- c("ExpUnit","Experiment")
Lev_2010_metadata$LibraryName <-  Lev_2010_metadata$ExpUnit
Lev_2010_metadata$Condition	<-  "Various"
Lev_2010_metadata$ShortName  <-  Lev_2010_metadata$ExpUnit
Lev_2010_metadata$TimePoint <-  0
Lev_2010_metadata$RNA_Replicate <-  1


common_cols2 <- intersect(colnames(Metadata_2exp_reduced),colnames(Lev_2010_metadata))


Metadata_3exp <- rbind(
  subset(Metadata_2exp_reduced, select = common_cols2), 
  subset(Lev_2010_metadata, select = common_cols2)
)


```

This is to see about contaminating sequence profile using Kraken, with Krona
for visualization. 
Oospore conversion time-course samples, especially the earliest 
time-points (T0, T12, T24) were primarily contaminating reads that did not map to 
the reference genome we have. From a blast search of the top 53 hits of unmapped 
reads for the T24 sample we saw that the taxa identified were 83% Paenibacillus. 
I want to see a profile for this in detail for my samples.
```{r}
pathKraken <- "/home/AAFC-AAC/girouxem/Downloads/kraken"
krakenDB <- "/home/AAFC-AAC/girouxem/krakenDB_first"

dir.create(paste(sharedPathAn, "krakenContaminationTesting", sep = ""), showWarnings=TRUE, recursive=FALSE)
pathKrakenTests <- paste(sharedPathAn, "krakenContaminationTesting/", sep = "")

prefix <- "kraken_testingQsubs"
cmd <- with(metadataAdapRem,
            paste(pathKraken,
                  " --preload ",
                  " --db ", krakenDB,
                  " --fastq-input",
                  #" --threads 12 ",
                  " --output ", paste(pathKrakenTests, "krakenTesting_", 
                                      metadataAdapRem$LibraryName, ".output", sep = ""),
                  " --paired ", paste(pathFastq, metadataAdapRem$processed5.R1.Fastq, sep = ""),
                  " ", paste(pathFastq, metadataAdapRem$processed5.R2.Fastq, sep = ""),
                  sep = ""))

# Note, if there are a lot of qsubs to run, some may terminate due to lack of
# available memory to run on the cluster. You can see which ones didn't run 
# by cd into the directory with the qsubs, and on the command line do:
# $  find . -type f -exec grep -H 'Cannot allocate memory' {} \;
# You can then run these separately after the others are complete.

suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix)
```

Clean-up step:
Remove the output files while keeping the qsub and bash file:
```{r}
system("/opt/gridengine/bin/linux-x64/qstat") # Remove qsub temp when qstat returns nothing.
RemoveQsubTempFiles(sharedPathAn, prefix)
```


To add the kraken output files to the metadata table:
```{r}
for(k in 1:nrow(metadataAdapRem)){
  metadataAdapRem$KrakenOutput <- paste("krakenTesting_", metadataAdapRem$LibraryName, ".output", sep = "") 
}
```


For users who want the full taxonomic name associated with each input sequence, 
use kraken-translate to produce two different output formats for classified sequences. 
The script operates on the output of kraken, like so:
kraken-translate --db $DBNAME sequences.kraken > sequences.labels
```{r}
prefix <- "krakenTranslateQsubs"
pathKrakenTranslate <- "/home/AAFC-AAC/girouxem/Downloads/kraken-translate"

cmd <- with(metadataAdapRem,
            paste(pathKrakenTranslate,
                  " --db ", krakenDB,
                  " ", paste(pathKrakenTests, metadataAdapRem$KrakenOutput, sep = ""),
                  " > ", paste(pathKrakenTests, "krakenTesting_", metadataAdapRem$LibraryName,
                               ".labels", sep = ""),
                  sep = ""))

suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix)
```

Clean-up step:
Remove the output files while keeping the qsub and bash file:
```{r}
system("/opt/gridengine/bin/linux-x64/qstat") # Remove qsub temp when qstat returns nothing.
RemoveQsubTempFiles(sharedPathAn, prefix)
```

To add the kraken output files to the metadata table:
```{r}
for(k in 1:nrow(metadataAdapRem)){
  metadataAdapRem$KrakenOutputLabels <- paste("krakenTesting_", metadataAdapRem$LibraryName, ".labels", sep = "") 
}
```

After blasting our sequences against our Kraken database, we tabulate the results
using the following:
```{r}
prefix <- "krakenReportQsubs"
pathKrakenResults <- "/home/AAFC-AAC/girouxem/Downloads/kraken-report"

cmd <- with(metadataAdapRem,
            paste(pathKrakenResults,
                  " --db ", krakenDB,
                  " ", paste(pathKrakenTests, metadataAdapRem$KrakenOutput, sep = ""),
                  " > ", paste(pathKrakenTests, "krakenTesting_", metadataAdapRem$LibraryName,
                               ".report", sep = ""),
                  sep = ""))

suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix)
```

Clean-up step:
Remove the output files while keeping the qsub and bash file:
```{r}
system("/opt/gridengine/bin/linux-x64/qstat") # Remove qsub temp when qstat returns nothing.
RemoveQsubTempFiles(sharedPathAn, prefix)
```

To visualize the kraken data, we can use krona:
```{r}
prefix <- "KrakenToKronaInFormatQsub-outputIn"
pathKronaImportTax <- "/home/AAFC-AAC/girouxem/Downloads/bin/ktImportTaxonomy"

cmd <- with(metadataAdapRem,
            paste("cut -f2,3 ", 
                  paste(pathKrakenTests, metadataAdapRem$KrakenOutput, sep = ""),
                  " > ", 
                  paste(pathKrakenTests, "krona_outputIn_", metadataAdapRem$LibraryName, ".in", sep = ""),
                  sep = ""))
suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix)
```
Clean-up step:
Remove the output files while keeping the qsub and bash file:
```{r}
system("/opt/gridengine/bin/linux-x64/qstat") # Remove qsub temp when qstat returns nothing.
RemoveQsubTempFiles(sharedPathAn, prefix)
```

To add the kraken output files to the metadata table:
```{r}
for(k in 1:nrow(metadataAdapRem)){
  metadataAdapRem$KronaIn <- paste("krona_outputIn_", metadataAdapRem$LibraryName, ".in", sep = "") 
}

```

Passing in to krona:
```{r}
prefix <- "kronaQsubs_OutputIn"
pathKronaImportTax <- "/home/AAFC-AAC/girouxem/Downloads/bin/ktImportTaxonomy"

cmd <- with(metadataAdapRem,
            paste(pathKronaImportTax, " ", 
                  paste(pathKrakenTests, metadataAdapRem$KronaIn, sep = ""),
                  " -o ", 
                  paste(pathKrakenTests, "krona_outputIn_", metadataAdapRem$LibraryName, ".out.html", sep = ""),
                  sep = ""))
suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix)
```
* Not done yet
Clean-up step:
Remove the output files while keeping the qsub and bash file:
```{r}
system("/opt/gridengine/bin/linux-x64/qstat") # Remove qsub temp when qstat returns nothing.
RemoveQsubTempFiles(sharedPathAn, prefix)
```

```{r}

rpkmCount_3exp <- merge(rpkmCount, LEV_2010_rpkm_df, by = "row.names", all.x = TRUE)

# somewhow forcing rpkmCount matrix into data creates a Row.names column
rownames(rpkmCount_3exp) <- rpkmCount_3exp$Row.names

rpkmCount_3exp$Row.names <- NULL

```



***House Keeping Genes***
Finding potential house-keeping (HK) genes from 3 experiments
```{r}

logRpkmCount      <- log10(rpkmCount_3exp+1)
logRpkmCount$mean <- apply(logRpkmCount, 1, mean)
logRpkmCount$min  <- apply(logRpkmCount, 1, min)
logRpkmCount$max  <- apply(logRpkmCount, 1, max)
logRpkmCount$SD   <- apply(logRpkmCount, 1, sd)

#<<<<<<< HEAD
potentialHK <- subset(logRpkmCount, min > 0 & SD < 0.25)
potentialHK <- potentialHK[order(-potentialHK$mean),]  

write.table(potentialHK, 
            file   = file.path(paste(sharedPath, sep = ""), "potential_houseKeepingGenes.csv"),
            append = FALSE, sep = ",", col.names = NA)
```



The following was done by Andre, and the shared path was not the same. I can't get "rentrez" to 
work for me, so I won't repeat this. Instead I will refer to tab-delinited file with the genes 
using a hard path reference.
This is to generate a list of GI queries to search for BLAST (makes it a lot faster)
```{r}
# Make a list of potential HK genes:
interestListHK <- rownames(potentialHK)

dir.create(paste(sharedPath, "Blast_HK", sep = ""), showWarnings = TRUE, recursive = FALSE)

# Read all transcripts:
interestHK <- readAAStringSet(pyuuProteins, format = "fasta") 

# fixes the names to match list
names(interestHK) <- gsub(" .*", "", names(interestHK), ignore.case = FALSE)

# makes a subset of only the potential
interestHK <- interestHK[which(names(interestHK) %in% interestListHK)]

writeXStringSet(interestHK, 
                file   = paste(sharedPath, "Blast_HK/HK_InterestSubset.fasta", sep = ""), 
                append = FALSE, format = "fasta") 

install.packages("rentrez")
library("rentrez") 
query <- "(Oomycetes[ORGN])"

webEnvSearch <- entrez_search(db = "protein", query, retmax = 999999)
webEnvSearch
length(webEnvSearch$ids)
write.table(webEnvSearch$ids, 
            file = paste(sharedPath,"Blast_HK/gilist.txt", sep = ""), 
            append = FALSE, quote = FALSE, row.names = FALSE, col.names = FALSE)
```

This is to prerform a BLAST on the best candidate House Keeping genes
```{r}
ncbiDbPath     <- "/isilon/biodiversity/reference/ncbi/blastdb/reference/nr/nr"
ncbiBlastnPath <- "/opt/bio/ncbi-blast+/bin/blastp"
blastnOutFrmt  <-  c("sseqid", "qseqid", "sacc", "pident", "length",
                     "mismatch", "gapopen", "qstart", "qend", "sstart",
                     "send", "evalue", "bitscore")
#maxTargetsSeqs <- 3
prefix   <- "Blast_HK"  # need to clean up by inserting below
node <- 1

cmd <- paste(ncbiBlastnPath,
             " -db ",     ncbiDbPath,
             " -query ",  paste(sharedPath, "Blast_HK/HK_InterestSubset.fasta", sep = ""),
             " -evalue  1.00E-90", 
             " -num_threads ", node,
             " -gilist ", sharedPath,   "Blast_HK/gilist.txt",
             " -outfmt '6 ", paste(blastnOutFrmt, collapse=" "), "'",
             " -out ",    paste(sharedPath, "Blast_HK/HK_InterestSubsetNt2.tab", sep = ""),
             sep = "")

suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix, node)

```


Andre did the next chunk, it's not working for me due to library("spider") error. 
This is to get the names of the genes and select/sort  the results by BLAST statistics
```{r}
install.packages("spider")
library("spider")

subsetHKinterestPath <- paste(sharedPath, 
                              "Blast_HK/HK_InterestSubsetNt2.tab", 
                              sep = "")

hitsHK <- read.table(subsetHKinterestPath, sep = "\t", header = FALSE,
                     comment.char = "", quote = "", as.is = TRUE) 
colnames(hitsHK) <- blastnOutFrmt

hitSeqHK <- read.GB(hitsHK$sacc, seq.names = hitsHK$sacc, species.names = TRUE, gene = TRUE,
                    access = TRUE, as.character = FALSE)
#rm("hitSeqHK")

hitGenesHK <- data.frame(attr(hitSeqHK, "gene"), attr(hitSeqHK, "accession_num"), 
                           stringsAsFactors = FALSE)

colnames(hitGenesHK) <- c("gene", "sacc")
length(unique(hitGenesHK$sacc))

hitsHK <- merge(unique(hitGenesHK), hitsHK, by  = "sacc",  all.y= TRUE)
temp   <- hitsHK[-grep("unnamed|hypothetical",hitsHK$gene),]
temp   <- temp[order(temp$qseqid,temp$evalue,-temp$bitscore),] 

write.table(temp, 
            file = file.path(paste(sharedPathAn, sep=""),"potential_houseKeepingGenes_BLAST.csv"),
            append = FALSE, sep = ",", col.names = NA)
```


*** Creating a big file for all plots ***

rpkmCount for the three experiments
```{r}
temp_to_melt <- cbind(rownames(rpkmCount_3exp),rpkmCount_3exp)

rpkmForBarPlots <- melt(temp_to_melt)


colnames(rpkmForBarPlots)[1] <- "SequenceID"
colnames(rpkmForBarPlots)[2] <- "ExpUnit"
colnames(rpkmForBarPlots)[3] <- "Read_Counts"


unique(rpkmForBarPlots$ExpUnit) # Notice ExpUnit "y"
#rpkmForBarPlots <- rpkmForBarPlots[rpkmForBarPlots$ExpUnit!="y",]
#unique(rpkmForBarPlots$ExpUnit) # Notice ExpUnit "y" is gone

#I took out library name because we added two libraries
metadataColsToMerge <- c("Experiment","TimePoint", "Condition", "RNA_Replicate", "ExpUnit") 

# For paired-end not merged:
rpkmForBarPlots_full <- merge(rpkmForBarPlots, unique(Metadata_3exp[, which(names(Metadata_3exp) %in% metadataColsToMerge)]),
                  by = "ExpUnit", all.x = TRUE)


# # For merged or single-end reads:
# cazyRpkm <- merge(rpkmForBarPlots, unique(metadataAdapRM[,c(metadataColsToMerge)]),
#                   by = "ExpUnit", all.x = TRUE)


write.table(rpkmForBarPlots_full, 
            file = file.path(paste(sharedPath, sep=""),"rpkmForBarPlots_full.csv"),
            append = FALSE, sep = ",", col.names = NA)

```



Ignore for now
```{r}

cazyRpkmSub  <- cazyRpkm[which(cazyRpkm$SequenceID 
                              %in% unique(cazyDataSub[,c("SequenceID")])), ]

# this duplicate some of the enzymes that are in two groups
cazyRpkmSub2 <- merge(cazyRpkmSub, cazyDataSub[,c("CazyFam", "SequenceID")],  
                      by = "SequenceID", all.x = TRUE)

cazyRpkmSub2$Secreted_CAZy <- paste(cazyRpkmSub2$CazyFam, cazyRpkmSub2$SequenceID, sep = "-")

cazyRpkmSub2$Experiment <- "OosporeConversion"
cazyRpkmSub2$Condition  <- "OosporeConversion" 
# I'm adding this because the oosporogenesis has a column called Condition and 
# I can't do rbind if not the same


nameReadDGEcsv <- "countsReadDGE.TopHat.SecretedCAZy.HiSeqAnalyses2.OosporeConversion.csv"
write.table(cazyRpkmSub2, file = file.path(sharedPathAn, nameReadDGEcsv),
            sep = ",", row.names = TRUE, col.names = NA, quote = FALSE)

# Try to get a csv that looks like cazyRpkmSub2
readDGEcsvOoGen  <- paste(sharedPath, "EmilyTryingtoMakeLikeOosConHiSeq2Layout_counts_readDGE_TopHat_secreted_CAZy.csv", sep = "")
readDGEcsvOoConv <- paste(sharedPathAn, 
                          "countsReadDGE.TopHat.SecretedCAZy.HiSeqAnalyses2.OosporeConversion.csv",
                          sep = "") 


csvOoGen <- read.table(file = readDGEcsvOoGen, sep = ",", header = TRUE, comment.char = "", 
                       quote = "", as.is = TRUE, stringsAsFactors = FALSE)

csvOoGen$Experiment <- "Oosporogenesis" # This is to make it like the oospore conversion csv
csvOoGen$LibraryName <- paste(csvOoGen$ExpUnit, "_", csvOoGen$RNASeq_Replicate, "rep", sep = "")

library("plyr")

colnames(csvOoGen)
colnames(cazyRpkmSub2)

#because there is no condition for covversion
cazyRpkmSub2$Condition <- NA

#rbind fill fixed the issue of the extra column.  It put NA for the missing values.
cazyRpkmSub4 <- rbind.fill(cazyRpkmSub2, csvOoGen)

# creates a new column with conditions and time combined
cazyRpkmSub4$Timepoint_Condition <- paste(cazyRpkmSub4$TimePoint,cazyRpkmSub4$Condition,sep="_")
cazyRpkmSub4$Timepoint_Condition <- sub("_NA","", cazyRpkmSub4$Timepoint_Condition, ignore.case = FALSE)

# order the factors so they get in the right order in the plot
cazyRpkmSub4$Timepoint_Condition <- 
  factor(cazyRpkmSub4$Timepoint_Condition, levels = c(timePoints,
        "3_Control", "3_Cholesterol","7_Control","7_Cholesterol","24_Control","24_Cholesterol"))

# Over here!!!
# > cazyRpkmSub4 <- rbind(cazyRpkmSub2, csvOoGen)
# Warning message:
# In `[<-.factor`(`*tmp*`, ri, value = c(24L, 3L, 3L, 7L, 3L, 3L,  :
#   invalid factor level, NA generated

```


=======
df <- merge(T_G_conversion_table, LEV_2010_rpkm, by = "Locus", all.x = TRUE)
>>>>>>> b272e7c0e0c1cf0ccd5efa8628919339a1d1fff8
Setting up for Differential Analysis: Generating CAZy families, all Secreted and CAZy secreted, 
as well as House-Keeping subsets

Selecting the subset of reads that are to be analysed (experiment-specific)
```{r}
# To look at all CAZy in Pyuu, not just secreted:
allCazyFasta     <- paste(shared_path, "References/Secretome/all_CAZy.fasta", sep = "")
allCazy          <- readAAStringSet(paste(allCazyFasta), format = "fasta")
allCazyNames     <- names(allCazy)
allCazyPyuuNames <- subset(allCazyNames, grepl("^Pyuu_", allCazyNames))
allCazyPyuuNames <- sub("_secreted$", "", allCazyPyuuNames)

# To look at only secreted CAZy in Pyuu:
secCazyFasta <- paste(referencesPath, "Secretome/CAZy_secreted.fasta", sep = "")
secretedCAZyAll   <- readAAStringSet(paste(secCazyFasta), format = "fasta")
secretedNames     <- names(secretedCAZyAll)
secretedPyuuNames <- subset(secretedNames, grepl("^Pyuu_", secretedNames))

# Name the names file from the above options that will be analyzed:
namesForPlots <- secretedPyuuNames
# namesForPlots <- allCazyPyuuNames
cazyDataSub <- data.frame(matrix(unlist(strsplit(as.character(namesForPlots), "_")), 
                                nrow  = length(namesForPlots), 
                                byrow = TRUE), stringsAsFactors = FALSE)

cazyDataSub$SequenceID   <- paste(cazyDataSub$X3, cazyDataSub$X4, sep = "_")
colnames(cazyDataSub)[2] <- "CazyFam"
cazyFam <- unique(cazyDataSub$CazyFam)
```

Counts per million from the gene counts from TopHat for Cazy data 
```{r}
countsForBarPlots <- melt(cpms, id.vars = colnames(cpms))

colnames(countsForBarPlots)[1] <- "SequenceID"
colnames(countsForBarPlots)[2] <- "ExpUnit"
colnames(countsForBarPlots)[3] <- "Read_Counts"

metadataColsToMerge <- c("LibraryName", "TimePoint", "RNASeq_Replicate", "ExpUnit") 

# For paired-end not merged:
cazyCounts <- merge(countsForBarPlots, unique(metadataAdapRem[,c(metadataColsToMerge)]),  
                    by = "ExpUnit", all.x = TRUE)

# # For merged or single-end reads:
# cazyCounts <- merge(countsForBarPlots, unique(metadataAdapRM[,c(metadataColsToMerge)]),  
#                     by = "ExpUnit", all.x = TRUE)

cazyCounts$TimePoint <- factor(cazyCounts$TimePoint, levels = c(timePoints))        
length(unique(cazyCounts$TimePoint))
cazyCountsSub <- cazyCounts[which(cazyCounts$SequenceID %in% unique(cazyDataSub[,c("SequenceID")])), ]
```

rpkmCount for Cazy
```{r}
rpkmForBarPlots <- melt(rpkmCount, id.vars = colnames(rpkmCount))
colnames(rpkmCount)
length(unique(rpkmForBarPlots$Var2)) 
colnames(rpkmForBarPlots)[1] <- "SequenceID"
colnames(rpkmForBarPlots)[2] <- "ExpUnit"
colnames(rpkmForBarPlots)[3] <- "Read_Counts"

unique(rpkmForBarPlots$ExpUnit) # Notice ExpUnit "y"
rpkmForBarPlots <- rpkmForBarPlots[rpkmForBarPlots$ExpUnit!="y",]
unique(rpkmForBarPlots$ExpUnit) # Notice ExpUnit "y" is gone

#I took out library name because we added two libraries
metadataColsToMerge <- c("TimePoint", "RNASeq_Replicate", "ExpUnit") 

# For paired-end not merged:
cazyRpkm <- merge(rpkmForBarPlots, unique(metadataAdapRem[,c(metadataColsToMerge)]),
                  by = "ExpUnit", all.x = TRUE)

# # For merged or single-end reads:
# cazyRpkm <- merge(rpkmForBarPlots, unique(metadataAdapRM[,c(metadataColsToMerge)]),
#                   by = "ExpUnit", all.x = TRUE)

cazyRpkm$TimePoint <- factor(cazyRpkm$TimePoint, levels = c(timePoints))        

cazyRpkmSub  <- cazyRpkm[which(cazyRpkm$SequenceID 
                              %in% unique(cazyDataSub[,c("SequenceID")])), ]

# this duplicate some of the enzymes that are in two groups
cazyRpkmSub2 <- merge(cazyRpkmSub, cazyDataSub[,c("CazyFam", "SequenceID")],  
                      by = "SequenceID", all.x = TRUE)

cazyRpkmSub2$Secreted_CAZy <- paste(cazyRpkmSub2$CazyFam, cazyRpkmSub2$SequenceID, sep = "-")

cazyRpkmSub2$Experiment <- "OosporeConversion"
cazyRpkmSub2$Condition  <- "OosporeConversion" 
# I'm adding this because the oosporogenesis has a column called Condition and 
# I can't do rbind if not the same


nameReadDGEcsv <- "countsReadDGE.TopHat.SecretedCAZy.HiSeqAnalyses2.OosporeConversion.csv"
write.table(cazyRpkmSub2, file = file.path(sharedPathAn, nameReadDGEcsv),
            sep = ",", row.names = TRUE, col.names = NA, quote = FALSE)

# Try to get a csv that looks like cazyRpkmSub2
readDGEcsvOoGen  <- paste(sharedPath, "EmilyTryingtoMakeLikeOosConHiSeq2Layout_counts_readDGE_TopHat_secreted_CAZy.csv", sep = "")
readDGEcsvOoConv <- paste(sharedPathAn, 
                          "countsReadDGE.TopHat.SecretedCAZy.HiSeqAnalyses2.OosporeConversion.csv",
                          sep = "") 


csvOoGen <- read.table(file = readDGEcsvOoGen, sep = ",", header = TRUE, comment.char = "", 
                       quote = "", as.is = TRUE, stringsAsFactors = FALSE)

csvOoGen$Experiment <- "Oosporogenesis" # This is to make it like the oospore conversion csv
csvOoGen$LibraryName <- paste(csvOoGen$ExpUnit, "_", csvOoGen$RNASeq_Replicate, "rep", sep = "")

library("plyr")

colnames(csvOoGen)
colnames(cazyRpkmSub2)

#because there is no condition for covversion
cazyRpkmSub2$Condition <- NA

#rbind fill fixed the issue of the extra column.  It put NA for the missing values.
cazyRpkmSub4 <- rbind.fill(cazyRpkmSub2, csvOoGen)

# creates a new column with conditions and time combined
cazyRpkmSub4$Timepoint_Condition <- paste(cazyRpkmSub4$TimePoint,cazyRpkmSub4$Condition,sep="_")
cazyRpkmSub4$Timepoint_Condition <- sub("_NA","", cazyRpkmSub4$Timepoint_Condition, ignore.case = FALSE)

# order the factors so they get in the right order in the plot
cazyRpkmSub4$Timepoint_Condition <- 
  factor(cazyRpkmSub4$Timepoint_Condition, levels = c(timePoints,
        "3_Control", "3_Cholesterol","7_Control","7_Cholesterol","24_Control","24_Cholesterol"))

# Over here!!!
# > cazyRpkmSub4 <- rbind(cazyRpkmSub2, csvOoGen)
# Warning message:
# In `[<-.factor`(`*tmp*`, ri, value = c(24L, 3L, 3L, 7L, 3L, 3L,  :
#   invalid factor level, NA generated

```


<<<<<<< HEAD
=======
Finding potential house-keeping (HK) genes
```{r}
colnames(rpkmCount)
logRpkmCount      <- data.frame(log10(rpkmCount[,1:17]+1))
logRpkmCount$mean <- apply(logRpkmCount, 1, mean)
logRpkmCount$min  <- apply(logRpkmCount, 1, min)
logRpkmCount$max  <- apply(logRpkmCount, 1, max)
logRpkmCount$SD   <- apply(logRpkmCount, 1, sd)

potentialHK <- subset(logRpkmCount, min > 0 & SD < 0.25)
potentialHK <- potentialHK[order(-potentialHK$mean),]  

write.table(potentialHK, 
            file   = file.path(paste(sharedPathAn, sep = ""), "potential_houseKeepingGenes.csv"),
            append = FALSE, sep = ",", col.names = NA)
```


This is to generate a list of GI queries to search for BLAST (makes it a lot faster)
```{r}
# Make a list of potential HK genes:
interestListHK <- rownames(potentialHK)

dir.create(paste(sharedPathAn, "Blast_HK", sep = ""), showWarnings = TRUE, recursive = FALSE)

# Read all transcripts:
interestHK <- readAAStringSet(pyuuProteins, format = "fasta") 

# fixes the names to match list
names(interestHK) <- gsub(" .*", "", names(interestHK), ignore.case = FALSE)

# makes a subset of only the potential
interestHK <- interestHK[which(names(interestHK) %in% interestListHK)]

writeXStringSet(interestHK, 
                file   = paste(sharedPathAn, "Blast_HK/HK_InterestSubset.fasta", sep = ""), 
                append = FALSE, format = "fasta") 

install.packages("rentrez")
library("rentrez") # Error in library("rentrez") : there is no package called ‘rentrez’
query <- "(Oomycetes[ORGN])"

webEnvSearch <- entrez_search(db = "protein", query, retmax = 999999)
webEnvSearch
length(webEnvSearch$ids)
write.table(webEnvSearch$ids, 
            file = paste(sharedPathAn,"Blast_HK/gilist.txt", sep = ""), 
            append = FALSE, quote = FALSE, row.names = FALSE, col.names = FALSE)
```

This is to prerform a BLAST on the best candidate House Keeping genes
```{r}
ncbiDbPath     <- "/isilon/biodiversity/reference/ncbi/blastdb/reference/nr/nr"
ncbiBlastnPath <- "/opt/bio/ncbi-blast+/bin/blastp"
blastnOutFrmt  <-  c("sseqid", "qseqid", "sacc", "pident", "length",
                     "mismatch", "gapopen", "qstart", "qend", "sstart",
                     "send", "evalue", "bitscore")
#maxTargetsSeqs <- 3
prefix   <- "Blast_HK"  # need to clean up by inserting below
node <- 1

cmd <- paste(ncbiBlastnPath,
             " -db ",     ncbiDbPath,
             " -query ",  paste(sharedPathAn, "Blast_HK/HK_InterestSubset.fasta", sep = ""),
             " -evalue  1.00E-90", 
             " -num_threads ", node,
             " -gilist ", sharedPathAn,   "Blast_HK/gilist.txt",
             " -outfmt '6 ", paste(blastnOutFrmt, collapse=" "), "'",
             " -out ",    paste(sharedPathAn, "Blast_HK/HK_InterestSubsetNt2.tab", sep = ""),
             sep = "")

suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix, node)
```

Andre did the next chunk, it's not working for me due to library("spider") error. 
This is to get the names of the genes and select/sort  the results by BLAST statistics
```{r}
install.packages('http://cran.r-project.org/src/contrib/Archive/Rcpp/Rcpp_0.12.5.tar.gz')
# manually delete: "/home/AAFC-AAC/girouxem/R/x86_64-unknown-linux-gnu-library/3.1/00LOCK-Rcpp"
install.packages('http://cran.r-project.org/src/contrib/Archive/Rcpp/Rcpp_0.12.5.tar.gz')
install.packages("spider")
library("spider")

subsetHKinterestPath <- paste(sharedPath, analysis, 
                              "HiSeq_Analyses/Blast_HK/HK_InterestSubsetNt.tab", 
                              sep = "")

hitsHK <- read.table(subsetHKinterestPath, sep = "\t", header = FALSE,
                     comment.char = "", quote = "", as.is = TRUE) 
colnames(hitsHK) <- blastnOutFrmt

hitSeqHK <- read.GB(hitsHK$sacc, seq.names = hitsHK$sacc, species.names = TRUE, gene = TRUE,
                    access = TRUE, as.character = FALSE)
#rm("hitSeqHK")

hitGenesHK <- data.frame(attr(hitSeqHK, "gene"), attr(hitSeqHK, "accession_num"), 
                           stringsAsFactors = FALSE)

colnames(hitGenesHK) <- c("gene", "sacc")
length(unique(hitGenesHK$sacc))

hitsHK <- merge(unique(hitGenesHK), hitsHK, by  = "sacc",  all.y= TRUE)
temp   <- hitsHK[-grep("unnamed|hypothetical",hitsHK$gene),]
temp   <- temp[order(temp$qseqid,temp$evalue,-temp$bitscore),] 

write.table(temp, 
            file = file.path(paste(sharedPathAn, sep=""),"potential_houseKeepingGenes_BLAST.csv"),
            append = FALSE, sep = ",", col.names = NA)
```
>>>>>>> b272e7c0e0c1cf0ccd5efa8628919339a1d1fff8


This pulls the first line of each of the BLAST hits.  Because they were sorted, the one with 
the lowest evalue and highest bit score is kept.
```{r}
# Emily had to skip some above and import Andre's tables directly 
# (import temp, from csv read.csv(), collect names, then):
genesHKNames <- paste(sharedPath, analysis, "HiSeq_Analyses/potential_houseKeepingGenes_BLAST.csv", 
                      sep = "")
temp         <- read.csv(genesHKNames, stringsAsFactors = FALSE)
bestHitHK    <- temp[!duplicated(temp$qseqid), ]  # duplicate removal
bestHitHK$HK_gene <- sub("_\\[.*","",bestHitHK$gene)
rpkmHKCount  <- rpkmCount[bestHitHK$qseqid, ]

rpkmHKforBarPlots <- melt(rpkmHKCount, id.vars = colnames(rpkmHKCount))
colnames(rpkmHKforBarPlots)[1] <- "SequenceID"
colnames(rpkmHKforBarPlots)[2] <- "ExpUnit"
colnames(rpkmHKforBarPlots)[3] <- "Read_Counts"

# temp <- merge(rpkmHKforBarPlots, bestHitHKs, by  = "sacc",  all.y= TRUE)

# I took out library name because we added two libraries
metadataColsToMerge <- c("TimePoint", "RNASeq_Replicate", "ExpUnit") 

# For paired-end not merged:
rpkmHKforBarPlots <- merge(rpkmHKforBarPlots, 
                           unique(metadataAdapRem[,c(metadataColsToMerge)]), 
                           by = "ExpUnit", all.x = TRUE)

# # For merged or single-end reads:
# rpkmHKforBarPlots <- merge(rpkmHKforBarPlots, unique(metadataAdapRM[,c(metadataColsToMerge)]), 
#                            by = "ExpUnit", all.x = TRUE)

unique(rpkmHKforBarPlots$ExpUnit) # Notice ExpUnit "y"
rpkmHKforBarPlots <- rpkmHKforBarPlots[rpkmHKforBarPlots$ExpUnit!="y",]
unique(rpkmForBarPlots$ExpUnit) # Notice ExpUnit "y" is gone

rpkmHKforBarPlots$TimePoint <- factor(rpkmHKforBarPlots$TimePoint, levels = c(timePoints))        

rpkmHKforBarPlots <- merge(rpkmHKforBarPlots, bestHitHK[,c("HK_gene", "qseqid")],
                           by.x = "SequenceID", by.y = "qseqid", all.x = TRUE)

rpkmHKforBarPlots$House_Keeping_genes <- paste(rpkmHKforBarPlots$SequenceID,
                                               rpkmHKforBarPlots$HK_gene, sep = "-")
```

**** Over here there is a problem - for Andre ( I left a useless loop there - sorry! André)
Housekeeping Bar plots
```{r}
library("ggplot2")
dir.create(paste(sharedPathAn, "HK_ggplots", sep = ""), showWarnings = TRUE, recursive = FALSE)
ggplotsPath <- paste(sharedPathAn, "HK_ggplots/", sep = "")

pdf(file   = paste(ggplotsPath, "HK_plots", ".pdf", sep = ""), 
    width  = 8, 
    height = 24)
bp  <- ggplot(aes(y = Read_Counts, x = TimePoint), data = rpkmHKforBarPlots) + 
  theme_bw() + 
  scale_y_log10(breaks=c(10, 100,1000)) +
  theme(axis.text.y = element_text(size  = 6, 
                                   hjust = 1, 
                                   vjust = 0.4)) +
  theme(axis.text.x = element_text(colour = 'black', 
                                   size   = 6, 
                                   angle  = 45, 
                                   hjust  = 1, 
                                   vjust  = 1)) +
  theme(axis.ticks = element_line(colour = 'black', 
                                  size   = 0.5)) +
  labs(y = "Total Number of Reads (rpkm-normalized)") +
  theme(axis.title.y = element_text(colour = 'black', 
                                    size   = 12, 
                                    angle  = 90, 
                                    hjust  = 0.5, 
                                    vjust  = 0.2, 
                                    face   = 'bold')) +
  theme(axis.title.x = element_text(colour = 'black', 
                                    size   = 12, 
                                    angle  = 0, 
                                    hjust  = 0.5, 
                                    vjust  =  -0.2, 
                                    face   = 'bold')) +
  geom_boxplot(position = position_dodge(0.8), 
               width    = 0.8, 
               outlier.size   = 1, 
               outlier.colour = "black", 
               outlier.shape  = 20) +
  stat_summary(fun.y = mean, 
               geom  = "point", 
               shape = 5, 
               size  = 1, 
               position = position_dodge(0.8))  +
  facet_wrap(~House_Keeping_genes,
             ncol = 4) +
  theme(strip.text.x = element_text(size   = 4, 
                                    face   = 'bold', 
                                    colour = "black", 
                                    angle  = 0))
print(bp)
dev.off()

```

<<<<<<< HEAD



=======
>>>>>>> b272e7c0e0c1cf0ccd5efa8628919339a1d1fff8
Bar plots per CAZy family group:
```{r}
library("ggplot2")

dirPlotName <- "ggplotsOosCon_and_OoGen"

# dir.create(paste(sharedPathAn, "ggplotsByCazyFam", sep = ""),
#            showWarnings = TRUE,
#            recursive    = FALSE)

# For plots for both time courses merged:
dir.create(paste(sharedPath, dirPlotName, sep = ""),
           showWarnings = TRUE,
           recursive    = FALSE)

#ggplotsByCazyFamPath <- paste(sharedPathAn, "ggplotsByCazyFam/", sep = "")
ggplotsByCazyFamPath <- paste(sharedPath, dirPlotName, "/", sep = "")

i <- 1

for(i in 1:length(cazyFam)){
# temp <- subset(cazyRpkmSub2, cazyRpkmSub2$CazyFam == cazyFam[i])
temp <- subset(cazyRpkmSub4, cazyRpkmSub4$CazyFam == cazyFam[i])

pdf(file   = paste(ggplotsByCazyFamPath, cazyFam[i], ".pdf", sep = ""),
    width  = 8, 
    height = 7*ceiling(nrow(temp)/90))

bp <- ggplot(aes(y = Read_Counts, x = Timepoint_Condition), data  = temp) +
  theme_bw() +
  theme(axis.text.y = element_text(size  = 6, 
                                   hjust = 1, 
                                   vjust = 0.4)) +
  theme(axis.text.x = element_text(colour = 'black', 
                                   size   = 6, 
                                   angle  = 45, 
                                   hjust  = 1, 
                                   vjust  = 1)) +
  theme(axis.ticks = element_line(colour = 'black', 
                                  size   = 0.5)) +
  labs(y = "Total Number of Reads (rpkm-normalized)") +
  theme(axis.title.y = element_text(colour = 'black', 
                                    size   = 12, 
                                    angle  = 90, 
                                    hjust  = 0.5, 
                                    vjust  = 0.2, 
                                    face   = 'bold')) +
  theme(axis.title.x = element_text(colour = 'black', 
                                    size   = 12, 
                                    angle  = 0, 
                                    hjust  = 0.5, 
                                    vjust  =  -0.2, 
                                    face   = 'bold')) +
  geom_boxplot(position = position_dodge(0.8), 
               width    = 0.8, 
               outlier.size   = 1, 
               outlier.colour = "black", 
               outlier.shape  = 20) +
  stat_summary(fun.y = mean, 
               geom  = "point", 
               shape = 5, 
               size  = 1, 
               position = position_dodge(0.8))  +
#   facet_wrap(~SequenceID, 
#   facet_grid(SequenceID~Experiment,
  facet_grid(SequenceID~Experiment,
             scales = "free") +
  theme(strip.text.x = element_text(size   = 6, 
                                    face   = 'bold.italic', 
                                    colour = "black", 
                                    angle  = 0))
print(bp)
dev.off()
}

```

Individual multiple Bar plots
```{r}
library("ggplot2")
dir.create(paste(sharedPathAn, "ggplots", sep = ""), showWarnings = TRUE, recursive = FALSE)
ggplotsPath <- paste(sharedPathAn, "ggplots/", sep = "")

cazyRpkmSub2$Secreted_CAZy <- paste(cazyRpkmSub2$CazyFam, cazyRpkmSub2$SequenceID, sep = "-")
forPlot <- unique(cazyRpkmSub2$Secreted_CAZy) 
i <- 1
for(i in 1:length(forPlot)){
  temp <- subset(cazyRpkmSub2, cazyRpkmSub2$Secreted_CAZy == forPlot[i])
  png(file   = paste(ggplotsPath, forPlot[i], "-secreted.png", sep = ""), 
      width  = 3, 
      height = 3, 
      units  = "in", 
      res    = 300, 
      bg     = "white")
  bp <- ggplot(aes(y = Read_Counts, x = TimePoint), data = temp) +
    theme_bw() +
    theme(axis.text.y = element_text(size  = 6, 
                                     hjust = 1, 
                                     vjust = 0.5)) +
    theme(axis.text.x = element_text(colour = 'black', 
                                     size   = 6, 
                                     angle  = 0, 
                                     hjust  = 0.5, 
                                     vjust  = 1)) +
    theme(axis.ticks = element_line(colour = 'black', 
                                    size   = 0.5)) +
    labs(y = "Total Number of Reads (rpkm-normalized)") +
    theme(axis.title.y = element_text(colour = 'black', 
                                      size   = 7, 
                                      angle  = 90, 
                                      hjust  = 0.5, 
                                      vjust  = 0.5, 
                                      face   = 'bold')) +
    theme(axis.title.x = element_text(colour = 'black', 
                                      size   = 7, 
                                      angle  = 0, 
                                      hjust  = 0.5, 
                                      vjust  = 0.5, 
                                      face   = 'bold')) +
    geom_boxplot(position = position_dodge(0.8), 
                 width    = 0.8, 
                 outlier.size   = 1, 
                 outlier.colour = "black", 
                 outlier.shape  = 20) +
    stat_summary(fun.y = mean, 
                 geom  = "point", 
                 shape = 5, 
                 size  = 1, 
                 position = position_dodge(0.8))  +
    facet_wrap(~Secreted_CAZy, 
               scales = "free", 
               ncol   = 1) +
    theme(strip.text.x = element_text(size   = 6, 
                                      face   = 'bold', 
                                      colour = "black", 
                                      angle  = 0))
  print(bp)
  dev.off()
}
```





                      

