RNA-Seq Read Processing and Alignment with TopHat2
===============================================================

Installing the required packages for R:
This is to be done in the R command line and not in R studio
source("http://www.Bioconductor.org/biocLite.R")
biocLite("BiocUpgrade")

Getting started in R:
Set the working directory > setwd("~/")
Check version installed
```{r}
library(knitr)
```

This will help us when finding our files to source functions:
```{r}
install.packages("rprojroot")
library(rprojroot)
# We specify ours is an RStudio project
# The root object contains a function that will help us locate our package r files
# regarless of our current working directory
root <- rprojroot::is_rstudio_project
scriptsPath <- root$make_fix_file(".")("R")
scripts <- dir(root$find_file("R", path = root$find_file()))
scriptsl <- paste(scriptsPath, scripts, sep = "//")
lapply(scriptsl, source)
```

Setting up the system environment to run qsub jobs from within RStudio
```{r}
Sys.setenv(SGE_ROOT="/opt/gridengine")
Sys.setenv(SGE_CELL="default")
Sys.setenv(SGE_ARCH="linux-x64")
Sys.setenv(SGE_EXECD_PORT="537")
Sys.setenv(SGE_QMASTER_PORT="536")
```

Many of the commands and approach are from:
Simon Anders et al., 2013. Count-based differential expression analysis of 
RNA sequencing data using R and Bioconductor. Nature protocols | VOL.8 NO.9 | 2013

Chunk: Analysis and Sequence Data Directory Setting
User:
Define the the folder in the shared folder that will hold the analyses of the time-course/dataset 
you will be working with. In our case, we have two different time-course experiments, Oosporogenesis 
and Oospore Conversion. Below we set which one the script will run analyses for. We also get the 
user to specify what the name of the directory that will hold the reads will be.
```{r}
# Define the path to the shared folder where the main working directory will be.
sharedPath <- "/isilon/biodiversity/users/shared/Pythium_ultimum_RNAseq/"

# What is the name of the overall experiment?
analysis   <- "Oospore_Conversion_TimeCourse/"

# Name of the directory you'll be keeping all the analyses for this pipeline:
# workDir    <- "MiSeq_Analyses_Paenibacillus_polymyxa"
# workDir    <- "HiSeq_Analyses2_Genomic_Mito"
workDir    <- "HiSeq_Analyses2_Paenibacillus_polymyxa"



#Name of the directory to keep the fastq files in:
# seqDataDir <- "MiSeq_data_Sci2"
seqDataDir <- "HiSeq_data2"

```

```{r}
sharedPath <- "/isilon/biodiversity/users/shared/Pythium_ultimum_RNAseq/"
# What is the name of the overall experiment?
analysis   <- "Oospore_Conversion_TimeCourse/"

# Name of the directory you'll be keeping all the analyses for this pipeline:
seqDataDir <- "MiSeq_data_Sci2"




workDir    <- "HiSeq_Analyses2"
# Set the path:
sharedPathAn <- paste(sharedPath, analysis, workDir, "/", sep ="")


```



Metadata oospore conversion
```{r}
# Specify which metadata table you've been using:
# finalMetadata <- metadataAdapRem
finalName_conv <- "OosporeConversion_Final_metadataTable_afterProcessing.csv"
# write.table(finalMetadata, file = finalName, 
#             append = FALSE, quote = FALSE, 
#             sep = ",", row.names = FALSE)
finalMetadata_conv <- read.csv(paste(sharedPath,"Oospore_Conversion_TimeCourse/", "HiSeq_Analyses2/", finalName_conv, sep=""),
                       stringsAsFactors = FALSE)

metadataAdapRem_conv <- finalMetadata_conv

metadataAdapRem_conv$AdapRMDustMinLen <- metadataAdapRem_conv$processed5.R1.Fastq

metadataAdapRem_conv$AdapRMDustMinLen <- gsub(".5processed\\.R1\\.fastq", ".5processed.log", 
                                        metadataAdapRem_conv$AdapRMDustMinLen)


```
Metadata oosporogenesis
```{r}
# Specify which metadata table you've been using:
# finalMetadata <- metadataAdapRem
#finalName <- "OosporeConversion_Final_metadataTable_afterProcessing.csv"
finalName_gen <- "Oosporogenesis_Final_metadataTable_afterProcessing.csv"
# write.table(finalMetadata, file = finalName, 
#             append = FALSE, quote = FALSE, 
#             sep = ",", row.names = FALSE)
finalMetadata_gen <- read.csv(paste(sharedPath,finalName_gen, sep=""), stringsAsFactors = FALSE)
metadataAdapRem_gen <- finalMetadata_gen
```


** Oospore conversion **
Ready to read the log file from all print seq logs

```{r}
printSeq <- "I_PrinSeqDustMinLen"
workDir    <- "HiSeq_Analyses2"
analysis   <- "Oospore_Conversion_TimeCourse/"

read_stats_conv <- data.frame()
j <- 1
for(j in 1:nrow(metadataAdapRem_conv)) {

  my_reads <- read.table(paste(sharedPath,analysis, workDir,"/", printSeq, "/", metadataAdapRem_conv$AdapRMDustMinLen[j], sep=""), sep="\t",
                       skip=3, stringsAsFactors = FALSE)
good_read_line <- my_reads[grep("Good sequences \\(pairs\\)\\:",my_reads[,1]),]
good_sequences <- sub("\\[.*\\)\\:\\ ","", good_read_line)
good_sequences <- as.numeric(gsub(",", "", good_sequences))
temp <- data.frame(metadataAdapRem_conv[j,1:4], metadataAdapRem_conv$AdapRMDustMinLen[j], good_sequences, stringsAsFactors = FALSE)
read_stats_conv <- rbind(read_stats_conv,temp)
}

```

** Oosporogenesis **
Ready to read the log file from all print seq logs

```{r}
printSeq_genesis <- "F_Print_Seq_trim"


read_stats_gen <- data.frame()
j <- 1
for(j in 1:nrow(metadataAdapRem_gen)) {

  my_reads_genesis <- read.table(paste(sharedPath, printSeq_genesis, "/", 
                                       "Processed_log_", metadataAdapRem_gen$LibraryName[j], sep=""), sep="\t",
                       skip=3, stringsAsFactors = FALSE)
good_read_line_g <- my_reads_genesis[grep("Good sequences \\(pairs\\)\\:",my_reads_genesis[,1]),]
good_sequences_g <- sub("\\[.*\\)\\:\\ ","", good_read_line_g)
good_sequences_g <- as.numeric(gsub(",", "", good_sequences_g))
temp_g <- data.frame(metadataAdapRem_gen[j,1:4], paste("Processed_log_",metadataAdapRem_gen$LibraryName[j],sep=""), good_sequences_g, stringsAsFactors = FALSE)
read_stats_gen <- rbind(read_stats_gen,temp_g)
}

colnames(read_stats_gen)[colnames(read_stats_gen) == "good_sequences_g"] <- "good_sequences"

```


Read counts from SamTools for oospore conversion - P ultimum
```{r}
library("edgeR")
# Identify the count files and read them into R using readDGE:

# What is the name of the overall experiment?
analysis   <- "Oospore_Conversion_TimeCourse/"

# Name of the directory you'll be keeping all the analyses for this pipeline:
#workDir    <- "MiSeq_Analyses_Paenibacillus_polymyxa"
workDir2    <- "HiSeq_Analyses2_Genomic_Mito"
#workDir    <- "HiSeq_Analyses2_Paenibacillus_polymyxa"

topHatDate <- ".2017-07-19"
stranded <- "no"
MINAQUAL <- 10
prefix   <- "L_HTSeq_Qsub"

# For paired-end not merged:
countsListTopHat <- paste(sharedPath, analysis, workDir2, "/", metadataAdapRem_conv$LibraryName, "/",
                          metadataAdapRem_conv$LibraryName,".TopHat",topHatDate,"/",
                          metadataAdapRem_conv$countfTopHat, "_MINAQUAL_" ,MINAQUAL, "_stranded_NO", sep = "")

# # For merged or single-end reads:
# countsListTopHat <- paste(sharedPathAn, metadataAdapRM$LibraryName, "/", metadataAdapRM$LibraryName,
#                           ".TopHat", topHatDate, "/", metadataAdapRM$countfTopHat, sep = "")

# Turn the count of the list into a data.frame
countsTopHat_Pult <- readDGE(countsListTopHat, header=FALSE)$counts
colnames(countsTopHat_Pult) <- basename(colnames(countsTopHat_Pult))
colnames(countsTopHat_Pult) <- sub("\\.TopHat.*","",colnames(countsTopHat_Pult))

grep("PYU1_R", rownames(countsTopHat_Pult))

colSums(countsTopHat_Pult[grep("PYU1_R", rownames(countsTopHat_Pult)),])

Pult_stat_conversion <- data.frame("Oospore_conversion",colnames(countsTopHat_Pult), 
                                   colSums(countsTopHat_Pult[grep("PYU1_R", rownames(countsTopHat_Pult)),]),
                                   colSums(countsTopHat_Pult[grep("rna5|rna17", rownames(countsTopHat_Pult)),]),
                                   colSums(countsTopHat_Pult[grep("rna1|rna2|rna3|rna4|rna6|rna7|rna8|rna9|rna10|rna11|rna12|rna13|rna14|rna15|rna16|rna18|rna19|rna20|rna21|rna22|rna23|rna24|rna25", rownames(countsTopHat_Pult)),]),
                                   colSums(countsTopHat_Pult[grep("PYU1_T", rownames(countsTopHat_Pult)),]),
                                   countsTopHat_Pult[grep("__no_feature", rownames(countsTopHat_Pult)),],
                                   countsTopHat_Pult[grep("__not_aligned", rownames(countsTopHat_Pult)),],
                                   countsTopHat_Pult[grep("__alignment_not_unique", rownames(countsTopHat_Pult)),])
colnames(Pult_stat_conversion) <- c("experiment","LibraryName","nuc_ribo","mito_ribo","mito_coding","P_ult_nuc_genes","no_feature","not_aligned","alignment_not_unique")


```

Read counts from SamTools for oospore conversion - Paenibacillus
```{r}
library("edgeR")
# Identify the count files and read them into R using readDGE:

# What is the name of the overall experiment?
analysis   <- "Oospore_Conversion_TimeCourse/"

# Name of the directory you'll be keeping all the analyses for this pipeline:
#workDir    <- "MiSeq_Analyses_Paenibacillus_polymyxa"
#workDir2    <- "HiSeq_Analyses2_Genomic_Mito"
#workDir2    <- "HiSeq_Analyses2_Paenibacillus_polymyxa"
workDir2    <- "HiSeq_Analyses2_Paenibacillus_from_RNAseq"


topHatDate <- ".2017-08-08"
stranded <- "no"
MINAQUAL <- 10
prefix   <- "L_HTSeq_Qsub"

# For paired-end not merged:
countsListTopHat_bac <- paste(sharedPath, analysis, workDir2, "/", metadataAdapRem_conv$LibraryName, "/",
                          metadataAdapRem_conv$LibraryName,".TopHat",topHatDate,"/",
                          metadataAdapRem_conv$countfTopHat, "_MINAQUAL_" ,MINAQUAL, "_stranded_NO", sep = "")

# # For merged or single-end reads:
# countsListTopHat <- paste(sharedPathAn, metadataAdapRM$LibraryName, "/", metadataAdapRM$LibraryName,
#                           ".TopHat", topHatDate, "/", metadataAdapRM$countfTopHat, sep = "")

# Turn the count of the list into a data.frame
countsTopHat_PaBac <- readDGE(countsListTopHat_bac, header=FALSE)$counts
colnames(countsTopHat_PaBac) <- basename(colnames(countsTopHat_PaBac))
colnames(countsTopHat_PaBac) <- sub("\\.TopHat.*","",colnames(countsTopHat_PaBac))

PaBac_conv <- data.frame(colnames(countsTopHat_PaBac),colSums(countsTopHat_PaBac))
colnames(PaBac_conv) <- c("LibraryName","Paenibacillus")

```

Merging conversion data
```{r}

Pult_stat_conversion <- merge(Pult_stat_conversion,PaBac_conv, by="LibraryName")

Pult_stat_conversion$total_mapped <- rowSums(Pult_stat_conversion[,c(3:7,9:10)])

Pult_stat_conversion <- merge(Pult_stat_conversion,read_stats_conv[,c(1,6)], by="LibraryName")

Pult_stat_conversion$unmapped <- Pult_stat_conversion$good_sequences - Pult_stat_conversion$total_mapped

Pult_stat_conversion$single_library <- substr(Pult_stat_conversion$LibraryName,1,nchar(as.character(Pult_stat_conversion$LibraryName))-3)

Pult_stat_conversion$Time <- gsub("T|\\-.*","",Pult_stat_conversion$LibraryName)

library(reshape2)

Pult_stat_conversion_agg_lib <- aggregate(. ~ single_library, data=Pult_stat_conversion[,3:14], FUN=sum)
Pult_stat_conversion_agg_time <- aggregate(. ~ Time, data=Pult_stat_conversion[,c(3:13,15)], FUN=sum)
Pult_stat_conversion_agg_time$Experiment <- "Conversion"

Pult_stat_conversion_melt <- melt(Pult_stat_conversion_agg_time[,c(1:5,9)], value.name = "Read_Counts")


```


```{r}

Pult_stat_conversion_melt <- Pult_stat_conversion_melt[grep("T7-", Pult_stat_conversion_melt$single_library),]
Pult_stat_conversion_agg <- Pult_stat_conversion_agg[grep("T7-", Pult_stat_conversion_agg$single_library),]
Pult_stat_conversion_agg$y <- 1.02

temp <- data.frame(x=c("T7-c3_BC29", "T7-c3_BC29", "T7-c3_BC29"), y=c(1.02, 1.02, 1.02), z=c(51, 101, 348))

library(ggplot2)
pdf(file = paste(sharedPath, "GGPLOT_basic_stats/conversion stats2.pdf",sep=""), width = 6, height =6 )
bp <- ggplot(Pult_stat_conversion_melt, aes(x =single_library, y = Read_Counts, fill = variable)) +
# geom_bar(stat = "identity", width=1) +
  geom_bar(stat = "identity",  position ="fill", color = "black") + 
   geom_col()+   
#  geom_text(aes(label = "AA"), size = 3, hjust = 0.5, vjust = 3, position =     "stack") +
 #  geom_text(aes(label = "AA"),   position = position_stack(vjust = 0.5))+
    #    geom_text(data=temp, aes(x=x, y=y, label=as.factor(z))) +
 #    geom_text(aes(y = Read_Counts, label = Read_Counts,  position ="fill"), vjust = 3) +
 # geom_text(label= "A")+
 # geom_text(data=Pult_stat_conversion_agg, mapping=aes(x=single_library, y=100, label="A"), size=4, vjust=3, hjust=0.5) +
 # geom_text(Pult_stat_conversion_agg, aes(label = ..count.., y= ..prop..), stat= "count", vjust = -.5) +
 # geom_text(Pult_stat_conversion_agg ,aes(y=..count.., label=single_library), vjust=-1) +
#
 # geom_text(aes(y = ..count.., label= ..prop..), stat= "count", vjust = -.5) +
 #   geom_text(aes(y = ..count.., label= ..prop..), stat= "count", vjust = -.5) +
 # geom_text(stat = "count", aes(label = ..count.., y = ..count..), vjust = -.5) +
#  geom_text(stat='count',aes(label=..count..),vjust=-1) +
#  geom_text(data=Pult_stat_conversion_agg, aes(x = as.factor(single_library),  y =y , label = as.factor(total_mapped))) +
  #   geom_text(data=temp, aes(x=x, y=y, label=as.factor(z)))
 # geom_text(stat='count',aes(label=..count..),vjust=-1) +
 # geom_bar(stat = "identity",  position = position_dodge()) +
  scale_y_continuous(labels = scales::percent) +
  theme_bw()+
#  scale_y_log10() +
  theme(axis.text.y = element_text(size  = 12, 
                                   hjust = 1, 
                                   vjust = 0.4)) +
  theme(axis.text.x = element_text(colour = 'black', 
                                   size   = 12, 
                                   angle  = 45, 
                                   hjust  = 1, 
                                   vjust  = 1)) +
  theme(axis.ticks = element_line(colour = 'black', 
                                  size   = 0.5)) +
  labs(y = "Percent of Number of Reads\n (rpkm-normalized)") + 
  labs(x = "Treatment") 
print(bp)
dev.off()

    set.seed(123)
    cat1<-sample(letters[1:3], 500, replace=T, prob=c(0.1, 0.2, 0.65))
    cat2<-sample(letters[4:8], 500, replace=T, prob=c(0.3, 0.4, 0.75, 0.5, 0.1))
    df <- data.frame(cat1, cat2)

temp <- data.frame(x=c("a", "b", "c"), y=c(1.02, 1.02, 1.02), z=c(51, 101, 348))

   ggplot(df, aes(x=cat1))+
    geom_bar(aes(fill = cat2),
                    position = 'fill',color = "black")+
    scale_y_continuous(labels = scales::percent)+
    labs ( y = "Percentage")+
      # this final line is me trying to add the label
      geom_text(data=temp, aes(x=x, y=y, label=as.factor(z)))
   
library(dplyr)

d <- data.frame(ageCat = rep(1:5, each=20)
                ,catCRP = rep(1:5,20))

p <- d %>% group_by(ageCat, catCRP) %>% summarise(n = n()) %>%
 mutate(p = n/sum(n))

p1 <- ggplot(p,aes(x=ageCat,y=p, fill=catCRP))+geom_bar(stat = "identity", position="fill") + theme_bw() 
p1 + geom_text(aes(label = paste0(n," (",p*100,"%)")), size = 3, hjust = 0.5, vjust = 3, position =     "stack")```

```


** Read counts from SamTools for oosporogenesis - P ultimum **
```{r}
library("edgeR")
# Identify the count files and read them into R using readDGE:

# What is the name of the overall experiment?
analysis   <- "Oosporogenesis_new_references/"

# Name of the directory you'll be keeping all the analyses for this pipeline:
#workDir    <- "MiSeq_Analyses_Paenibacillus_polymyxa"
workDir2    <- "HiSeq_Analyses2_Genomic_Mito"
#workDir    <- "HiSeq_Analyses2_Paenibacillus_polymyxa"

topHatDate <- ".2017-08-08"
stranded <- "no"
MINAQUAL <- 10
prefix   <- "L_HTSeq_Qsub"

# For paired-end not merged:
countsListTopHat_gen <- paste(sharedPath, analysis, workDir2, "/", metadataAdapRem_gen$LibraryName, "/",
                          metadataAdapRem_gen$LibraryName,".TopHat",topHatDate,"/",
                          metadataAdapRem_gen$countfTopHat, "_MINAQUAL_" ,MINAQUAL, "_stranded_NO", sep = "")

# # For merged or single-end reads:
# countsListTopHat <- paste(sharedPathAn, metadataAdapRM$LibraryName, "/", metadataAdapRM$LibraryName,
#                           ".TopHat", topHatDate, "/", metadataAdapRM$countfTopHat, sep = "")

# Turn the count of the list into a data.frame
countsTopHat_Pult_gen <- readDGE(countsListTopHat_gen, header=FALSE)$counts
colnames(countsTopHat_Pult_gen) <- basename(colnames(countsTopHat_Pult_gen))
colnames(countsTopHat_Pult_gen) <- sub("\\.TopHat.*","",colnames(countsTopHat_Pult_gen))

grep("PYU1_R", rownames(countsTopHat_Pult_gen))

colSums(countsTopHat_Pult_gen[grep("PYU1_R", rownames(countsTopHat_Pult_gen)),])

Pult_stat_genesis <- data.frame("Oospore_genesis",colnames(countsTopHat_Pult_gen), 
                                   colSums(countsTopHat_Pult_gen[grep("PYU1_R", rownames(countsTopHat_Pult_gen)),]),
                                   colSums(countsTopHat_Pult_gen[grep("rna5|rna17", rownames(countsTopHat_Pult_gen)),]),
                                   colSums(countsTopHat_Pult_gen[grep("rna1|rna2|rna3|rna4|rna6|rna7|rna8|rna9|rna10|rna11|rna12|rna13|rna14|rna15|rna16|rna18|rna19|rna20|rna21|rna22|rna23|rna24|rna25", rownames(countsTopHat_Pult_gen)),]),
                                   colSums(countsTopHat_Pult_gen[grep("PYU1_T", rownames(countsTopHat_Pult_gen)),]),
                                   countsTopHat_Pult_gen[grep("__no_feature", rownames(countsTopHat_Pult_gen)),],
                                   countsTopHat_Pult_gen[grep("__not_aligned", rownames(countsTopHat_Pult_gen)),],
                                   countsTopHat_Pult_gen[grep("__alignment_not_unique", rownames(countsTopHat_Pult_gen)),])
colnames(Pult_stat_genesis) <- c("experiment","LibraryName","nuc_ribo","mito_ribo","mito_coding","P_ult_nuc_genes","no_feature","not_aligned","alignment_not_unique")


```

Read counts from SamTools for oosporogenesis - Paenibacillus
```{r}
library("edgeR")
# Identify the count files and read them into R using readDGE:

# What is the name of the overall experiment?
analysis   <- "Oosporogenesis_new_references/"

# Name of the directory you'll be keeping all the analyses for this pipeline:
#workDir    <- "MiSeq_Analyses_Paenibacillus_polymyxa"
#workDir2    <- "HiSeq_Analyses2_Genomic_Mito"
#workDir2    <- "HiSeq_Analyses2_Paenibacillus_polymyxa"
workDir2    <- "HiSeq_Analyses2_Paenibacillus_from_RNAseq"


topHatDate <- ".2017-08-08"
stranded <- "no"
MINAQUAL <- 10
prefix   <- "L_HTSeq_Qsub"

# For paired-end not merged:
countsListTopHat_gen_bac <- paste(sharedPath, analysis, workDir2, "/", metadataAdapRem_gen$LibraryName, "/",
                          metadataAdapRem_gen$LibraryName,".TopHat",topHatDate,"/",
                          metadataAdapRem_gen$countfTopHat, "_MINAQUAL_" ,MINAQUAL, "_stranded_NO", sep = "")

# # For merged or single-end reads:
# countsListTopHat <- paste(sharedPathAn, metadataAdapRM$LibraryName, "/", metadataAdapRM$LibraryName,
#                           ".TopHat", topHatDate, "/", metadataAdapRM$countfTopHat, sep = "")

# Turn the count of the list into a data.frame
countsTopHat_PaBac_gen <- readDGE(countsListTopHat_gen_bac, header=FALSE)$counts
colnames(countsTopHat_PaBac_gen) <- basename(colnames(countsTopHat_PaBac_gen))
colnames(countsTopHat_PaBac_gen) <- sub("\\.TopHat.*","",colnames(countsTopHat_PaBac_gen))

PaBac_gen <- data.frame(colnames(countsTopHat_PaBac_gen),colSums(countsTopHat_PaBac_gen))
colnames(PaBac_gen) <- c("LibraryName","Paenibacillus")

```

Merging oosporogenesis data
```{r}

Pult_stat_genesis <- merge(Pult_stat_genesis,PaBac_gen, by="LibraryName")

Pult_stat_genesis$total_mapped <- rowSums(Pult_stat_genesis[,c(3:7,9:10)])

Pult_stat_genesis <- merge(Pult_stat_genesis,read_stats_gen[,c(1,6)], by="LibraryName")

Pult_stat_genesis$unmapped <- Pult_stat_genesis$good_sequences - Pult_stat_genesis$total_mapped

#because number is negative
Pult_stat_genesis$unmapped <- 0

Pult_stat_genesis$single_library <- Pult_stat_genesis$LibraryName

Pult_stat_genesis$Time <- gsub("T|[0-9]\\_.*","",Pult_stat_genesis$LibraryName)

library(reshape2)

Pult_stat_genesis_agg_lib <- aggregate(. ~ single_library, data=Pult_stat_genesis[,3:14], FUN=sum)
Pult_stat_genesis_agg_time <- aggregate(. ~ Time, data=Pult_stat_genesis[,c(3:13,15)], FUN=sum)

Pult_stat_genesis_melt <- melt(Pult_stat_genesis_agg_time[,c(1:5,9)], value.name = "Read_Counts")

Pult_stat_genesis_agg_time$Experiment <- "Genesis"


```


Combine Oosporogenesis and Conversion
```{r}


Two_experiments <- rbind(Pult_stat_genesis_agg_time,Pult_stat_conversion_agg_time)

Two_experiments <- Two_experiments[!(Two_experiments$Time =="7" | Two_experiments$Time =="neg"),]

Two_experiments$Trt <- sub("^.*\\-|[0-9]*","",Two_experiments$Time)
Two_experiments$Time <- as.numeric(sub("\\-.*","",Two_experiments$Time))

Two_experiments <- Two_experiments[order(-rank(Two_experiments$Experiment),Two_experiments$Trt,Two_experiments$Time),]

for(j in 1:nrow(Two_experiments)) {
if (Two_experiments$Experiment[j] == "Conversion") {
Two_experiments$everything_mapped[j] <- Two_experiments$nuc_ribo[j] + Two_experiments$mito_ribo[j] +
  Two_experiments$mito_coding[j] + Two_experiments$P_ult_nuc_genes[j] + Two_experiments$no_feature[j] +
  Two_experiments$alignment_not_unique[j] + Two_experiments$Paenibacillus[j]
} else {
 Two_experiments$everything_mapped[j] <-  Two_experiments$good_sequences[j]
}}

Two_experiments$percent_everything_mapped <- Two_experiments$everything_mapped/Two_experiments$good_sequences

Two_experiments$P_ult_Total_ribo <- Two_experiments$nuc_ribo + Two_experiments$mito_ribo
Two_experiments$percent_Pult_all_ribo_genes_mapped_over_all_reads <- Two_experiments$P_ult_Total_ribo /Two_experiments$good_sequences
Two_experiments$percent_Pult_all_ribo_genes_mapped_over_all_mapped <- Two_experiments$P_ult_Total_ribo /Two_experiments$everything_mapped

Two_experiments$percent_Pult_nuc_coding_genes_mapped_over_all_reads <- Two_experiments$P_ult_nuc_genes/Two_experiments$good_sequences
Two_experiments$percent_Pult_nuc_coding_genes_mapped_over_all_mapped <- Two_experiments$P_ult_nuc_genes/Two_experiments$everything_mapped

Two_experiments$percent_Paenibacillus_contigs_mapped_over_total_reads <- Two_experiments$Paenibacillus/Two_experiments$good_sequences
Two_experiments$percent_Paenibacillus_contigs_mapped_over_total_mapped <- Two_experiments$Paenibacillus/Two_experiments$everything_mapped

write.table(Two_experiments, paste(sharedPath, "GGPLOT_basic_stats/Counts_for_genesis_and_conversion.csv",sep=""),  append = FALSE, quote = FALSE,sep = ",", row.names = FALSE)



```

