Generating pairwise plots of maximum gene expression between oosporogenesis and oospore conversion
======================================================================

Path to the shared folder:
```{r}
sharedPath <- "/isilon/biodiversity/users/shared/Pythium_ultimum_RNAseq/"
```

Read file with all the data.  This file was made after normalizing oosporogenesis and oospore conversion together with EdgeR
```{r}
rpkmPlotsFullImport <- read.table(paste(sharedPath, 
                                        "rpkmForBarPlots_full.csv", 
                                        sep = ""), 
                                  sep = ",", 
                                  header = TRUE, 
                                  as.is = TRUE, 
                                  stringsAsFactors = FALSE)

#This is to create a lingle variable that creates a single variable that separates the 7 oosporegenesis treatments
rpkmPlotsFullImport$xLab <- paste(rpkmPlotsFullImport$Condition,
                                  rpkmPlotsFullImport$TimePoint, 
                                  sep = "_")
unique(rpkmPlotsFullImport$xLab)

# To clean up the other experiments
rpkmPlotsFullImport$xLab <- sub("OoCTimeCourse_", "", rpkmPlotsFullImport$xLab)
rpkmPlotsFullImport$xLab <- sub("Various_0", "Various", rpkmPlotsFullImport$xLab)
unique(rpkmPlotsFullImport$xLab)
unique(rpkmPlotsFullImport$Experiment)

# Creates orderes factors (very important for plotting)
labOrdered <- c("Various", "Control_0", "Control_3", "Cholesterol_3", 
                "Control_7", "Cholesterol_7", "Control_24", "Cholesterol_24", 
                "0", "12", "24", "48", "72", "120", "240") 
rpkmPlotsFullImport$xLab <- factor(rpkmPlotsFullImport$xLab, levels = c(labOrdered))   
levels(rpkmPlotsFullImport$xLab)

expOrdered <- c("Mycelium", "Oosporogenesis", "OosporeConversion") 
rpkmPlotsFullImport$Experiment <- factor(rpkmPlotsFullImport$Experiment, 
                                         levels = c(expOrdered))
levels(rpkmPlotsFullImport$Experiment)
rm(expOrdered)
rm(labOrdered)

# note, there is a bug in the way RNA replicate was carried forward (probably because levels were used instead of actual numbers), this is why Exp_Unit is being used to rebuild it here.
rpkmPlotsFullImport$RNA_Rep <- sub("^T","", rpkmPlotsFullImport$ExpUnit)
rpkmPlotsFullImport$RNA_Rep <- sub("^[0-9]+-","", rpkmPlotsFullImport$RNA_Rep)
rpkmPlotsFullImport$RNA_Rep <- sub("^c|^x","", rpkmPlotsFullImport$RNA_Rep)
rpkmPlotsFullImport$RNA_Rep <- sub("\\.","", rpkmPlotsFullImport$RNA_Rep)
rpkmPlotsFullImport$RNA_Rep <- sub("^[A-z]*.[A-z]$","1", rpkmPlotsFullImport$RNA_Rep)
rpkmPlotsFullImport$RNA_Rep <- sub("emp35C","1", rpkmPlotsFullImport$RNA_Rep)

levels(as.factor(rpkmPlotsFullImport$RNA_Rep))
```


Add a column in the large data file about being CAZy (secreted or not) on non CAZy
```{r}

mainFolder <- "GGPLOTS_FROM_FULL_DATA_continuous_X/from_Emily_24_March_2017"


listLoci <- read.table(paste(sharedPath, mainFolder, "/CAZy_of_interest_list.csv", sep = ""), 
                                  sep = ",",header = TRUE, as.is = TRUE,  stringsAsFactors = FALSE)

# Note that I am not using the fread command from data.table.  This one creates a real problem later as column indexing no longer works with data,
# table that has the feature of usinf<g column names instead.  A real problem if you want to use columns in a loop.
# library(data.table)
# listLoci <- fread(paste(sharedPath, mainFolder, "/CAZy_of_interest_list.csv", sep = ""),
#                   sep = "auto", header = TRUE)
```


Show duplicated secreted genes
```{r}
Secreted_duplicated <-  listLoci[which(listLoci$Secreted=="secreted"),]
Secreted_duplicated[duplicated(Secreted_duplicated$SequenceID),]
```





Aggregate together CAZy families for genes with more than one family, and remove duplicated
```{r}
dat_agg <- aggregate(listLoci[,c(2,3)], by=list(listLoci$SequenceID),FUN=c)
for(i in 1:length(dat_agg$CAZyFamily)){
  dat_agg$CAZyFamily2[i] <-paste(unlist(dat_agg$CAZyFamily[i]), collapse="-")
}

dat_agg$SequenceID <- NULL
colnames(dat_agg)[1] <- "SequenceID"
dat_agg$CAZyFamily <- NULL
colnames(dat_agg)[2] <- "CAZyFamily"

#remove duplicated loci
listLoci <- listLoci[!duplicated(listLoci$SequenceID),]
listLoci$CAZyFamily <- NULL
#replace names with concatenated names
listLoci <- merge(listLoci, dat_agg, by= "SequenceID")
```


Integrate gene loci CAZy and secreted data into main data table and write non-CAZy for NA
```{r}
rpkmPlotsFullImport <- merge(listLoci[,c(1,3,6)], rpkmPlotsFullImport[,2:10], by = "SequenceID", all.y = TRUE)
rpkmPlotsFullImport$Secreted[is.na(rpkmPlotsFullImport$Secreted)] <- "non-CAZy"
rpkmPlotsFullImport$CAZyFamily[is.na(rpkmPlotsFullImport$CAZyFamily)] <- "non-CAZy"
```


Create Main plot folder
```{r}
mainFolder <- "GGPLOTS_STATS_FROM_FULL_DATA_pairwise_comparisons"

if (dir.exists(paste(sharedPath, mainFolder , sep = ""))) {} else {
  dir.create(paste(sharedPath, mainFolder , sep = ""), showWarnings = TRUE, recursive = FALSE)  }
```

Create a folder for the plot outputs:
```{r}
analysisDir <- "29_Aug_2017_annotated_script"
if (dir.exists(paste(sharedPath, mainFolder , "/", analysisDir, sep = ""))) {} else {
  dir.create(paste(sharedPath, mainFolder , "/", analysisDir, sep = ""), showWarnings = TRUE, recursive = FALSE)  }

analysisPath <- paste(sharedPath, mainFolder , "/", analysisDir, sep = "")
```

Creating log10 + 1 read
```{r}
library(doBy)
rpkmPlotsFullImport$log_Read_Counts <- log10(rpkmPlotsFullImport$Read_Counts + 1)
head(rpkmPlotsFullImport)
```

Create a Table with mean by experimental unit and remove time 0 (except for mycelium=Various).
```{r}
mean_per_exp <- summaryBy(log_Read_Counts ~ SequenceID + CAZyFamily + Secreted + Experiment + Condition + TimePoint, data=rpkmPlotsFullImport, FUN = c(mean) )

mean_per_exp_no_time0  <- mean_per_exp[which(mean_per_exp$TimePoint != 0 | mean_per_exp$Condition =="Various"),]

mean_per_exp_no_time0[1:20,]
```


Find the maximum for each experiment, keeping cholesterol and control separate.
```{r}
max_per_exp <- summaryBy(log_Read_Counts.mean ~ SequenceID + CAZyFamily + Secreted + Experiment + Condition, data=mean_per_exp_no_time0, FUN = c(max) )
head(max_per_exp)
```

Reorganize this table to have different columns for experiment, so one line per gene
```{r}
library(reshape2)
counts_for_pairwise_no_T0 <- dcast(data = max_per_exp, SequenceID + CAZyFamily + Secreted  ~ Condition, value.var = "log_Read_Counts.mean.max")
head(counts_for_pairwise_no_T0)

```


Subset only secreted CAZy.  Carried all data up to here in case one wanted to do more plots with other genes.
```{r}
subset_count <-  counts_for_pairwise_no_T0[which(counts_for_pairwise_no_T0$Secreted=="secreted"),]

```




```{r}
library(ggplot2)
#Condition.vars <- c("Cholesterol", "Control", "OoCTimeCourse", "Various")


p0 <-   ggplot(aes(y= OoCTimeCourse, x= Cholesterol, colour=CAZyFamily), data = subset_count) +
   ylab("Oospore Conversion for time with maximum count (log RPKM +1)") +
     xlab("Oosporogenesis for cholesterol treatment with maximum count (log RPKM +1)") +
    geom_point() +
  geom_abline(intercept = 0, slope = 1) +
  theme_bw() +
  theme(axis.text.x = element_text(colour = 'black', size = 9, angle = 0, hjust = 0.5, vjust = 1.0, face= "plain")) +
  theme(axis.title.y = element_text(colour = 'black', size = 12, angle = 90, hjust = 0.5, vjust = 1.0, face= "plain")) +
    theme(axis.title.x = element_text(colour = 'black', size = 12, angle = 0, hjust = 0.5, vjust = 1.0, face= "plain")) 
 print(p0)
```


Vector plot
```{r}
  svg(file   = paste(sharedPath, "/", mainFolder, "/", analysisDir, "/",  "Secreted_CZAy_correlations.svg", sep=""), 
      width  = 10, 
      height = 8, 
      bg     = "white")
print(p0)
dev.off()
```

tiff plot
```{r}
  png(file   = paste(sharedPath, "/", mainFolder, "/", analysisDir, "/",  "Secreted_CZAy_correlations.png", sep=""), 
      width  = 10, 
      height = 8, 
      units  = "in", 
      res    = 600, 
      bg     = "white")
print(p0)
dev.off()
```
pdf plot
```{r}
pdf(file = paste(sharedPath, "/", mainFolder, "/", analysisDir, "/",  "Secreted_CZAy_correlations.pdf", sep="") , width = 10, height = 8)
print(p0)
dev.off()
```

Correlation plot by family
```{r}
p1 <-   ggplot(aes(y= OoCTimeCourse, x= Various, colour=Secreted), data = subset_count) +
    geom_point() +
  geom_abline(intercept = 0, slope = 1) +
    facet_wrap(~CAZyFamily, ncol=4, scales = "free") +
  theme_bw() +
  theme(axis.text.x = element_text(colour = 'black', size = 9, angle = 0, hjust = 1, vjust = 1.0, face= "italic")) +
  theme(strip.text.x = element_text(size = 8)) 
print(p1)
```


```{r}
pdf(file = paste(sharedPath, "/", mainFolder, "/", analysisDir, "/", "facet_correlations.pdf", sep="") , width = 8, height = 40)
print(p1)
dev.off()

write.table(subset_count,  paste(sharedPath, "/", mainFolder, "/", analysisDir, "/", "Table_for_correlations.csv", sep=""), append = FALSE, quote = TRUE, sep = ",", 
        col.names = NA, row.names = TRUE)
```


With the four different variables
```{r}
library("GGally")
g1 <- ggpairs(data=subset_count, columns= 4:7, mapping=ggplot2::aes(colour = Secreted), lower=list(combo=wrap("facethist",binwidth=1,  alpha = 0.3)))
print(g1)
```

Print to pdf
```{r}
pdf(file = paste(sharedPath, "/", mainFolder, "/", analysisDir, "/", "Matrix_plot_correlations.pdf", sep="") , width = 8, height = 8)
print(g1)
dev.off()

```












