Title
========================================================
This will help us when finding our files to source functions:
```{r}
install.packages("rprojroot")
install.packages("data.table")
install.packages("seqinr")
install.packages("Biostrings")
```

```{r}
library(rprojroot)
# We specify ours is an RStudio project
# The root object contains a function that will help us locate our package r files
# regarless of our current working directory
root <- rprojroot::is_rstudio_project
scriptsPath <- root$make_fix_file(".")("R")
scripts <- dir(root$find_file("R", path = root$find_file()))
scriptsl <- paste(scriptsPath, scripts, sep = "//")
lapply(scriptsl, source)
```

Set the paths for output plots, tables, figures, and graphs:
```{r}
figPath <- root$make_fix_file(".")("Figures")
dataPath <- root$make_fix_file(".")("data")
```
Ignore this path Andre, it's for where I keep the proteomes in my GitHub file structure, and I can't push this to the reporitory because the files are too big. Instead, the path is given to our shared drive in the metadata_file that gets read inthe chunk after this one.
```{r}
ref_transcriptomes_folder <- paste(dataPath, "/referenceProteomes/", sep="")
```

This table has all the paths to the files in our shared drive.
```{r}
library(Biostrings)

metadata_file <- read.table(paste(dataPath, "/referencesMetadataFinal.csv", sep=""),
                         header=TRUE, 
                         sep = ",",
                         skip = 0, 
                         stringsAsFactors = FALSE 
)
```

In this chunk I add some paths, the dbSppPath is redundant, but keep it bause I call those columns later on - I'll need to fix that later on.
```{r}
metadata_file <- metadata_file[order(metadata_file$Species),]
metadata_file$Species
metadata_file$dbSppPath <- paste("/isilon/biodiversity/users/shared/Pythium_ultimum_RNAseq/CAZy_prediction_test/output/",
                                 metadata_file$Species, "/",
                                 metadata_file$Species, ".CAZy.out.dm", sep = "")
metadata_file$ProtPath <- paste(ref_transcriptomes_folder, metadata_file$ProteinsFasta, ".gz", sep = "")
```

For some reason, none of the pythium spp except pyuu have cazy sequences that pass signalP4.1.
Get fasta file for CAZy of the pythium species, to see where problem with sequences
may be.
In the following I was trying to see why I was losing all the other pythiums when running the secretome pipeline. I solved this issue, so you can skip this chunk and go the one after it.
```{r}
library(reshape2)
library(data.table)
sppAbbr <- c("Ath", "Ddi", "Ehux", "Ha", "Phatr", "Phin", "Phni", "Phpa", "Phra", "Phso", "pyap", "pyar", 
             "pyir", "pyiw", "pyus", "pyuu", "pyve", "Sce", "Thaps")
metadata_file$SppAbbr <- sppAbbr
pySppList <- c("Pythium")
metadataPyspp <- metadata_file[metadata_file$Group %in% pySppList, ]

dbCanPySppList <- metadataPyspp$ParsedmtblPath
i <- 1

test <- lapply(dbCanPySppList, fread, sep = "auto", header = FALSE)
namesCols <- c("family.hmm", "hmm.length", "query.id", "query.length", "e.val",
               "hmm.start", "hmm.end", "query.start", "query.end", "coverage")
for(i in 1:length(test)){
  setnames(test[[i]], 1:10, namesCols)
  }
for(i in 1:length(test)){
test[[i]]$Species <- basename(dbCanPySppList[i])
test[[i]]$Species <- sub(".out.dm.ps", "", test[[i]]$Species)
print(nrow(test[[i]]))
}
summary(test)
View(test[[7]])
lapply(test, function(i) setkey(i))
testAll <- Reduce(function(...) merge(..., all = T), test)
testAll$family.hmm <- sub(".hmm", "", testAll$family.hmm)

write.table(testAll, 
           file = file.path(paste(dataPath, "/dbCAN_Subset_PythiumRefs.csv", sep = "")),
           append    = FALSE, 
           sep       = ",",
           quote     = FALSE,
           row.names = TRUE,
           col.names = NA)
```


In this chunk I am making a big dbCAN table that has the cazy.hmm families for all the reference proteomes I passed through the cazy prediction pipeline.
This is different from the first dbCAN table (which had only select CAZy families). For this big one, I go to the raw output for each species, and then merge it into a single big table.
Then I add some extra columns that may help me parse data based on species, group and class.
```{r}
# To make also get a table for all species, not just pythiums:
allParsedTablesList <- metadata_file$ParsedmtblPath
test2 <- lapply(allParsedTablesList, fread, sep = "auto", header = FALSE)
namesCols <- c("family.hmm", "hmm.length", "query.id", "query.length", "e.val",
               "hmm.start", "hmm.end", "query.start", "query.end", "coverage")
for(i in 1:length(test2)){setnames(test2[[i]], 1:10, namesCols)}
for(i in 1:length(test2)){
  test2[[i]]$Species <- basename(allParsedTablesList[i])
  test2[[i]]$Species <- sub(".out.dm.ps", "", test2[[i]]$Species)
  print(nrow(test2[[i]]))}
lapply(test2, function(i) setkey(i))
testAll2 <- Reduce(function(...) merge(..., all = T), test2)
testAll2$family.hmm <- sub(".hmm", "", testAll2$family.hmm)
testAll2$Group <- testAll2$Species
setkeyv(testAll2, "Group")
key(testAll2)
testAll2[.("Arabidopsis_thaliana"), Group := "Plant"]
setkeyv(testAll2, "Group")
testAll2[.("Dictyostelium_discoideum"), Group := "Mycetozoa"]
setkeyv(testAll2, "Group")
testAll2[like(Group, "Pyth"), Group := "Pythium"]
setkeyv(testAll2, "Group")
testAll2[like(Group, "Phy"), Group := "Phytophthora"]
setkeyv(testAll2, "Group")
testAll2[like(Group, "Hya"), Group := "Phytophthora"]
setkeyv(testAll2, "Group")
testAll2[like(Group, "Emi"), Group := "Haptophyte"]
setkeyv(testAll2, "Group")
testAll2[like(Group, "Phae"), Group := "Diatom"]
setkeyv(testAll2, "Group")
testAll2[like(Group, "Tha"), Group := "Diatom"]
setkeyv(testAll2, "Group")
testAll2[like(Group, "Sac"), Group := "Fungi"]	
testAll2$NewSeqName <- paste(testAll2$family.hmm, testAll2$query.id, sep = "_")
setkeyv(testAll2, "Group")
testAll2$Class <- testAll2$Group
setkeyv(testAll2, "Class")
testAll2[like(Class, "Py"), Class := "Oomycete"]
setkeyv(testAll2, "Class")
testAll2[like(Class, "Phy"), Class := "Oomycete"]
write.table(testAll2, 
           file = file.path(paste(dataPath, "/dbCAN_Subset_AllRefs.csv", sep = "")),
           append = FALSE, sep  = ",", quote = FALSE, row.names = TRUE, col.names = NA)
```

Don't run the next chunk - you don't need it.
```{r}
dir.create(paste(getwd(), "/data/secretome", sep = ""), showWarnings = TRUE, recursive = FALSE)
for(i in 1:length(metadata_file$Species)){
  dir.create(paste(dataPath, "/secretome/", species[i], sep = ""), showWarnings = TRUE, recursive = FALSE)
}
```

Here we read in the big dbCAN table for all the species and all their predicted CAZy:
```{r}
library("dplyr")
dbCAN_tempFile_AllRefs <- fread(paste(dataPath, "/dbCAN_Subset_AllRefs.csv", sep = ""), 
                                sep="auto", header = TRUE)

# I don't need the extra coordinate daa, so I'm taking a subset of the table:
dbCAN_tempFile_AllRefs <- subset(dbCAN_tempFile_AllRefs, 
                                 select=c("family.hmm", "query.id", "Species", "Group", "NewSeqName", "Class"))
# Rename some columns:
colnames(dbCAN_tempFile_AllRefs) <- c("cazyFam", "sequence", "species", "group", "newSeqName", "class")

# Add a column with the sequence name in the same format Andre had it in ggplots script:
dbCAN_tempFile_AllRefs$NewNames <- paste(dbCAN_tempFile_AllRefs$newSeqName, 
                                         dbCAN_tempFile_AllRefs$species, sep = "_")
# get a list of paths for where the reference proteomes are:
listProtRefFasta <- metadata_file$ProtPath

# Read in all the fasta proteomes as a list:
AA_all_refs_fasta <- lapply(listProtRefFasta, readAAStringSet, format = "fasta")

# Rename all the sequences so that only the sequenceID is kept, and not the extra annotations in the name: 
for(i in 1:length(AA_all_refs_fasta)){
  names(AA_all_refs_fasta[[i]]) <- sub(" .*","",names(AA_all_refs_fasta[[i]]))
}

# Check one of the sets and see if the names are okay:
names(AA_all_refs_fasta[[15]])

# The next is a relic from Andre's where he indexed the CAZy families:
dbCAN_tempFile_AllRefs$cazyFam <- as.factor(dbCAN_tempFile_AllRefs$cazyFam)

# There are duplicates in our table from dbCAN. This is because
# there are genes that had more than one hit for the same domain (i.e., two GH5 domains on the same sequence)
# This redundancy is not needed for me in this script. What I want is for there to be one line
# for each unique CAZy-sequence pair. 
# i.e., it's okay to have both cazy1-gene1, and cazy2-gene1, but I don't want
# cazy1-gene1 and cazy1-gene1.
# So we keep the distinct rows:
dbCAN_tempFile_AllRefs <- distinct(dbCAN_tempFile_AllRefs)

# Write this to a new table:
write.table(dbCAN_tempFile_AllRefs, 
           file = file.path(paste(dataPath, "/dbCAN_Subset_AllRefs_columnSubset.csv", sep = "")),
           append = FALSE, sep  = ",", quote = FALSE, row.names = TRUE, col.names = NA)

# Now I'm going to create a big fasta file that has all the cazy sequences for all reference species:
cazyFamTemp <- unique(dbCAN_tempFile_AllRefs$cazyFam) # Get all the CAZy families 
cazyNamesAllList <- dbCAN_tempFile_AllRefs$sequence # Get all the sequence names
rm(subsetAllCAZyFasta) # I need to make sure this is empty first
i <- 1
for(i in 1:length(AA_all_refs_fasta)){
subsetAllCAZyFasta <- AAStringSet(AA_all_refs_fasta[[i]][names(AA_all_refs_fasta[[i]]) %in% cazyNamesAllList])
if (i == 1){AA_allFasta <- subsetAllCAZyFasta} else {AA_allFasta <- append(AA_allFasta, subsetAllCAZyFasta)}
}
writeXStringSet(AA_allFasta, file=paste(dataPath, "/secretome/BigCAZySet_rawSeqNames.fasta", sep = ""),
                append=FALSE, format="fasta")
# The above fasta file doesn't have the family and species included in names, only the sequence id.

# This takes up a lot of space in my global environment, since I've written it to file, I can remove it:
rm(AA_all_refs_fasta)
```


Signalp 4.1
```{r}
#make a directory for where the signalp-4.1 data gff files will be kept:
dir.create(paste(dataPath, "/secretome/SignalP4_gff_data", sep = ""), showWarnings = TRUE, recursive = FALSE)
```

```{r}
#If I need to specify what gff path: 
signalp4_gff_path <- paste(dataPath,"/secretome/SignalP4_gff_data", sep = "")

#Path to signalp-4.1 on the cluster:
signalp_cluster_path <- "/opt/bio/signalp/signalp"


#Record the signalp commands desired:
output <- "short"
org_type <- "euk"
#To get help for signalp on the cluster:
cmd <- paste(signalp_cluster_path, "-h")
system(cmd)

#To run signalp4 on the entire fasta set with all spp in it instead of sets for individual spp, try:
#new_names_fasta_all_path <- paste(dataPath, "/secretome/All_Big_CAZy_Refs.fasta", sep = "")
bigCAZyFastaRawNames <- paste(dataPath, "/secretome/BigCAZySet_rawSeqNames.fasta", sep = "")
# rm(AA_allFasta) # since I'm reading it back in, don't need this one in rglobal env
cmd <- paste(signalp_cluster_path,
             " -f ", output,
             " -t ", org_type,
             " -n ", paste(signalp4_gff_path, "/BigCAZySet_rawSeqNames_fasta_signalp4.gff ", sep = ""),
             bigCAZyFastaRawNames,
             sep = "")
cmd
system(cmd)
```

To look at the gff files generated from signalp-4.1:
```{r}
sp4_gff_list <- list.files(path = signalp4_gff_path, 
                           pattern = glob2rx("BigCAZySet_rawSeqNames_fasta_signalp4.gff"), 
                           full.names = T)
sp4_gff_list

# To see the first 10 lines of the gff file:
cmd <- paste("tail", "-1000", sp4_gff_list)
system(cmd)

# Note: the first three lines of the gff file are headers, then the sequences are in the remaining rows.
# To get the names of the sequences from the gff files:
gff_seq_names_sp4 <- read.table(sp4_gff_list,
                                header = FALSE, 
                                sep = "\t",
                                skip = 3, 
                                stringsAsFactors = FALSE 
                                ) #skip 3 is to ignore the first 3 lines (these are headers), read.table uses the xls library.
list_gff_seq_names_sp4 <- gff_seq_names_sp4[,1] #To get list of only the seq names, which are in col 1.
list_gff_seq_names_sp4

# Check to ensure not duplicates, using the whole list (length), and unique:
length(unique(list_gff_seq_names_sp4)) #If all are unique, then the number unique will be equal to the total in list_gff_seq_names_sp4
# list_gff_seq_names_sp4 <- unique(list_gff_seq_names_sp4)
```

Moving on, if all are unique:
Get fasta sequences that pass signalp-4.1 using the seq names from the gff files:
1. Triage the original protein fasta file and get a subset for only those passing signalp4 from our names list:
```{r}
# AA_fasta_Big_AAstring <- readAAStringSet(new_names_fasta_all_path, format = "fasta")
AA_fasta_Big_AAstring <- readAAStringSet(bigCAZyFastaRawNames, format = "fasta")
names(AA_fasta_Big_AAstring)
all_fasta_signalp4_subset <- AA_fasta_Big_AAstring[which(names(AA_fasta_Big_AAstring) %in% list_gff_seq_names_sp4)]
length(all_fasta_signalp4_subset)
unique(all_fasta_signalp4_subset)
```

2. Write a new fasta file for only those seqs passing signalp-4.1, with raw names:
```{r}
writeXStringSet(all_fasta_signalp4_subset, 
                file=paste(dataPath, "/secretome/SignalP4_gff_data/BigCAZySet_rawSeqNames_SignalP4.fasta", sep = ""), 
                append=FALSE, format="fasta", width = 400)

all_fasta_signalp4_subset <- readAAStringSet(paste(dataPath, "/secretome/SignalP4_gff_data/BigCAZySet_rawSeqNames_SignalP4.fasta", sep = ""), format="fasta")
unique(names(all_fasta_signalp4_subset))
```

Retrieve the names of the sequences for Pyuu that pass signalP4.1:
Note: I could do this for the whole set as well, but recall that signalP3.0 is 
on the browser and limits runs to 2000 sequences at a time.
```{r}
library(BSgenome)
subPyuu <- all_fasta_signalp4_subset[grep("PYU1_.*", names(all_fasta_signalp4_subset))]
length(subPyuu)
pyuuSignalP4Names <- names(subPyuu)
# This shows me that I have identified 157 Pyuu CAZy genes out of the total 805 in pyuuCazyGenes
all_Pyuu_signalp4_subset <- AA_fasta_Big_AAstring[which(names(AA_fasta_Big_AAstring) %in% pyuuSignalP4Names)]
writeXStringSet(all_Pyuu_signalp4_subset, 
                file=paste(dataPath, "/secretome/SignalP4_gff_data/Pyuu_rawSeqNames_SignalP4.fasta", sep = ""), 
                append=FALSE, format="fasta")

# To do this for all Pythium and Phytophthora species:
setkeyv(dbCAN_tempFile_AllRefs, "class")
length(unique(dbCAN_tempFile_AllRefs$sequence))

oomyCazyNames <- dbCAN_tempFile_AllRefs[.("Oomycete"), "sequence", with = FALSE]
oomyCazyNamesList <- as.list(oomyCazyNames)
subNames <- names(all_fasta_signalp4_subset)
oomySignalP4Names <- intersect(oomyCazyNames$sequence, subNames)
oomySignalp4AASet <- all_fasta_signalp4_subset[which(names(all_fasta_signalp4_subset) %in% oomySignalP4Names)]
writeXStringSet(oomySignalp4AASet, 
                file=paste(dataPath, "/secretome/SignalP4_gff_data/AllOomycete_rawSeqNames_SignalP4.fasta", sep = ""), 
                append=FALSE, format="fasta")
```

Signalp 3.0
Had to split the fasta file to input into SignalP3.0 in the Shell:
$  awk 'BEGIN {n_seq=0;} /^>/ {if(n_seq%2000==0){file=sprintf("AllOomycete_rawSeqNames_SignalP4_%d.fasta",n_seq);} print >> file; n_seq++; next;} { print >> file; }' < AllOomycete_rawSeqNames_SignalP4.fasta

Passed the file "all_fasta_signalp4_subset.fasta" through signalp 3.0 on the web server:
http://www.cbs.dtu.dk/services/SignalP-3.0/
For Eukaryotes, method=Both (Neural Networks, NN, and Hidden Markov Models, HMM), No graphics,
Output format = short (no graphics), Truncation = 70 residues
Selected all the data starting at the first header line. Pasted this into Word
Replaced all whitespce (^w), with a comma (,). Deleted the first row. Removed the exta #, in front of "name".
Saved the word file, then copied it to an excel sheet, and saved this as a csv (comma delimited file), "all_signalp3_data.csv"
csv col names before and after I edited to remove special charactes:
name,Cmax,pos,?,Ymax,pos,?,Smax,pos,?,Smean,?,D,?,name,!,Cmax,pos,?,Sprob,?
name,Cmax,pos,Cres,Ymax,pos,Yres,Smax,pos,Sres,Smean,Smeanres,D,Dres,name,Exclam,Cmax,pos,HMMCres,Sprob,HMMSres
```{r}
dir.create(paste(dataPath, "/secretome/SignalP3_data", sep=""), 
           showWarnings = TRUE, recursive = FALSE)
signalp3_data_path <- paste(dataPath, "/secretome/SignalP3_data/", sep = "")
signalp3_Oomy <- fread(paste(signalp3_data_path, "Oomy_fasta_signalP3_rawNames.csv", sep=""), 
                             header = TRUE, sep = "auto")
sigp3_col_names <- names(signalp3_Oomy)
names <- c("name","Cmax","pos","Cres","Ymax","pos","Yres","Smax","pos","Sres","Smean","Smeanres","D","Dres","name.1","Exclam","Cmax","pos","HMMCres","Sprob","HMMSres")
colnames(signalp3_Oomy) <- names
sig3_df <- subset(signalp3_Oomy, select=c("name", "Dres", "name.1", "Exclam"))
colnames(sig3_df) <- c("NN_Name", "NN_result", "HMM_Name", "HMM_result")
setkeyv(sig3_df, c("NN_result", "HMM_result"))
key(sig3_df)
sig3_df_subset <- sig3_df[.("Y", "S"), "HMM_Name", with = FALSE] # Don't use NN_Name bec.it's incomplete
sig3Names <- as.list(sig3_df_subset$HMM_Name)
sig4Names <- as.list(names(oomySignalp4AASet)) # names of those passing signalp 4.1 above
sig3Names <- intersect(sig3Names, sig4Names)
sigP3AASet <- AA_fasta_Big_AAstring[which(names(AA_fasta_Big_AAstring) %in% sig3Names)]
sigP3AASet_truncated <- AAStringSet(strtrim(sigP3AASet, width=400))
attributes(sigP3AASet_truncated)
writeXStringSet(sigP3AASet_truncated, 
                file=paste(dataPath, "/secretome/SignalP3_data/AllOomycete_rawSeqNames_SignalP3-truncated.fasta", 
                           sep = ""), 
                append=FALSE, format="fasta")
```

 
TargetP:
```{r}
dir.create(paste(dataPath, "/secretome/TargetP1-1", sep=""), showWarnings = TRUE, recursive = FALSE)
```

It is not possible to get targetP data for the Pythiums other than Pyuu because TargetP
truncates the names to 20 cahracters, and then I can't tell the sequences apart. So I run 
TargetP on fasta subsets for all the other species, and then do those for the other
pythiums as a separate batch and note which one do not pass.
For that I need to go to the targetP output and since the table output gives the results 
of each sequence in the same order aas the input fasta file, I match up the lines inthe table with the fasta entries in the fasta file.
```{r}
targetP_data_path <- paste(dataPath, "/secretome/TargetP1-1/", sep="")
targetP_subset_fasta_path <- paste(shared_path,"TargetP1-1/all_fasta_targetP1-1_subset.fasta", sep="")
#cd to the signalp3 directory and do the folllwoing command in the shell:
#awk 'BEGIN {n_seq=0;} /^>/ {if(n_seq%100==0){file=sprintf("signalp3_fasta_subset_%d.fa",n_seq);} print >> file; n_seq++; next;} { print >> file; }' < all_fasta_signalp3_subset.fasta

#Manually passed seqs through the web server for TargetP1-1: http://www.cbs.dtu.dk/services/TargetP/
#file with notes and sequences that did not pass is in the targetP folder. Instead of logging all that are secreted, 
#take note of those that are not (only a few are not, and so easier to find and delete them)


TargetP_data_path <- paste(dataPath, "/secretome/TargetP1-1/", sep = "")
TargetP_Oomy <- fread(paste(TargetP_data_path, "targetP_doesnotIncludePythiumsOtherThanPyuuBecOfNameTruncatedProblem.csv", sep=""), 
                             header = TRUE, sep = "auto")
setkeyv(TargetP_Oomy, "Loc")
# Find those not passing:
targetPoomySubset <- TargetP_Oomy[!"S", "Name", with = FALSE]

# Find those not passing target P for the other pythiums with long names:
failTargetP_otherPySpp <- c("maker-pag1_scaffold_19-snap-gene-0.67-mRNA-1",
                            "maker-pag1_scaffold_236-snap-gene-0.13-mRNA-1",
                            "maker-pag1_scaffold_323-snap-gene-0.15-mRNA-1",
                            "maker-pag1_scaffold_1025-snap-gene-0.8-mRNA-1",
                            "maker-pag1_scaffold_1111-snap-gene-0.4-mRNA-1",
                            "maker-pir_contig_423-snap-gene-0.16-mRNA-1",
                            "maker-pir_contig_605-snap-gene-0.8-mRNA-1",
                            "maker-pir_contig_1340-fgenesh-gene-0.1-mRNA-1",
                            "maker-pug3_contig_115-snap-gene-0.12-mRNA-1",
                            "snap_masked-pug3_contig_164-abinit-gene-0.16-mRNA-1",
                            "maker-pug3_contig_1359-snap-gene-0.5-mRNA-1",
                            "fgenesh-pve_contig_26-abinit-gene-0.169-mRNA-1",
                            "maker-pve_contig_37-fgenesh-gene-0.13-mRNA-1",
                            "maker-pve_contig_166-fgenesh-gene-0.14-mRNA-1",
                            "maker-pve_contig_345-fgenesh-gene-0.0-mRNA-1")

targetPFailNames <- as.list(targetPoomySubset$Name)
targetPFailNames2 <- union(targetPFailNames, failTargetP_otherPySpp)
sig3NamesIntermediate <- setdiff(sig3Names, targetPFailNames2) # 1458 entries, now the failPysremoved
targetPNames <- sig3NamesIntermediate
length(targetPNames)

targetPAASet <- AA_fasta_Big_AAstring[which(names(AA_fasta_Big_AAstring) %in% targetPNames)]
writeXStringSet(targetPAASet, 
                file=paste(dataPath, "/secretome/TargetP1-1/targetP_Oomycetes.fasta", 
                           sep = ""), 
                append=FALSE, format="fasta")
```

TMHMM2.0 http://www.cbs.dtu.dk/services/TMHMM/
Need to pass the fasta file for sequences passing TargetP1.1 through the web server for TMHMM2.0
Restrictions:
At most 10,000 sequences and 4,000,000 amino acids per submission; each sequence not more than 8,000 amino acids. 
Criteria to pass tmhmm:
0 or 1 tmhmm only
for those with a tmhmm, transmembrane portion must overlap 10 or more amino acids with the first 35
residues at the N-terminal, as these tmhmms are likely signal peptides.
```{r}
dir.create(paste(dataPath, "/secretome/TMHMM_data", sep=""), showWarnings = TRUE, recursive = FALSE)
tmhmm_data_path <- paste(dataPath,"/secretome/TMHMM_data/", sep="")

tmhmm_all_data <- read.table(paste(tmhmm_data_path, "tmhmm_data.csv", sep=""), 
                                header = FALSE,
                                sep = ",",
                                quote = "\"",
                                #dec = ".",
                                stringsAsFactors = FALSE)

#To see the first 10 lines of the data file:
cmd <- paste("head", "-10", paste(tmhmm_data_path,"tmhmm_data.csv", sep=""))
system(cmd)

#To add column names to the tmhmm data table:
colnames(tmhmm_all_data) <- c("name", "length", "exp_num_aa", "aa_in_first_60", "PredHel", "Topology")

tmhmm_subset_no_tmhmm <- subset(tmhmm_all_data, tmhmm_all_data[,5] =="PredHel=0", select = c(name, PredHel, Topology))
tmhmm_subset_1tmhmm <- subset(tmhmm_all_data, tmhmm_all_data[,5] == "PredHel=1")

tmhmm_subset_1tmhmm$Topology <- sub("Topology=","", tmhmm_subset_1tmhmm$Topology, ignore.case=F)
tmhmm_subset_1tmhmm$Topology <- sub("i","", tmhmm_subset_1tmhmm$Topology, ignore.case=F)
tmhmm_subset_1tmhmm$Topology <- sub("o","", tmhmm_subset_1tmhmm$Topology, ignore.case=F)
#tmhmm_subset_1tmhmm$Topology <- sub("^[0-9]+-","", tmhmm_subset_1tmhmm$Topology, ignore.case=F)
tmhmm_subset_1tmhmm$Topology <- sub("-[0-9]+","", tmhmm_subset_1tmhmm$Topology, ignore.case=F)

#Need to make the values in tmhmm_subset_1tmhmm$Topology numeric so that I can make a subset 
#according to the topology values (greater than function depends on it being numeric)
tmhmm_subset_1tmhmm$Topology <- as.numeric(tmhmm_subset_1tmhmm$Topology)
tmhmm_1hel_subset <- subset(tmhmm_subset_1tmhmm, Topology < 27, select = c(name, PredHel, Topology))
#Now make the values in Topology in tmhmm_1hel_subset characters again
tmhmm_1hel_subset$Topology <- as.character(tmhmm_1hel_subset$Topology)
#Double check it:
is.character(tmhmm_1hel_subset$Topology)


#Combine tmhmm_subset_no_tmhmm and tmhmm_1hel_subset
tmhmm_subset <- rbind(tmhmm_subset_no_tmhmm, tmhmm_1hel_subset)
rownames(tmhmm_subset)
#Order the tmhmm data by row.names column
tmhmm_subset <- tmhmm_subset[ order(as.numeric(row.names(tmhmm_subset))),]
#Give more meaningfull name to topology for sequences with no tmhmm:
tmhmm_subset$Topology <- sub("Topology=o", "no_tmhmm", tmhmm_subset$Topology, ignore.case=F)
#Save the data table:
write.table(tmhmm_subset, 
            file = paste(tmhmm_data_path,"tmhmm_subset_table.csv", sep=""), 
            append = FALSE, 
            sep = ",", 
            col.names=NA)


#make a fasta file using the names of the sequences in tmhmm_subset:
tmhmm_subset_names <- tmhmm_subset[,c("name")]
#Check to ensure not duplicates, using the whole list (length), and unique:
length(unique(tmhmm_subset_names))

tmhmmNames <- intersect(targetPNames, tmhmm_subset_names)
#1. Triage the original protein fasta file and get a subset for only those passing tmhmm from our names list:
all_fasta_tmhmm_subset <- AA_fasta_Big_AAstring[which(names(AA_fasta_Big_AAstring) %in% tmhmmNames)]


#2. Write a new fasta file for only those seqs passing tmhmm:
writeXStringSet(all_fasta_tmhmm_subset, 
                file=paste(tmhmm_data_path, "tmhmm_fasta_subset.fasta", sep=""),
                append=FALSE, format="fasta")
tmhmm_fasta_subset <- readAAStringSet(paste(tmhmm_data_path, "tmhmm_fasta_subset.fasta", sep=""), 
                                      format="fasta")
```

ScanProSite:
```{r}
dir.create(paste(dataPath, "/secretome/ER_ScanP_Pass", sep=""), showWarnings = TRUE, recursive = FALSE)
scanP_data_path <- paste(dataPath,"/secretome/ER_ScanP_Pass/", sep="")
#Moving on, if all are unique:
###################################################################################################
#ScanProsite tool http://prosite.expasy.org/scanprosite/
#To pass this scan, need to not have a ER_TARGET: Endoplasmic reticulum targeting sequence(PS00014)
# Output format: Matchlist
# 
# Hits for PS00014 (ER_TARGET) [PROSITE (release 20.124)] motif on all diUvGg3yzu custom database sequences :
#   
#   found: 9 hits in 9 sequences
# 
# Graphical view (graphical view with feature detection)
# shaded alignment of hits
# 
# Hits for PS00014|ER_TARGET (pattern) Endoplasmic reticulum targeting sequence :
#   Pattern: [KRHQSA]-[DENQ]-E-L>
#   Can't estimate number of random matches in ~ 100'000 sequences (50'000'000 residues): constrained Motifs are not handled

#Results: The following have ER retention signals and need to be removed from our final list:
erRetSig <- c("EGZ27990", "ETI40213", "KUF87062", "PITG_03200T0", "PYU1_T001074", 
              "maker-pir_contig_70-fgenesh-gene-0.2-mRNA-1", 
              "maker-piw_contig_156-fgenesh-gene-0.4-mRNA-1",
              "maker-piw_contig_4-snap-gene-0.51-mRNA-1",
              "maker-pve_contig_12-fgenesh-gene-0.21-mRNA-1",
              "maker-pve_contig_186-snap-gene-0.15-mRNA-1")
erPassNames <- setdiff(tmhmmNames, erRetSig)

#1. Triage the original protein fasta file:
all_ScanPrositeER_pass <- AA_fasta_Big_AAstring[which(names(AA_fasta_Big_AAstring) %in% erPassNames)]


#2. Write a new fasta file for only those seqs passing scanProsite for no ER retention signals:
writeXStringSet(all_ScanPrositeER_pass, 
                file=paste(scanP_data_path, "scanP_ER_subset.fasta", sep=""),
                append=FALSE, format="fasta")
scanP_fasta_subset <- readAAStringSet(paste(scanP_data_path, "scanP_ER_subset.fasta", sep=""), 
                                      format="fasta")
#3. Write the list of sequence names as a text file for only those seqs passing scanProsite for no ER retention signals:
write.table(erPassNames, 
            file = paste(dataPath, "/secretome/ER_ScanP_Pass/seqsPassingER_secreted_all_oomy_refs.txt", sep = ""), 
            append = FALSE, quote = FALSE, row.names = FALSE, col.names = FALSE)

#4. Write one that has only the final secreted Pyuu sequences as well:
secPyuu <- scanP_fasta_subset[grep("PYU1_.*", names(scanP_fasta_subset))]
writeXStringSet(secPyuu, 
                file=paste(scanP_data_path, "scanP_ER_subset_Pyuu_Only.fasta", sep=""),
                append=FALSE, format="fasta")
writeXStringSet(secPyuu, 
                file=paste(dataPath, "/secretome/final_secreted_cazy_Pyuu.fasta", sep=""),
                append=FALSE, format="fasta")

```

> length(names(AA_all_refs_fasta))
[1] 0
> length(names(AA_allFasta))
[1] 10846
> length(names(bigCAZyFastaRawNames))
[1] 0
> length(names(all_fasta_signalp4_subset))
[1] 3945
From this point on, oomycete set only:
> length(names(oomySignalp4AASet))
[1] 2487
> length(names(sigP3AASet))
[1] 2473
> length(names(targetPAASet))
[1] 2439
> length(names(all_fasta_tmhmm_subset))
[1] 2080
> length(names(all_ScanPrositeER_pass))
[1] 2070

I found 10 additional secreted CAZy in Pyuu after rerunning the script.
The new CAZy secreted that I found that were not included when looking only at Marcello's list
are: 
"PYU1_T001944" "PYU1_T001945" "PYU1_T001949" "PYU1_T002783" "PYU1_T003536" "PYU1_T003553" "PYU1_T005497"
"PYU1_T005498" "PYU1_T010744" "PYU1_T013244"

Out of these, 126 Pyuu pass this pipeline.
```{r}
# Retrieve as a list all the PYU1_ matches - these are secreted Pyuu seqs
secPyuu <- scanP_fasta_subset[grep("PYU1_.*", names(scanP_fasta_subset))]
length(secPyuu)
secPyuuNames <- names(secPyuu)
length(unique(secPyuuNames)) # 126 secreted CAZy that I found newly.


# How does the secreted list compare to the secreted CAZy I had for Pyuu before?
oldSecCAZyAA <- readAAStringSet(paste(sharedPath, "References/Secretome/CAZy_secreted.fasta", sep = ""),
                                format = "fasta")  

oldSecPyuuAA <- oldSecCAZyAA[grep("Pyuu_.*", names(oldSecCAZyAA))]
oldPyuuCAZyNames <- data.table(names(oldSecPyuuAA))
newCols <- c("spp", "cazyFam", "Prefix", "seqNum", "secreted")
oldPyuuNames <- oldPyuuCAZyNames[, c(newCols) := tstrsplit(V1, "_", fixed = TRUE)]
oldPyuuNames$sequence <- paste(oldPyuuNames$Prefix, oldPyuuNames$seqNum, sep = "_")
oldPyuuNamesList <- as.list(oldPyuuNames$sequence)
length(oldPyuuNamesList)
length(unique(oldPyuuNamesList)) # 120 unique secreted genes from Marcello's Pyuu CAZy
oldPyuuNamesList <- unique(oldPyuuNamesList)
length(oldPyuuNamesList)

newSeqFound <- setdiff(secPyuuNames, oldPyuuNamesList)
newSeqFound <- sort(newSeqFound)
newSeqFound

# How does my secreted CAZy list compare to those predicted by Marcello?
marcelloCAZyPyuuannotations <- fread(paste(sharedPath, "References/Final_Table_CAZyme_050813.csv", 
                                 sep = ""), sep = "auto", header = FALSE, skip = 1)
namesColMar <- as.character(as.list(marcelloCAZyPyuuannotations[1,]))
colnames(marcelloCAZyPyuuannotations)[1:22] <- namesColMar
marcelloCAZyPyuuannotations <- marcelloCAZyPyuuannotations[-1,]
setkeyv(marcelloCAZyPyuuannotations, "Signal.P")
marcelloSecreted <- marcelloCAZyPyuuannotations[.("SignalP")]

length(unique(marcelloSecreted$SequenceID)) # Result is 212

# What are the genes I pulled out with the new dbCAN run?
setkeyv(dbCAN_tempFile_AllRefs2, "species")
dbCANPyuuseqs <- dbCAN_tempFile_AllRefs2[.("Pythium_ultimum_var_ultimum")]
length(dbCANPyuuseqs$sequence)
dbCANPyuuseqsNames <- unique(dbCANPyuuseqs$sequence)
length(dbCANPyuuseqsNames) # there are 411 unique sequences captured
# What are the genes Marcello had for Pyuu CAZy that are not in this new dbCAN list?
marcelloCAZyPyuuNames <- as.list(unique(marcelloCAZyPyuuannotations$SequenceID))
marcelloAnti <- setdiff(marcelloCAZyPyuuNames, dbCANPyuuseqsNames)
length(marcelloAnti) # result is 14 genes
```
The following genes in Marcello's CAZy are not in the new CAZy dbCAN run:
"PYU1_T001230"
"PYU1_T001534"
"PYU1_T006496"
"PYU1_T009260"
"PYU1_T010847"
"PYU1_T013043"
"PYU1_T001521"
"PYU1_T008198"
"PYU1_T008359"
"PYU1_T009242"
"PYU1_T010177"
"PYU1_T011610"
"PYU1_T012899"
"PYU1_T014041"

Continuing on from the previous chunk.
```{r}
# Read in the latest dbCAN table that has the extra columns, in case it's no 
# longer in the environment:
dbCAN_tempFile_AllRefs2 <- fread(paste(dataPath, 
                                      "/dbCAN_Subset_AllRefs_columnSubset.csv", 
                                      sep = ""), 
                                sep="auto", header = TRUE)
colnames(dbCAN_tempFile_AllRefs2)
dbCAN_tempFile_AllRefs2 <- dbCAN_tempFile_AllRefs2[,2:8] # To remove extra "V1" column
setkeyv(dbCAN_tempFile_AllRefs2, "sequence")
key(dbCAN_tempFile_AllRefs2)
dbCANPyuuSec <- dbCAN_tempFile_AllRefs2[.(c(secPyuuNames))]


# Write this table for our records:
# Note, this table doesn't have the start and stop positions of the raw dbCAN file.
# If that is what you want, read in the following table and repeat the previous lines,
# dbCAN_CAZySubset_AllRefs.csv - you'll have to add the extra name columns again.
write.table(dbCANPyuuSec, 
           file = file.path(paste(dataPath, "/dbCAN_SecretedPyuu_Final_subsetColumns.csv", sep = "")),
           append = FALSE, sep  = ",", quote = FALSE, row.names = TRUE, col.names = NA)

# Now use this table in the ggplots script to get these plots or continue below:
cazyPyuuAll2 <- dbCANPyuuSec[,1:2]


T_G_conversion_table <- fread(paste(sharedPath,
                                    "References/T_G_name_conversion_from_gff_file.csv", 
                                    sep = ""),
                              sep = "auto",
                              stringsAsFactors = FALSE)

length(unique(T_G_conversion_table$T_code))
length(unique(T_G_conversion_table$G_code))

T_G_conversion_table$G_code[duplicated(T_G_conversion_table$G_code)]
T_G_conversion_table$T_codes[duplicated(T_G_conversion_table$T_code)]

colnames(T_G_conversion_table)[colnames(T_G_conversion_table)=="G_code"] <- "Locus"

# Can't read the next table becuase it requires openxlsx, not installing in my RStudio
# for some reason. So I went to the document and saved the sheat I want (S9) as a 
# csv, and use data.table fread to get it.:
LEV_2010_annotations <- fread(paste(sharedPath, "References/Genome_Biology_paper_AdditionalDataFile3.csv", 
                                 sep = ""), sep = "auto", header = FALSE, skip = 1)
namesCol <- as.character(as.list(LEV_2010_annotations[1,]))
colnames(LEV_2010_annotations)[1:10] <- namesCol
colnames(LEV_2010_annotations)[2] <- 	"Functional.Annotation" # This is because there is a space in the name to remove
LEV_2010_annotations <- LEV_2010_annotations[-1,]

# Now that our data.table looks the same as per Andre's code, moe on:
LEV_2010_df <- merge(T_G_conversion_table, LEV_2010_annotations, by = "Locus", all.x = TRUE)

colnames(cazyPyuuAll2)[2] <- "T_code"

CAZy_of_interest <- merge(cazyPyuuAll2, LEV_2010_df[,2:3], by = "T_code", all.x = TRUE)

CAZy_of_interest$geneInfo <- paste(CAZy_of_interest$T_code, CAZy_of_interest$Functional.Annotation, sep = " - ")
CAZy_of_interest$geneInfoTab <- paste(CAZy_of_interest$T_code, CAZy_of_interest$Functional.Annotation, sep = " -\n ")

colnames(listLoci)
listLoci2 <- data.table(CAZy_of_interest$T_code,CAZy_of_interest$cazyFam,CAZy_of_interest$geneInfo,
                        CAZy_of_interest$geneInfoTab) 
colnames(listLoci2) <- c("SequenceID", "CAZyFamily", "geneInfo", "geneInfoTab")
listName2 <- "all_secretedPyuuCAZy"
write.table(listLoci2,
            file = file.path(paste(dataPath, "/secretome/", listName2, ".csv", sep = "")),
            append = FALSE, sep  = ",", quote = FALSE, row.names = TRUE, col.names = NA)

```

Creates a data subset for plotting and exports this subset.  
This can help us when we want to diagnosticate outliers.
```{r}
temp <- merge(listLoci2, rpkmPlotsFullImport, by = "SequenceID", all.x = TRUE)
temp$Experiment <- sub("OosporeConversion", "Oospore Conversion", temp$Experiment)

# To order rows
temp <- temp[order(temp$geneInfo, temp$xLab),] 

# To select and order columns
selectCol <- c("SequenceID", "CAZyFamily", "geneInfo", "geneInfoTab", "ExpUnit", "Condition", 
               "TimePoint", "RNA_Replicate", "Experiment", "xLab", "Read_Counts") 

temp <- subset(temp, select = selectCol)

write.table(temp, paste(dataPath, "/secretome/", listName2,"_all.csv", sep = ""),
            append = FALSE, sep = ",", col.names = NA)
```


Single file, multiple Bar plots for gene in listLoci
```{r}
library("ggplot2")

svg(file   = paste(dataPath, "/secretome/", listName2,  ".svg", sep = ""),
    width  = 8, 
    height = 7*ceiling(nrow(temp)/200))
    # 200 in seems to be the maximum length for a pdf file, so I used sgv
    #height = 200)

bp <- ggplot(aes(y = Read_Counts, x = xLab), data  = temp) +
  theme_bw() +
  theme(axis.text.y = element_text(size  = 6, 
                                   hjust = 1, 
                                   vjust = 0.4)) +
  theme(axis.text.x = element_text(colour = 'black', 
                                   size   = 6, 
                                   angle  = 45, 
                                   hjust  = 1, 
                                   vjust  = 1)) +
  theme(axis.ticks = element_line(colour = 'black', 
                                  size   = 0.5)) +
  labs(y = "Total Number of Reads\n (rpkm-normalized)") + 
  labs(x = "Treatment") +
  theme(axis.title.y = element_text(colour = 'black', 
                                    size   = 10, 
                                    angle  = 90, 
                                    hjust  = 0.5, 
                                    vjust  = 0.2, 
                                    face   = 'bold')) +
  theme(axis.title.x = element_text(colour = 'black', 
                                    size   = 10, 
                                    angle  = 0, 
                                    hjust  = 0.5, 
                                    vjust  =  -0.2, 
                                    face   = 'bold')) +
  geom_boxplot(position = position_dodge(0.8), 
               width    = 0.8, 
               outlier.size   = 1, 
               outlier.colour = "black", 
               outlier.shape  = 20) +
  stat_summary(fun.y = mean, 
               geom  = "point", 
               shape = 5, 
               size  = 1, 
               position = position_dodge(0.8))  +
#   facet_wrap(~SequenceID, 
#   facet_grid(SequenceID~Experiment,
  facet_grid(geneInfoTab~Experiment,
             scales = "free", space = "free_x") +
  theme(strip.text.x = element_text(size   = 6, 
                                    face   = 'bold.italic', 
                                    colour = "black", 
                                    angle  = 0))
print(bp)
dev.off()

```

*** Over here Andre - just ran this after you left, 19March2017:
Bar plots per CAZy family group:
```{r}
library("ggplot2")
cazyFam <- unique(listLoci2$CAZyFamily)
dirPlotName <- "ggplotsByCAZyFam"

dir.create(paste(dataPath, "/secretome/", dirPlotName, sep = ""), 
           showWarnings = TRUE, recursive = FALSE)

ggplotsByCazyFamPath <- paste(dataPath, "/secretome/", dirPlotName, "/", sep = "")

i <- 1
for(i in 1:length(cazyFam)){
  temp3 <- subset(temp, temp$CAZyFamily == cazyFam[i])
  temp3$Experiment <- sub("Mycelium", "Mycel.", temp3$Experiment )

pdf(file   = paste(ggplotsByCazyFamPath, cazyFam[i], ".pdf", sep = ""),
    width  = 8, 
    height = 7*ceiling(nrow(temp3)/90))

bp <- ggplot(aes(y = Read_Counts, x = xLab), data  = temp3) +
theme_bw() +
  theme(axis.text.y = element_text(size  = 8, 
                                   hjust = 1, 
                                   vjust = 0.4)) +
  theme(axis.text.x = element_text(colour = 'black', 
                                   size   = 8, 
                                   angle  = 45, 
                                   hjust  = 1, 
                                   vjust  = 1)) +
  theme(axis.ticks = element_line(colour = 'black', 
                                  size   = 0.5)) +
  labs(y = "Total Number of Reads\n(rpkm-normalized)") +
  labs(x = "Treatment") +
  theme(axis.title.y = element_text(colour = 'black', 
                                    size   = 8, 
                                    angle  = 90, 
                                    hjust  = 0.5, 
                                    vjust  = 1.5, 
                                    face   = 'bold')) +
  theme(axis.title.x = element_text(colour = 'black', 
                                    size   = 10, 
                                    angle  = 0, 
                                    hjust  = 0.5, 
                                    vjust  =  0, 
                                    face   = 'bold')) +
  geom_boxplot(position = position_dodge(0.8), 
               width    = 0.8, 
               outlier.size   = 1, 
               outlier.colour = "black", 
               outlier.shape  = 20) +
  stat_summary(fun.y = mean, 
               geom  = "point", 
               shape = 5, 
               size  = 0.5, 
               position = position_dodge(0.8))  +
#   facet_wrap(~SequenceID, 
#   facet_grid(SequenceID~Experiment,
  facet_grid(geneInfoTab~Experiment,
             scales = "free", space= "free_x") +
  theme(strip.text.x = element_text(size   = 9, 
                                    face   = 'bold', 
                                    colour = "black", 
                                    angle  = 0)) +
  theme(strip.text.y = element_text(size   = 7, 
                                    face   = 'bold', 
                                    colour = "black", 
                                    angle  = -90))
print(bp)
dev.off()
}

```
Chunk to generate bar plots for individual genes and output them
to the specified folder for "dataPath"
```{r}
library("ggplot2")

dir.create(paste(dataPath, "/secretome/Individual_plots", sep = ""), 
           showWarnings = TRUE, recursive = FALSE)
i<- 1
for(i in 1:length(listLoci2$geneInfoTab)){
  temp2 <- subset(temp, temp$SequenceID == listLoci2$SequenceID[i])
  temp2$Experiment <- sub("Mycelium", "Mycel.", temp2$Experiment )
  png(file   = paste(dataPath, "/secretome/Individual_plots/", 
                     listLoci2$SequenceID[i], ".png", sep = ""), 
      width  = 7, 
      height = 3, 
      units  = "in", 
      res    = 300, 
      bg     = "white")
  bp <- ggplot(aes(y = Read_Counts, x = xLab), data  = temp2) +
  theme_bw() +
  theme(axis.text.y = element_text(size  = 8, 
                                   hjust = 1, 
                                   vjust = 0.4)) +
  theme(axis.text.x = element_text(colour = 'black', 
                                   size   = 8, 
                                   angle  = 45, 
                                   hjust  = 1, 
                                   vjust  = 1)) +
  theme(axis.ticks = element_line(colour = 'black', 
                                  size   = 0.5)) +
  labs(y = "Total Number of Reads\n(rpkm-normalized)") +
  labs(x = "Treatment") +
  theme(axis.title.y = element_text(colour = 'black', 
                                    size   = 8, 
                                    angle  = 90, 
                                    hjust  = 0.5, 
                                    vjust  = 1.5, 
                                    face   = 'bold')) +
  theme(axis.title.x = element_text(colour = 'black', 
                                    size   = 10, 
                                    angle  = 0, 
                                    hjust  = 0.5, 
                                    vjust  =  0, 
                                    face   = 'bold')) +
  geom_boxplot(position = position_dodge(0.8), 
               width    = 0.8, 
               outlier.size   = 1, 
               outlier.colour = "black", 
               outlier.shape  = 20) +
  stat_summary(fun.y = mean, 
               geom  = "point", 
               shape = 5, 
               size  = 0.5, 
               position = position_dodge(0.8))  +
#   facet_wrap(~SequenceID, 
#   facet_grid(SequenceID~Experiment,
  facet_grid(geneInfoTab~Experiment,
             scales = "free", space= "free_x") +
  theme(strip.text.x = element_text(size   = 9, 
                                    face   = 'bold', 
                                    colour = "black", 
                                    angle  = 0)) +
  theme(strip.text.y = element_text(size   = 7, 
                                    face   = 'bold', 
                                    colour = "black", 
                                    angle  = -90))
print(bp)
dev.off()
}
```


Let's make a table of our secreted Pyuu sequence names, with some annotations
```{r}
#cmd <- paste("pfscan ",)

sapply(marcelloCAZyPyuuannotations,class)          

marcelloCAZy <- unique(marcelloCAZyPyuuannotations$no..counted)
length(marcelloCAZy)
egCAZy <- unique(dbCANPyuuseqs$cazyFam)
length(egCAZy)
newCAZy2016 <- setdiff(egCAZy, marcelloCAZy)
length(newCAZy2016)


```
